{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dc5a27c0-9172-4f92-bc69-6bb64436d116",
   "metadata": {},
   "outputs": [],
   "source": [
    "from oligopoolio import * "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "86a7550c-9c60-4f98-a855-ccaa4a09025b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyswarms as ps\n",
    "import numpy as np\n",
    "import itertools\n",
    "from primer3 import calc_hairpin, calc_homodimer\n",
    "from functools import partial\n",
    "\n",
    "def gc_content(seq):\n",
    "    return (seq.count('G') + seq.count('C')) / len(seq)\n",
    "\n",
    "def objective_function(x, seq):\n",
    "    penalties = []\n",
    "    for row in x:\n",
    "        # The row contains the specified cut sites for the sequence \n",
    "        cut_sites = row\n",
    "        # \n",
    "        results = check_secondary_structure(seq)\n",
    "        homodimer_tm = results['homodimer']['homodimer_dg']\n",
    "        hairpin_tm = results['hairpin']['hairpin_dg']\n",
    "        \n",
    "        # Penalize if Tm deviates from 60\n",
    "        primer_tm = primer3.bindings.calcTm(seq)\n",
    "\n",
    "        tm_penalty = np.abs(primer_tm - tm_optimal)\n",
    "        \n",
    "        # Penalize if GC content is far from 0.5\n",
    "        gc_penalty = np.abs(gc - gc_optimal)*30\n",
    "        # Example: simple penalty for homopolymer runs\n",
    "        homodimer_tm = 0 if homodimer_tm > 0 else np.abs(homodimer_tm)\n",
    "        penalties.append(tm_penalty + gc_penalty + homodimer_tm + clamp_penalty)\n",
    "    return np.array(penalties)\n",
    "\n",
    "    \n",
    "def check_secondary_structure(sequence, temp=55):\n",
    "    \"\"\"\n",
    "    Check secondary structures like hairpins and homodimers in a given primer sequence.\n",
    "    \n",
    "    Args:\n",
    "        sequence (str): The DNA sequence of the primer to analyze.\n",
    "        \n",
    "    Returns:\n",
    "        dict: Results for hairpin and homodimer properties.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Check for hairpin structure\n",
    "        hairpin_result = calc_hairpin(sequence, temp_c=temp)\n",
    "        hairpin_info = {\n",
    "            \"hairpin_found\": hairpin_result.structure_found,\n",
    "            \"hairpin_tm\": hairpin_result.tm,\n",
    "            \"hairpin_dg\": hairpin_result.dg/1000,\n",
    "            \"hairpin_dh\": hairpin_result.dh/1000,\n",
    "            \"hairpin_ds\": hairpin_result.ds,\n",
    "        }\n",
    "\n",
    "        # Check for homodimer structure\n",
    "        homodimer_result = calc_homodimer(sequence, temp_c=temp)\n",
    "        homodimer_info = {\n",
    "            \"homodimer_found\": homodimer_result.structure_found,\n",
    "            \"homodimer_tm\": homodimer_result.tm,\n",
    "            \"homodimer_dg\": homodimer_result.dg/1000,\n",
    "            \"homodimer_dh\": homodimer_result.dh/1000,\n",
    "            \"homodimer_ds\": homodimer_result.ds,\n",
    "        }\n",
    "    except Exception as e:\n",
    "        u.warn_p([f\"Warning: issue with secondary structure check. \", sequence, e])\n",
    "        hairpin_info = {\"hairpin_found\": False, \"hairpin_tm\": None, \"hairpin_dg\": None, \"hairpin_dh\": None, \"hairpin_ds\": None}\n",
    "        homodimer_info = {\"homodimer_found\": False, \"homodimer_tm\": None, \"homodimer_dg\": None, \"homodimer_dh\": None, \"homodimer_ds\": None}\n",
    "    # Combine results\n",
    "    return {\"hairpin\": hairpin_info, \"homodimer\": homodimer_info}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95901350-7da9-4e6e-b286-d84941894797",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sciutil import SciUtil\n",
    "import random\n",
    "import Levenshtein\n",
    "from collections import defaultdict\n",
    "\n",
    "u = SciUtil()\n",
    "\n",
    "gc_optimal = 0.5\n",
    "tm_optimal = 67\n",
    "structure_dg = 0 # aim for greater than 0\n",
    "\n",
    "# Probably best to use something like pyswarm \n",
    "def objective_function(x, seq):\n",
    "    penalties = []\n",
    "    for row in x:\n",
    "        # The row contains the specified cut sites for the sequence \n",
    "        cut_sites = row\n",
    "        # \n",
    "        results = check_secondary_structure(seq)\n",
    "        homodimer_tm = results['homodimer']['homodimer_dg']\n",
    "        hairpin_tm = results['hairpin']['hairpin_dg']\n",
    "        \n",
    "        # Penalize if Tm deviates from 60\n",
    "        primer_tm = primer3.bindings.calcTm(seq)\n",
    "\n",
    "        tm_penalty = np.abs(primer_tm - tm_optimal)\n",
    "        \n",
    "        # Penalize if GC content is far from 0.5\n",
    "        gc_penalty = np.abs(gc - gc_optimal)*30\n",
    "        # Example: simple penalty for homopolymer runs\n",
    "        homodimer_tm = 0 if homodimer_tm > 0 else np.abs(homodimer_tm)\n",
    "        penalties.append(tm_penalty + gc_penalty + homodimer_tm + clamp_penalty)\n",
    "    return np.array(penalties)\n",
    "    \n",
    "options = {'c1': 0.5, 'c2': 0.3, 'w': 0.9} \n",
    "seq_ids = defaultdict(list)\n",
    "all_sequences = defaultdict(list)\n",
    "\n",
    "for seq in seqs:\n",
    "    seq_len = len(seq)\n",
    "    num_splits = int(math.ceil(seq_len / min_segment_length))\n",
    "    # Make sure it is even\n",
    "    if num_splits % 2 == 0:\n",
    "        num_splits += 1\n",
    "        \n",
    "    bounds = []\n",
    "\n",
    "    optimizer = ps.single.GlobalBestPSO(\n",
    "        n_particles=50,\n",
    "        dimensions=primer_len,\n",
    "        options=options,\n",
    "        bounds=bounds\n",
    "    )\n",
    "    \n",
    "    best_cost, best_pos = optimizer.optimize(partial(objective_function, seq=sample), iters=200)\n",
    "    \n",
    "except:\n",
    "    print(sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "id": "c2900399-def5-4c46-ae59-30a1f7a16e38",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pyswarms as ps\n",
    "from functools import partial\n",
    "\n",
    "\n",
    "\n",
    "# Helper to compute fragment start-end pairs\n",
    "def get_fragments(cuts, seq_len, num_fragments):\n",
    "    cuts = np.sort(np.round(cuts).astype(int))\n",
    "    starts = [0] + list(cuts)\n",
    "    fragments = []\n",
    "\n",
    "    for i in range(num_fragments):\n",
    "        start = starts[i]\n",
    "        # estimate end based on next start minus overlap\n",
    "        if i < num_fragments - 1:\n",
    "            next_start = starts[i + 1]\n",
    "            end = next_start + max_overlap\n",
    "        else:\n",
    "            end = seq_len\n",
    "        end = min(end, seq_len)\n",
    "        fragments.append((start, end))\n",
    "    return fragments\n",
    "\n",
    "# Objective: minimize fragment length variance, and penalize invalid overlap\n",
    "def objective_function(x, seq_len, num_fragments, seq, tm_optimal=62):\n",
    "    penalties = []\n",
    "    for particle in x:\n",
    "        fragments = get_fragments(particle, seq_len, num_fragments)\n",
    "        frag_lengths = [end - start for start, end in fragments]\n",
    "        \n",
    "        # Overlap check\n",
    "        olap_penalty = 0\n",
    "        # Also keep track of the overlaps so that we can select ones which aren't similar\n",
    "        overlaps = [] \n",
    "        temp_penalty = 0\n",
    "        seq_len_penalty = 0\n",
    "        homodimer_penalty = 0\n",
    "        repeat_penalty = 0\n",
    "        for i in range(len(fragments) - 1):\n",
    "            start1, end1 = fragments[i]\n",
    "            start2, end2 = fragments[i + 1]\n",
    "            overlap = end1 - start2\n",
    "            # Check the temperature of the overlap and the dimerization \n",
    "            results = check_secondary_structure(seq[start2: end1])\n",
    "            homodimer_tm = results['homodimer']['homodimer_dg']\n",
    "            hairpin_tm = results['hairpin']['hairpin_dg']\n",
    "            \n",
    "            # Penalize if Tm deviates from 60\n",
    "            primer_tm = primer3.bindings.calcTm(seq[start2: end1])\n",
    "            \n",
    "            tm_penalty = np.abs(primer_tm - tm_optimal)\n",
    "            if overlap < min_overlap or overlap > max_overlap:\n",
    "                olap_penalty += 10 * abs(overlap - (min_overlap + max_overlap)/2)\n",
    "\n",
    "            # Have a tolerance for the TM (i.e. is 6C ok?)\n",
    "            if tm_penalty > 5:\n",
    "                temp_penalty += 2*tm_penalty\n",
    "            else:\n",
    "                temp_penalty += tm_penalty\n",
    "            if end1 - start1 > max_seq_len:# or end1 - start1 < min_seq_len:\n",
    "                seq_len_penalty += 10*abs(max_seq_len- (end1 - start1))\n",
    "            elif end1 - start1 < min_seq_len: \n",
    "                seq_len_penalty += 10*abs(min_seq_len- (end1 - start1))\n",
    "            if end2 - start2 > max_seq_len:# or end1 - start1 < min_seq_len:\n",
    "                seq_len_penalty += 10*abs(max_seq_len- (end2 - start2))\n",
    "            elif end2 - start2 < min_seq_len: \n",
    "                seq_len_penalty += 10*abs(min_seq_len- (end2 - start2)) \n",
    "            if homodimer_tm < -3:\n",
    "                # Add an extra penalty if it's outside the range\n",
    "                homodimer_penalty += 2*abs(homodimer_tm)\n",
    "            else:\n",
    "                homodimer_penalty += abs(homodimer_tm)\n",
    "            overlap_seq = seq[start2: end1]\n",
    "            if overlap_seq[:3] in ['AAA', 'TTT', 'CCC', 'GGG'] or overlap_seq[-3:] in ['AAA', 'TTT', 'CCC', 'GGG']:\n",
    "                repeat_penalty += 20\n",
    "            overlaps.append(seq[start2: end1])\n",
    "        \n",
    "        # Want to do levenstein distance and make sure there is at least an edit distance of 8 otherwise these are too similar.\n",
    "        min_dist = 100000\n",
    "        for i, x in enumerate(overlaps):\n",
    "            for j, y in enumerate(overlaps):\n",
    "                if i != j:\n",
    "                    dist = Levenshtein.distance(x, y)\n",
    "                    if dist < min_dist:\n",
    "                        min_dist = dist\n",
    "        dist_penalty = 27 - min_dist\n",
    "        if min_dist < 8:\n",
    "            dist_penalty += 27\n",
    "        # Variance penalty\n",
    "        mean_len = np.mean(frag_lengths)\n",
    "        variance = np.sqrt(np.sum([(l - mean_len)**2 for l in frag_lengths]))\n",
    "        penalties.append(variance + olap_penalty + homodimer_penalty + temp_penalty + seq_len_penalty + dist_penalty + repeat_penalty)\n",
    "    return np.array(penalties)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "b72e48b8-b72d-455d-b5d2-e7799926c030",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-06-24 11:32:36,157 - pyswarms.single.global_best - INFO - Optimize for 100 iters with {'c1': 0.5, 'c2': 0.5, 'w': 0.7}\n",
      "pyswarms.single.global_best: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████|100/100, best_cost=82.4\n",
      "2025-06-24 11:33:01,879 - pyswarms.single.global_best - INFO - Optimization finished | best cost: 82.36328492184145, best pos: [204.04128056 102.4249867  644.49720023 380.0298606  740.12362654\n",
      " 294.55862508 552.4081172  454.93127964 837.87533045]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Best cut sites: [(0, np.int64(129)), (np.int64(102), np.int64(231)), (np.int64(204), np.int64(322)), (np.int64(295), np.int64(407)), (np.int64(380), np.int64(482)), (np.int64(455), np.int64(579)), (np.int64(552), np.int64(671)), (np.int64(644), np.int64(767)), (np.int64(740), np.int64(865)), (np.int64(838), 958)]\n",
      "Fragments: ['gaaataattttgtttaactttaagaaggagatatacatATGCGCAAGCTTTCTCGTTGTGAGGACGGCTACGTCAAGATTCAGGGGATCTACATTTACTATAAAGTGTGCAAAGCCGAGAATGAGAAAG', 'AAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCACGGTGGACCCGGCATGAGCCACGACTACTTGCTTTCCCTGACCGATTTAGCTGAGAAGGGGATTACGGTTTTATTCTATGACC', 'AGGGGATTACGGTTTTATTCTATGACCAATTCGGCTGTGGACGTTCAGAAGAACCTGAGAAGGAGAAATTTACGATCGACTATGGAGTGGAAGAAGCAGAAGCGGTGAAAAAAAACAT', 'AGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTATTCCTGATGGGATCGTCCTATGGTGGGGCTTTAGCTCTGGCCTACGCTGTTAAGTATCAAGCACAT', 'GCCTACGCTGTTAAGTATCAAGCACATTTGAAGGGCCTGATCATCTCGGGCGGATTATCATCTGTGCCTCTGACAGTAAAAGAAATGCAGCGCTTGATTGAC', 'GTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACCGCAACGCGATCCGCAAGTATGGCGAAGTGGGGGATTATCAGAACCCTGAATACCAGGAGGCTGTGAACTATTTCTATC', 'ACCAGGAGGCTGTGAACTATTTCTATCACCAGCATTTGTTACGCTCTGAAGACTGGCCCCCCGAGGTTTTGAAAAGCCTGGAGTACGCAGAAGAGCGCAATGTGTACCGCACAATGAAC', 'GAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAATCACTGGAACTATCCGCGACTGGGATATCACCGATAAGATCGGGATTATCAGTGTTCCTACTCTTATTACGGTCGGT', 'AGTGTTCCTACTCTTATTACGGTCGGTGAGTTTGACGAGGTAACGCAAAATGTCGCAGAAGTTATCCACTCCAAAATCGATAATTCCCAGTTAATTGTATTTAAAGCATGTAGCCACCTGACCAT', 'ATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAGTACAACCGCATTTTATTACAGTTCATCGAAAAGAACATTCTCGAGCACCACCACCACCACCACTGAgatccggc']\n"
     ]
    }
   ],
   "source": [
    "import Levenshtein\n",
    "\n",
    "# Parameters\n",
    "seq = \"gaaataattttgtttaactttaagaaggagatatacatATGCGCAAGCTTTCTCGTTGTGAGGACGGCTACGTCAAGATTCAGGGGATCTACATTTACTATAAAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCACGGTGGACCCGGCATGAGCCACGACTACTTGCTTTCCCTGACCGATTTAGCTGAGAAGGGGATTACGGTTTTATTCTATGACCAATTCGGCTGTGGACGTTCAGAAGAACCTGAGAAGGAGAAATTTACGATCGACTATGGAGTGGAAGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTATTCCTGATGGGATCGTCCTATGGTGGGGCTTTAGCTCTGGCCTACGCTGTTAAGTATCAAGCACATTTGAAGGGCCTGATCATCTCGGGCGGATTATCATCTGTGCCTCTGACAGTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACCGCAACGCGATCCGCAAGTATGGCGAAGTGGGGGATTATCAGAACCCTGAATACCAGGAGGCTGTGAACTATTTCTATCACCAGCATTTGTTACGCTCTGAAGACTGGCCCCCCGAGGTTTTGAAAAGCCTGGAGTACGCAGAAGAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAATCACTGGAACTATCCGCGACTGGGATATCACCGATAAGATCGGGATTATCAGTGTTCCTACTCTTATTACGGTCGGTGAGTTTGACGAGGTAACGCAAAATGTCGCAGAAGTTATCCACTCCAAAATCGATAATTCCCAGTTAATTGTATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAGTACAACCGCATTTTATTACAGTTCATCGAAAAGAACATTCTCGAGCACCACCACCACCACCACTGAgatccggc\"\n",
    "seq_len = len(seq)\n",
    "min_overlap = 16\n",
    "max_overlap = 27\n",
    "max_seq_len = 130\n",
    "min_seq_len = 80\n",
    "optimal_seq_len = 100\n",
    "\n",
    "num_fragments = int(math.ceil(seq_len / optimal_seq_len))\n",
    "# Make sure it is even\n",
    "if num_fragments % 2 == 1:\n",
    "    num_fragments += 1\n",
    "        \n",
    "# Define bounds for cuts\n",
    "num_cuts = num_fragments - 1\n",
    "lower_bounds = np.arange(1, num_cuts + 1) * 2  # avoid first char\n",
    "upper_bounds = np.full(num_cuts, seq_len - 2)\n",
    "\n",
    "# Run PSO\n",
    "bounds = (lower_bounds, upper_bounds)\n",
    "optimizer = ps.single.GlobalBestPSO(\n",
    "    n_particles=150,\n",
    "    dimensions=num_cuts,\n",
    "    options={'c1': 0.5, 'c2': 0.5, 'w': 0.7},\n",
    "    bounds=bounds\n",
    ")\n",
    "\n",
    "best_cost, best_pos = optimizer.optimize(partial(objective_function, seq_len=seq_len, num_fragments=num_fragments, seq=seq), iters=100)\n",
    "\n",
    "# Decode result\n",
    "cuts = np.sort(np.round(best_pos).astype(int))\n",
    "fragments = get_fragments(cuts, seq_len, num_fragments)\n",
    "fragment_strings = [seq[start:end] for start, end in fragments]\n",
    "\n",
    "print(\"Best cut sites:\", fragments)\n",
    "print(\"Fragments:\", fragment_strings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "5f11473d-c113-4def-ba4a-1c7469a595e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = \"ATGCGCAAGCTTTCTCGTTGTGAGGACGGCTACGTCAAGATTCAGGGGATCTACATTTACTATAAAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCACGGTGGACCCGGCATGAGCCACGACTACTTGCTTTCCCTGACCGATTTAGCTGAGAAGGGGATTACGGTTTTATTCTATGACCAATTCGGCTGTGGACGTTCAGAAGAACCTGAGAAGGAGAAATTTACGATCGACTATGGAGTGGAAGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTATTCCTGATGGGATCGTCCTATGGTGGGGCTTTAGCTCTGGCCTACGCTGTTAAGTATCAAGCACATTTGAAGGGCCTGATCATCTCGGGCGGATTATCATCTGTGCCTCTGACAGTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACCGCAACGCGATCCGCAAGTATGGCGAAGTGGGGGATTATCAGAACCCTGAATACCAGGAGGCTGTGAACTATTTCTATCACCAGCATTTGTTACGCTCTGAAGACTGGCCCCCCGAGGTTTTGAAAAGCCTGGAGTACGCAGAAGAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAATCACTGGAACTATCCGCGACTGGGATATCACCGATAAGATCGGGATTATCAGTGTTCCTACTCTTATTACGGTCGGTGAGTTTGACGAGGTAACGCAAAATGTCGCAGAAGTTATCCACTCCAAAATCGATAATTCCCAGTTAATTGTATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAGTACAACCGCATTTTATTACAGTTCATCGAAAAGAACATTCTCGAGCACCACCACCACCACCACTGA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "5e62e385-5ebd-4a51-822a-05463db05d2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the overlaps by making the plasmid map\n",
    "from oligopoolio import * \n",
    "oligos = fragment_strings\n",
    "output_file = 'test.gb'\n",
    "prev_oligo = None\n",
    "genbank_file = 'base-pet22b-base-anm.gb'\n",
    "seq_id = 'test'\n",
    "# Add in the optimzed sequence\n",
    "translation_label = f\"Insert_{seq_id}\"\n",
    "reverse = False  # Set to True for reverse feature, False for forward\n",
    "insert_position = 5193\n",
    "reverse = False\n",
    "record = insert_sequence_with_translation(genbank_file, None, insert_position, seq, translation_label, reverse)\n",
    "bsa = False\n",
    "rows = []\n",
    "# If a genbank file was provided also just add in the new sequnece\n",
    "if len(oligos) > 0:\n",
    "    for i, oligo in enumerate(oligos):\n",
    "        seq = oligo\n",
    "        # CHeck that there is an overlap with the previous sequence and that it is not too short\n",
    "        # Also make sure we swap the directions of the oligos so they automatically anneal\n",
    "        # Also assert that the start is a methionine (and if not warn it... )\n",
    "        primer_overlap = None\n",
    "        primer_tm = None\n",
    "        primer_len = None\n",
    "        homodimer_tm = None\n",
    "        hairpin_tm = None\n",
    "        if prev_oligo:  \n",
    "            # Get the overlap with the previous sequence\n",
    "            match = SequenceMatcher(None, prev_oligo, seq).find_longest_match()\n",
    "            primer_overlap = prev_oligo[match.a:match.a + match.size]\n",
    "            # Analyze the primer sequence\n",
    "            results = check_secondary_structure(primer_overlap)\n",
    "            homodimer_tm = results['homodimer']['homodimer_dg']\n",
    "            hairpin_tm = results['hairpin']['hairpin_dg']\n",
    "            primer_tm = primer3.bindings.calcTm(primer_overlap)\n",
    "            primer_len = len(primer_overlap)\n",
    "\n",
    "        prev_oligo = seq\n",
    "        orig_seq = seq\n",
    "        strand = 1\n",
    "        # Changed this to be the opposite To\n",
    "        if bsa == True:\n",
    "            if i % 2 == 1 and i != 0:\n",
    "                seq = str(Seq(seq).reverse_complement())\n",
    "                strand = -1\n",
    "        else:\n",
    "            if i % 2 == 0:\n",
    "                seq = str(Seq(seq).reverse_complement())\n",
    "                strand = -1\n",
    "        oligo_tm = primer3.bindings.calcTm(seq)\n",
    "        #if genbank_file:\n",
    "        insert_features_from_oligos(record, f\"{seq_id}_oligo_{i}\", orig_seq, strand, oligo_tm, None)\n",
    "        rows.append([seq_id, oligo, seq, len(seq), oligo_tm, strand, primer_overlap, primer_tm, primer_len, homodimer_tm, hairpin_tm, oligo[2]])\n",
    "    if genbank_file:\n",
    "        output_file = genbank_file.replace('.', f'_{seq_id}.')\n",
    "        if genbank_file:\n",
    "            record.name = seq_id  # type: ignore \n",
    "            SeqIO.write(record, f'{output_file}', \"genbank\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "295d30e7-23a8-4570-bce5-c1f22927baa7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>oligo_id</th>\n",
       "      <th>oligo_sequence</th>\n",
       "      <th>oligo_length</th>\n",
       "      <th>oligo_tm</th>\n",
       "      <th>strand</th>\n",
       "      <th>primer_overlap_with_previous</th>\n",
       "      <th>overlap_tm_5prime</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>overlap_homodimer_dg</th>\n",
       "      <th>overlap_hairpin_dg</th>\n",
       "      <th>original_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>test</td>\n",
       "      <td>gaaataattttgtttaactttaagaaggagatatacatATGCGCAA...</td>\n",
       "      <td>CTTTCTCATTCTCGGCTTTGCACACTTTATAGTAAATGTAGATCCC...</td>\n",
       "      <td>129</td>\n",
       "      <td>74.650485</td>\n",
       "      <td>-1</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>test</td>\n",
       "      <td>AAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCA...</td>\n",
       "      <td>AAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCA...</td>\n",
       "      <td>129</td>\n",
       "      <td>81.704538</td>\n",
       "      <td>1</td>\n",
       "      <td>AAGTGTGCAAAGCCGAGAATGAGAAAG</td>\n",
       "      <td>62.231401</td>\n",
       "      <td>27.0</td>\n",
       "      <td>-2.210118</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>test</td>\n",
       "      <td>AGGGGATTACGGTTTTATTCTATGACCAATTCGGCTGTGGACGTTC...</td>\n",
       "      <td>ATGTTTTTTTTCACCGCTTCTGCTTCTTCCACTCCATAGTCGATCG...</td>\n",
       "      <td>118</td>\n",
       "      <td>78.597154</td>\n",
       "      <td>-1</td>\n",
       "      <td>AGGGGATTACGGTTTTATTCTATGACC</td>\n",
       "      <td>58.534689</td>\n",
       "      <td>27.0</td>\n",
       "      <td>-0.959718</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>test</td>\n",
       "      <td>AGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTA...</td>\n",
       "      <td>AGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTA...</td>\n",
       "      <td>112</td>\n",
       "      <td>79.263673</td>\n",
       "      <td>1</td>\n",
       "      <td>AGAAGCAGAAGCGGTGAAAAAAAACAT</td>\n",
       "      <td>60.752486</td>\n",
       "      <td>27.0</td>\n",
       "      <td>-1.666098</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>test</td>\n",
       "      <td>GCCTACGCTGTTAAGTATCAAGCACATTTGAAGGGCCTGATCATCT...</td>\n",
       "      <td>GTCAATCAAGCGCTGCATTTCTTTTACTGTCAGAGGCACAGATGAT...</td>\n",
       "      <td>102</td>\n",
       "      <td>79.737765</td>\n",
       "      <td>-1</td>\n",
       "      <td>GCCTACGCTGTTAAGTATCAAGCACAT</td>\n",
       "      <td>61.495744</td>\n",
       "      <td>27.0</td>\n",
       "      <td>-1.336169</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>test</td>\n",
       "      <td>GTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACC...</td>\n",
       "      <td>GTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACC...</td>\n",
       "      <td>124</td>\n",
       "      <td>80.997246</td>\n",
       "      <td>1</td>\n",
       "      <td>GTAAAAGAAATGCAGCGCTTGATTGAC</td>\n",
       "      <td>60.276860</td>\n",
       "      <td>27.0</td>\n",
       "      <td>-4.656577</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>test</td>\n",
       "      <td>ACCAGGAGGCTGTGAACTATTTCTATCACCAGCATTTGTTACGCTC...</td>\n",
       "      <td>GTTCATTGTGCGGTACACATTGCGCTCTTCTGCGTACTCCAGGCTT...</td>\n",
       "      <td>119</td>\n",
       "      <td>81.968434</td>\n",
       "      <td>-1</td>\n",
       "      <td>ACCAGGAGGCTGTGAACTATTTCTATC</td>\n",
       "      <td>60.117426</td>\n",
       "      <td>27.0</td>\n",
       "      <td>-2.838108</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>test</td>\n",
       "      <td>GAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAA...</td>\n",
       "      <td>GAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAA...</td>\n",
       "      <td>123</td>\n",
       "      <td>80.781631</td>\n",
       "      <td>1</td>\n",
       "      <td>GAGCGCAATGTGTACCGCACAATGAAC</td>\n",
       "      <td>65.174207</td>\n",
       "      <td>27.0</td>\n",
       "      <td>-4.342888</td>\n",
       "      <td>-0.226057</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>test</td>\n",
       "      <td>AGTGTTCCTACTCTTATTACGGTCGGTGAGTTTGACGAGGTAACGC...</td>\n",
       "      <td>ATGGTCAGGTGGCTACATGCTTTAAATACAATTAACTGGGAATTAT...</td>\n",
       "      <td>125</td>\n",
       "      <td>78.231318</td>\n",
       "      <td>-1</td>\n",
       "      <td>AGTGTTCCTACTCTTATTACGGTCGGT</td>\n",
       "      <td>60.971734</td>\n",
       "      <td>27.0</td>\n",
       "      <td>0.392122</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>test</td>\n",
       "      <td>ATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAG...</td>\n",
       "      <td>ATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAG...</td>\n",
       "      <td>120</td>\n",
       "      <td>78.734479</td>\n",
       "      <td>1</td>\n",
       "      <td>ATTTAAAGCATGTAGCCACCTGACCAT</td>\n",
       "      <td>60.985601</td>\n",
       "      <td>27.0</td>\n",
       "      <td>-1.096238</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     id                                           oligo_id  \\\n",
       "0  test  gaaataattttgtttaactttaagaaggagatatacatATGCGCAA...   \n",
       "1  test  AAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCA...   \n",
       "2  test  AGGGGATTACGGTTTTATTCTATGACCAATTCGGCTGTGGACGTTC...   \n",
       "3  test  AGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTA...   \n",
       "4  test  GCCTACGCTGTTAAGTATCAAGCACATTTGAAGGGCCTGATCATCT...   \n",
       "5  test  GTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACC...   \n",
       "6  test  ACCAGGAGGCTGTGAACTATTTCTATCACCAGCATTTGTTACGCTC...   \n",
       "7  test  GAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAA...   \n",
       "8  test  AGTGTTCCTACTCTTATTACGGTCGGTGAGTTTGACGAGGTAACGC...   \n",
       "9  test  ATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAG...   \n",
       "\n",
       "                                      oligo_sequence  oligo_length   oligo_tm  \\\n",
       "0  CTTTCTCATTCTCGGCTTTGCACACTTTATAGTAAATGTAGATCCC...           129  74.650485   \n",
       "1  AAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCA...           129  81.704538   \n",
       "2  ATGTTTTTTTTCACCGCTTCTGCTTCTTCCACTCCATAGTCGATCG...           118  78.597154   \n",
       "3  AGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTA...           112  79.263673   \n",
       "4  GTCAATCAAGCGCTGCATTTCTTTTACTGTCAGAGGCACAGATGAT...           102  79.737765   \n",
       "5  GTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACC...           124  80.997246   \n",
       "6  GTTCATTGTGCGGTACACATTGCGCTCTTCTGCGTACTCCAGGCTT...           119  81.968434   \n",
       "7  GAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAA...           123  80.781631   \n",
       "8  ATGGTCAGGTGGCTACATGCTTTAAATACAATTAACTGGGAATTAT...           125  78.231318   \n",
       "9  ATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAG...           120  78.734479   \n",
       "\n",
       "   strand primer_overlap_with_previous  overlap_tm_5prime  overlap_length  \\\n",
       "0      -1                         None                NaN             NaN   \n",
       "1       1  AAGTGTGCAAAGCCGAGAATGAGAAAG          62.231401            27.0   \n",
       "2      -1  AGGGGATTACGGTTTTATTCTATGACC          58.534689            27.0   \n",
       "3       1  AGAAGCAGAAGCGGTGAAAAAAAACAT          60.752486            27.0   \n",
       "4      -1  GCCTACGCTGTTAAGTATCAAGCACAT          61.495744            27.0   \n",
       "5       1  GTAAAAGAAATGCAGCGCTTGATTGAC          60.276860            27.0   \n",
       "6      -1  ACCAGGAGGCTGTGAACTATTTCTATC          60.117426            27.0   \n",
       "7       1  GAGCGCAATGTGTACCGCACAATGAAC          65.174207            27.0   \n",
       "8      -1  AGTGTTCCTACTCTTATTACGGTCGGT          60.971734            27.0   \n",
       "9       1  ATTTAAAGCATGTAGCCACCTGACCAT          60.985601            27.0   \n",
       "\n",
       "   overlap_homodimer_dg  overlap_hairpin_dg original_sequence  \n",
       "0                   NaN                 NaN                 a  \n",
       "1             -2.210118            0.000000                 G  \n",
       "2             -0.959718            0.000000                 G  \n",
       "3             -1.666098            0.000000                 A  \n",
       "4             -1.336169            0.000000                 C  \n",
       "5             -4.656577            0.000000                 A  \n",
       "6             -2.838108            0.000000                 C  \n",
       "7             -4.342888           -0.226057                 G  \n",
       "8              0.392122            0.000000                 T  \n",
       "9             -1.096238            0.000000                 T  "
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oligo_df = pd.DataFrame(rows, columns=[\"id\", \"oligo_id\", \"oligo_sequence\", \"oligo_length\",  \"oligo_tm\", \"strand\",\"primer_overlap_with_previous\", \"overlap_tm_5prime\", \"overlap_length\", \n",
    "                                            \"overlap_homodimer_dg\", \"overlap_hairpin_dg\", \"original_sequence\"])\n",
    "oligo_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "a9bc2d6e-483d-4ade-95b5-5cfb2e6a0d4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AAGTGTGCAAAGCCGAGAATGAGAAAG',\n",
       " 'AGGGGATTACGGTTTTATTCTATGACC',\n",
       " 'AGAAGCAGAAGCGGTGAAAAAAAACAT',\n",
       " 'TAGCTCTGGCCTACGCTGTTAAGTATC',\n",
       " 'CGGATTATCATCTGTGCCTCTGACAGT',\n",
       " 'TGGGGGATTATCAGAACCCTGAATACC',\n",
       " 'CTGGAGTACGCAGAAGAGCGCAATGTG',\n",
       " 'GATCGGGATTATCAGTGTTCCTACTCT',\n",
       " 'GTTAATTGTATTTAAAGCATGTAGCCA']"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "overlaps = [x for x in oligo_df['primer_overlap_with_previous'].values if x != None]\n",
    "overlaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "db56d5ee-3602-46f7-8b5c-419feafd4b0e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Want to do levenstein distance and make sure there is at least an edit distance of 8 otherwise these are too similar.\n",
    "min_dist = 100000\n",
    "for i, x in enumerate(overlaps):\n",
    "    for j, y in enumerate(overlaps):\n",
    "        if i != j:\n",
    "            dist = Levenshtein.distance(x, y)\n",
    "            if dist < min_dist:\n",
    "                min_dist = dist\n",
    "dist_penalty = 0\n",
    "if dist < 8:\n",
    "    dist_penalty = 100\n",
    "    print(min_dist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "cddf0370-4b55-4b1c-b197-0d974982658f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min_dist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e058085-462a-4b95-b30a-c3d658939277",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dnaweaver as dw\n",
    "import time\n",
    "import dnachisel as dnachisel \n",
    "from sciutil import SciUtil\n",
    "from Bio.Seq import Seq\n",
    "from difflib import SequenceMatcher\n",
    "import primer3\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import pandas as pd\n",
    "import math\n",
    "from dnachisel import *\n",
    "\n",
    "u = SciUtil()\n",
    "\n",
    "from primer3 import calc_hairpin, calc_homodimer\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "def insert_sequence_with_translation(input_file, output_file, insert_position, new_sequence, translation_label, reverse=False):\n",
    "    \"\"\"\n",
    "    Insert a new sequence at a specific position in a GenBank file, add a translation annotation,\n",
    "    and adjust existing annotations to avoid overlap.\n",
    "\n",
    "    Args:\n",
    "        input_file (str): Path to the original GenBank file.\n",
    "        output_file (str): Path to save the modified GenBank file.\n",
    "        insert_position (int): Position to insert the new sequence (0-based).\n",
    "        new_sequence (str): The DNA sequence to insert.\n",
    "        translation_label (str): Label for the translation feature.\n",
    "        reverse (bool): Whether the feature should be on the reverse strand.\n",
    "    \"\"\"\n",
    "    # Read the original GenBank file\n",
    "    record = SeqIO.read(input_file, \"genbank\")\n",
    "    \n",
    "    # Insert the new sequence at the specified position\n",
    "    if reverse:\n",
    "        new_sequence = str(Seq(new_sequence).reverse_complement())  # Reverse complement the sequence if needed\n",
    "    record.seq = record.seq[:insert_position] + Seq(new_sequence) + record.seq[insert_position:]\n",
    "    \n",
    "    # Adjust existing annotations to avoid overlap\n",
    "    inserted_length = len(new_sequence)\n",
    "    for feature in record.features:\n",
    "        if feature.location.start >= insert_position:\n",
    "            feature.location = FeatureLocation(\n",
    "                feature.location.start + inserted_length,\n",
    "                feature.location.end + inserted_length,\n",
    "                strand=feature.location.strand\n",
    "            )\n",
    "    \n",
    "    # Create the feature label\n",
    "    strand_label = \" (reverse)\" if reverse else \" (forward)\"\n",
    "    full_label = translation_label + strand_label\n",
    "    \n",
    "    # Add a feature for the inserted sequence\n",
    "    start = insert_position\n",
    "    end = insert_position + len(new_sequence)\n",
    "    feature = SeqFeature(\n",
    "        location=FeatureLocation(start, end, strand=-1 if reverse else 1),\n",
    "        type=\"CDS\",  # CDS type for coding sequences\n",
    "        qualifiers={\n",
    "            \"label\": full_label,\n",
    "            \"translation\": Seq(new_sequence).translate(table=11)  # Translation for the sequence\n",
    "        }\n",
    "    )\n",
    "    record.features.append(feature)\n",
    "    # Save the modified GenBank file\n",
    "    if output_file:\n",
    "        SeqIO.write(record, output_file, \"genbank\")\n",
    "        print(f\"Updated GenBank file saved as {output_file}\")\n",
    "    \n",
    "    return record\n",
    "\n",
    "def insert_features_from_oligos(record, seq_id, seq, strand, tm, output_file):\n",
    "    \"\"\"\n",
    "    Insert features into a GenBank file based on oligo_df.\n",
    "\n",
    "    Args:\n",
    "        genbank_file (str): Path to the original GenBank file.\n",
    "        output_file (str): Path to save the updated GenBank file.\n",
    "        oligo_df (pd.DataFrame): DataFrame with oligo information. \n",
    "                                 Expected columns: ['id', 'oligo_id', 'oligo_sequence', 'oligo_length', ...].\n",
    "    \"\"\"\n",
    "    # Iterate through the oligo DataFrame to add features\n",
    "    start = record.seq.find(seq.upper())\n",
    "    if start == -1:\n",
    "        print(f\"Warning: Oligo sequence {seq_id} not found in the GenBank sequence.\")\n",
    "    \n",
    "    end = start + len(seq)\n",
    "    feature = SeqFeature(\n",
    "        location=FeatureLocation(start, end, strand=strand),\n",
    "        type=\"misc_feature\",\n",
    "        qualifiers={\n",
    "            \"label\": f\"{seq_id} {'(reverse)' if strand == -1 else '(forward)'}\",\n",
    "            \"note\": f\"Length: {len(seq)}, TM: {tm}\"\n",
    "        }\n",
    "    )\n",
    "    record.features.append(feature)\n",
    "    \n",
    "    # Save the updated GenBank file\n",
    "    if output_file:\n",
    "        SeqIO.write(record, output_file, \"genbank\")\n",
    "        print(f\"Updated GenBank file saved as {output_file}\")\n",
    "    return record\n",
    "    \n",
    "def check_secondary_structure(sequence, temp=55):\n",
    "    \"\"\"\n",
    "    Check secondary structures like hairpins and homodimers in a given primer sequence.\n",
    "    \n",
    "    Args:\n",
    "        sequence (str): The DNA sequence of the primer to analyze.\n",
    "        \n",
    "    Returns:\n",
    "        dict: Results for hairpin and homodimer properties.\n",
    "    \"\"\"\n",
    "    try:\n",
    "        # Check for hairpin structure\n",
    "        hairpin_result = calc_hairpin(sequence, temp_c=temp)\n",
    "        hairpin_info = {\n",
    "            \"hairpin_found\": hairpin_result.structure_found,\n",
    "            \"hairpin_tm\": hairpin_result.tm,\n",
    "            \"hairpin_dg\": hairpin_result.dg/1000,\n",
    "            \"hairpin_dh\": hairpin_result.dh/1000,\n",
    "            \"hairpin_ds\": hairpin_result.ds,\n",
    "        }\n",
    "\n",
    "        # Check for homodimer structure\n",
    "        homodimer_result = calc_homodimer(sequence, temp_c=temp)\n",
    "        homodimer_info = {\n",
    "            \"homodimer_found\": homodimer_result.structure_found,\n",
    "            \"homodimer_tm\": homodimer_result.tm,\n",
    "            \"homodimer_dg\": homodimer_result.dg/1000,\n",
    "            \"homodimer_dh\": homodimer_result.dh/1000,\n",
    "            \"homodimer_ds\": homodimer_result.ds,\n",
    "        }\n",
    "    except Exception as e:\n",
    "        u.warn_p([f\"Warning: issue with secondary structure check. \", sequence, e])\n",
    "        hairpin_info = {\"hairpin_found\": False, \"hairpin_tm\": None, \"hairpin_dg\": None, \"hairpin_dh\": None, \"hairpin_ds\": None}\n",
    "        homodimer_info = {\"homodimer_found\": False, \"homodimer_tm\": None, \"homodimer_dg\": None, \"homodimer_dh\": None, \"homodimer_ds\": None}\n",
    "    # Combine results\n",
    "    return {\"hairpin\": hairpin_info, \"homodimer\": homodimer_info}\n",
    "\n",
    "def build_oligos(seq_id: str, sequence: str, output_directory: str, min_gc=0.3, max_gc=0.7, min_tm=55, max_tm=70, min_segment_length=40, max_segment_length=100, max_length=1500):\n",
    "    \"\"\" Use DNAweaver to build oligos \"\"\"\n",
    "    # Here we use a comercial supplier but don't actually care. \n",
    "    cheap_dna_offer = dw.CommercialDnaOffer(\n",
    "        name=\"CheapDNA.\",\n",
    "        sequence_constraints=[\n",
    "            dw.NoPatternConstraint(enzyme=\"BsaI\"),\n",
    "            dw.SequenceLengthConstraint(max_length=4000),\n",
    "            dw.GcContentConstraint(min_gc=min_gc, max_gc=max_gc)\n",
    "        ],\n",
    "        pricing=dw.PerBasepairPricing(0.10),\n",
    "    )\n",
    "\n",
    "    oligo_dna_offer = dw.CommercialDnaOffer(\n",
    "        name=\"OliGoo\",\n",
    "        sequence_constraints=[\n",
    "            dw.GcContentConstraint(min_gc=min_gc, max_gc=max_gc),\n",
    "            dw.SequenceLengthConstraint(max_length=4000),\n",
    "        ],\n",
    "        pricing=dw.PerBasepairPricing(0.07),\n",
    "        memoize=True\n",
    "    )\n",
    "\n",
    "    oligo_assembly_station = dw.DnaAssemblyStation(\n",
    "        name=\"Oligo Assembly Station\",\n",
    "        assembly_method=dw.OligoAssemblyMethod(\n",
    "            overhang_selector=dw.TmSegmentSelector(\n",
    "                min_size=15, max_size=25, min_tm=min_tm, max_tm=max_tm\n",
    "            ),\n",
    "            min_segment_length=min_segment_length,\n",
    "            max_segment_length=max_segment_length,\n",
    "            sequence_constraints=[dw.SequenceLengthConstraint(max_length=4000)],\n",
    "            duration=8,\n",
    "            cost=30,\n",
    "        ),\n",
    "        supplier=oligo_dna_offer,\n",
    "        coarse_grain=20,\n",
    "        a_star_factor=\"auto\",\n",
    "        memoize=True,\n",
    "    )\n",
    "\n",
    "    assembly_station = dw.DnaAssemblyStation(\n",
    "        name=\"Gibson Assembly Station\",\n",
    "        assembly_method=dw.GibsonAssemblyMethod(\n",
    "            overhang_selector=dw.TmSegmentSelector(min_tm=min_tm, max_tm=max_tm),\n",
    "            min_segment_length=min_segment_length,\n",
    "            max_segment_length=max_segment_length + 20, # add a bit of a buffer\n",
    "        ),\n",
    "        supplier=[cheap_dna_offer, oligo_assembly_station],\n",
    "        logger=\"bar\",\n",
    "        coarse_grain=100,\n",
    "        fine_grain=10,\n",
    "        a_star_factor=\"auto\",\n",
    "    )\n",
    "    \n",
    "    print(\"Looking for the best assembly plan...\")\n",
    "    t0 = time.time()\n",
    "    quote = assembly_station.get_quote(sequence, with_assembly_plan=True)\n",
    "    assembly_plan_report = quote.to_assembly_plan_report()\n",
    "    assembly_plan_report.write_full_report(f\"{output_directory}/oligo_assembly_plan_{seq_id}.zip\")\n",
    "    original_sequence = assembly_plan_report.plan.sequence\n",
    "    # Then get the sequence \n",
    "    rows = []\n",
    "    for oligo in assembly_plan_report.plan.assembly_plan:\n",
    "        # If this was chosen then choose it\n",
    "        if oligo.accepted:\n",
    "            rows.append([oligo.id, oligo.sequence, original_sequence])\n",
    "    return rows\n",
    "\n",
    "def get_oligos(df, protein_column, id_column, output_directory, forward_primer: str, reverse_primer: str, sequence_end: str, min_overlap=10, min_gc=0.3, \n",
    "               max_gc=0.7, min_tm=55, max_tm=70, min_segment_length=90, max_segment_length=130, max_length=1500, genbank_file=None,\n",
    "               insert_position=0, simple=False, optimize=True, bsa=False):\n",
    "    \"\"\" Get the oligos for a dataframe:\n",
    "    sequence_end is the end of the sequence i.e. TAA, TGA, etc or a histag \n",
    "    \"\"\"\n",
    "    rows = []   \n",
    "    record = None\n",
    "    for seq_id, protein_sequence in df[[id_column, protein_column]].values:\n",
    "        # Add on the primers that the user has provided\n",
    "        if optimize:\n",
    "            optimzed_sequence = codon_optimize(protein_sequence, min_gc, max_gc) + sequence_end\n",
    "        else:\n",
    "            optimzed_sequence = protein_sequence + sequence_end\n",
    "            u.dp([f\"Added sequence end with HIS and stop to: {seq_id}, {sequence_end}\"])\n",
    "\n",
    "        if genbank_file:\n",
    "            # Add in the optimzed sequence\n",
    "            translation_label = f\"Insert_{seq_id}\"\n",
    "            reverse = False  # Set to True for reverse feature, False for forward\n",
    "            record = insert_sequence_with_translation(genbank_file, None, insert_position, optimzed_sequence, translation_label, reverse)\n",
    "            \n",
    "        if optimzed_sequence[:3] != \"ATG\":\n",
    "            u.dp([f\"Warning: {seq_id} does not start with a methionine. \", optimzed_sequence[:3]])\n",
    "            if 'ATG' not in forward_primer:\n",
    "                u.warn_p([f\"Warning: {seq_id} does not start with a methionine. AND you don't have a methonine in your primer!!\", forward_primer])\n",
    "                print(\"We expect the primer to be in 5 to 3 prime direction.\")\n",
    "        # ALso check the end and or the reverse primer check for the three ones\n",
    "        if optimzed_sequence[-3:] != \"TAA\" and optimzed_sequence[-3:] != \"TGA\" and optimzed_sequence[-3:] != \"TAG\":\n",
    "            u.dp([f\"Warning: {seq_id} does not end with a stop codon. \", optimzed_sequence[-3:]])\n",
    "            if 'TAA' not in reverse_primer and \"TGA\" not in reverse_primer and \"TAG\" not in reverse_primer:\n",
    "                u.warn_p([f\"Warning: {seq_id} does not end with a stop codon. AND you don't have a stop codon in your primer!!\", reverse_primer])\n",
    "                print(\"We expect the primer to be in 5 to 3 prime direction.\")\n",
    "        codon_optimized_sequence = forward_primer + optimzed_sequence + reverse_primer\n",
    "        #try:\n",
    "        # Check now some simple things like that there is \n",
    "        if simple:\n",
    "            oligos = build_simple_oligos(seq_id, codon_optimized_sequence, min_segment_length, max_segment_length)\n",
    "        else:\n",
    "            oligos = build_oligos(seq_id, codon_optimized_sequence, output_directory, min_gc, max_gc, min_tm, max_tm, min_segment_length, max_segment_length, max_length)\n",
    "        # except Exception as e:\n",
    "        #     u.warn_p([f\"Warning: {seq_id} did not have any oligos built. \", e])\n",
    "        #     oligos = []\n",
    "        prev_oligo = None\n",
    "        # If a genbank file was provided also just add in the new sequnece\n",
    "        if len(oligos) > 0:\n",
    "            for i, oligo in enumerate(oligos):\n",
    "                seq = oligo[1]\n",
    "                # CHeck that there is an overlap with the previous sequence and that it is not too short\n",
    "                # Also make sure we swap the directions of the oligos so they automatically anneal\n",
    "                # Also assert that the start is a methionine (and if not warn it... )\n",
    "                primer_overlap = None\n",
    "                primer_tm = None\n",
    "                primer_len = None\n",
    "                homodimer_tm = None\n",
    "                hairpin_tm = None\n",
    "                if prev_oligo:  \n",
    "                    # Get the overlap with the previous sequence\n",
    "                    match = SequenceMatcher(None, prev_oligo, seq).find_longest_match()\n",
    "                    primer_overlap = prev_oligo[match.a:match.a + match.size]\n",
    "                    # Analyze the primer sequence\n",
    "                    results = check_secondary_structure(primer_overlap)\n",
    "                    homodimer_tm = results['homodimer']['homodimer_dg']\n",
    "                    hairpin_tm = results['hairpin']['hairpin_dg']\n",
    "                    primer_tm = primer3.bindings.calcTm(primer_overlap)\n",
    "                    primer_len = len(primer_overlap)\n",
    "\n",
    "                prev_oligo = seq\n",
    "                orig_seq = seq\n",
    "                strand = 1\n",
    "                # Changed this to be the opposite To\n",
    "                if bsa == True:\n",
    "                    if i % 2 == 1 and i != 0:\n",
    "                        seq = str(Seq(seq).reverse_complement())\n",
    "                        strand = -1\n",
    "                else:\n",
    "                    if i % 2 == 0:\n",
    "                        seq = str(Seq(seq).reverse_complement())\n",
    "                        strand = -1\n",
    "                oligo_tm = primer3.bindings.calcTm(seq)\n",
    "                if genbank_file:\n",
    "                    insert_features_from_oligos(record, f\"{seq_id}_oligo_{i}\", orig_seq, strand, oligo_tm, None)\n",
    "                rows.append([seq_id, oligo[0], seq, len(seq), oligo_tm, strand, primer_overlap, primer_tm, primer_len, homodimer_tm, hairpin_tm, oligo[2]])\n",
    "            if genbank_file:\n",
    "                output_file = genbank_file.replace('.', f'_{seq_id}.')\n",
    "                if genbank_file:\n",
    "                    record.name = seq_id  # type: ignore \n",
    "                    SeqIO.write(record, f'{output_directory}/{output_file}', \"genbank\")\n",
    "        else:\n",
    "            u.warn_p([f\"Warning: {seq_id} did not have any oligos built. \", optimzed_sequence])\n",
    "            rows.append([seq_id, None, optimzed_sequence, len(optimzed_sequence), None, None, None, None, None, None, None, None])\n",
    "        \n",
    "    oligo_df = pd.DataFrame(rows, columns=[\"id\", \"oligo_id\", \"oligo_sequence\", \"oligo_length\",  \"oligo_tm\", \"strand\",\"primer_overlap_with_previous\", \"overlap_tm_5prime\", \"overlap_length\", \n",
    "                                            \"overlap_homodimer_dg\", \"overlap_hairpin_dg\", \"original_sequence\"])\n",
    "    \n",
    "        \n",
    "    return oligo_df\n",
    "\n",
    "\n",
    "def build_simple_oligos(seq_id: str, sequence: str, min_segment_length=90, max_segment_length=130, overlap_len=18, optimal_temp=62, overlap_buffer=10):\n",
    "    # Basically get the splits size\n",
    "    seq_len = len(sequence)\n",
    "    num_splits = int(math.ceil(seq_len / min_segment_length))\n",
    "    # Make sure it is even\n",
    "    if num_splits % 2 == 0:\n",
    "        num_splits += 1\n",
    "    remainder = seq_len % num_splits\n",
    "    print(num_splits, remainder)        \n",
    "    # Now we want to go through the new part length and while remainder is greater then 0 we distribute this across the splits\n",
    "    max_part_len = math.floor(seq_len/num_splits)\n",
    "    split_counts = {}\n",
    "    for i in range(0, num_splits + 1):\n",
    "        split_counts[i] = max_part_len\n",
    "    # Now distribute the remainder\n",
    "    split_count = 0\n",
    "    for i in range(0, remainder + 1):\n",
    "        split_counts[split_count] += 1\n",
    "        split_count += 1\n",
    "        # Iterate through this again\n",
    "        if split_count == num_splits + 1:\n",
    "            split_count = 0\n",
    "    prev_cut = 0\n",
    "    rows = []\n",
    "    finished = False\n",
    "    for i in split_counts:\n",
    "        part_len = split_counts[i]\n",
    "        cut = prev_cut + part_len\n",
    "        oligo = sequence[prev_cut:cut]\n",
    "        # Calculate the tm and we'll check that we get the \"best\" one i.e. closest to 62 deg\n",
    "        # Get the overlap with the previous sequence\n",
    "        best_oligo = oligo\n",
    "        best_tm_diff = 10000\n",
    "        best_cut = cut\n",
    "        part_len_diff = 0\n",
    "        best_pl = 0 # The buffer of the overlap - was 10 by default\n",
    "        for pl in range(overlap_buffer, 0, -1):\n",
    "            for j in range(0, max_segment_length - min_segment_length):\n",
    "                cut = prev_cut + part_len + pl + j\n",
    "                oligo = sequence[prev_cut:cut]\n",
    "                primer_overlap = oligo[-1 * (overlap_len + pl):]\n",
    "                # Analyze the primer sequence\n",
    "                primer_tm = (primer3.bindings.calcTm(primer_overlap) + primer3.bindings.calcTm(str(Seq(primer_overlap).reverse_complement())))/2\n",
    "                results = check_secondary_structure(primer_overlap)\n",
    "                # Want to have the opp a high homodimer TM\n",
    "                homodimer_tm = -1 * results['homodimer']['homodimer_dg']\n",
    "                # Since the homodimer is negative and we're trying to minimize invert this.\n",
    "                if homodimer_tm < 1:\n",
    "                    # Ignore this unless it's actually a problem (otherwise we want to just care about temperature)\n",
    "                    homodimer_tm = 0\n",
    "                if (abs(primer_tm - optimal_temp) + homodimer_tm) < best_tm_diff:\n",
    "                    best_tm_diff = abs(primer_tm - optimal_temp) + homodimer_tm\n",
    "                    best_oligo = oligo\n",
    "                    best_cut = cut\n",
    "                    part_len_diff = j + pl  # Add the new offset and the overlap buffer (here we allow 18 to 29)\n",
    "                    best_pl = pl\n",
    "        # check the left over size\n",
    "        if len(sequence[best_cut:]) < 30:\n",
    "            # Add on the last bit and just have a longer final oligo\n",
    "            rows.append([f'{seq_id}_{i}', best_oligo + sequence[best_cut:], sequence, prev_cut, best_cut + len(sequence[best_cut:]), part_len + part_len_diff, 'last'])\n",
    "            finished = True\n",
    "            break\n",
    "        rows.append([f'{seq_id}_{i}', best_oligo, sequence, prev_cut, best_cut, part_len + part_len_diff, 'normal'])\n",
    "        prev_cut = best_cut - overlap_len - best_pl\n",
    "    # Add in the last one\n",
    "    oligo = sequence[prev_cut:]\n",
    "    if not finished:\n",
    "        part_len = len(oligo)\n",
    "        cut = len(sequence)\n",
    "        if len(oligo) < 18:\n",
    "            u.warn_p([\"Last oligo very short,.... check this!\", f'{seq_id}_{split_counts + 1}', oligo, sequence, prev_cut, cut, part_len, 'last'])\n",
    "        print(prev_cut, part_len, len(sequence), oligo)\n",
    "        rows.append([f'{seq_id}_{split_counts}', oligo, sequence, prev_cut, cut, part_len, 'last'])\n",
    "    return rows\n",
    "\n",
    "\n",
    "def codon_optimize(protein_sequence: str, min_gc=0.3, max_gc=0.7):\n",
    "    \"\"\" Codon optimize the protein sequence using DNA chisel: https://github.com/Edinburgh-Genome-Foundry/DnaChisel\"\"\"\n",
    "    seq = dnachisel.reverse_translate(protein_sequence)\n",
    "    problem = dnachisel.DnaOptimizationProblem(\n",
    "        sequence=seq,\n",
    "        constraints=[\n",
    "            AvoidPattern(\"BsaI_site\"), # type: ignore\n",
    "            EnforceGCContent(mini=min_gc, maxi=max_gc, window=50), # type: ignore\n",
    "            EnforceTranslation(location=(0, len(seq))), # type: ignore\n",
    "            AvoidStopCodons(location=(0, len(seq)-3)) # type: ignore Let there be stop codons in the last bit\n",
    "        ],\n",
    "        objectives=[CodonOptimize(species='e_coli', location=(0, len(seq)))] # type: ignore\n",
    "    )\n",
    "    # SOLVE THE CONSTRAINTS, OPTIMIZE WITH RESPECT TO THE OBJECTIVE\n",
    "    problem.resolve_constraints()\n",
    "    problem.optimize()\n",
    "\n",
    "    # PRINT SUMMARIES TO CHECK THAT CONSTRAINTS PASS\n",
    "    print(problem.constraints_text_summary())\n",
    "    print(problem.objectives_text_summary())\n",
    "\n",
    "    # GET THE FINAL SEQUENCE (AS STRING OR ANNOTATED BIOPYTHON RECORDS)\n",
    "    final_sequence = problem.sequence  # string\n",
    "    final_record = problem.to_record(with_sequence_edits=True)\n",
    "    print(protein_sequence)\n",
    "    print(final_sequence)\n",
    "    return final_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "92b7c94b-da05-4eb8-a7b2-23d53746c7d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pyswarms as ps\n",
    "from primer3 import bindings\n",
    "from functools import partial\n",
    "import math\n",
    "\n",
    "# --- Parameters ---\n",
    "min_len, max_len = 80, 130\n",
    "min_olap, max_olap = 18, 27\n",
    "tm_target = 62\n",
    "\n",
    "def get_tm(seq):\n",
    "    try:\n",
    "        return bindings.calcTm(seq)\n",
    "    except:\n",
    "        return 100  # Penalize invalid sequences\n",
    "\n",
    "def compute_fragments(cut_sites, seq):\n",
    "    cut_sites = np.sort(np.round(cut_sites).astype(int))\n",
    "    cut_sites = [0] + list(cut_sites) + [len(seq)]\n",
    "\n",
    "    fragments = []\n",
    "    overlaps = []\n",
    "\n",
    "    for i in range(len(cut_sites)-1):\n",
    "        frag = seq[cut_sites[i]:cut_sites[i+1]]\n",
    "        fragments.append(frag)\n",
    "        if i < len(cut_sites) - 2:\n",
    "            # Define overlap with next fragment\n",
    "            overlap = seq[cut_sites[i+1]-max_olap : cut_sites[i+1]]\n",
    "            overlaps.append(overlap)\n",
    "\n",
    "    return fragments, overlaps\n",
    "\n",
    "def objective_function(x, seq):\n",
    "    penalties = []\n",
    "    for particle in x:\n",
    "        particle = np.clip(particle, 0, len(seq))  # Ensure inside sequence\n",
    "        fragments, overlaps = compute_fragments(particle, seq)\n",
    "\n",
    "        total_penalty = 0\n",
    "        for frag in fragments:\n",
    "            if not (min_len <= len(frag) <= max_len):\n",
    "                total_penalty += 100 * abs(len(frag) - (min_len + max_len)/2)\n",
    "\n",
    "        for olap in overlaps:\n",
    "            if not (min_olap <= len(olap) <= max_olap):\n",
    "                total_penalty += 100\n",
    "            tm = get_tm(olap)\n",
    "            total_penalty += abs(tm - tm_target) * 10\n",
    "\n",
    "        penalties.append(total_penalty)\n",
    "    return np.array(penalties)\n",
    "\n",
    "# Example: Optimize cut sites for a single sequence\n",
    "def optimize_sequence(seq, num_fragments):\n",
    "    num_cuts = num_fragments - 1\n",
    "    seq_len = len(seq)\n",
    "\n",
    "    # Approx evenly spaced initial bounds (naive)\n",
    "    margin = max_len + max_olap\n",
    "    lower_bound = np.array([i * min_len for i in range(1, num_fragments)])\n",
    "    upper_bound = np.array([seq_len - (num_fragments - i) * min_len for i in range(1, num_fragments)])\n",
    "\n",
    "    bounds = (lower_bound, upper_bound)\n",
    "    print(bounds)\n",
    "    \n",
    "    # optimizer = ps.single.GlobalBestPSO(\n",
    "    #     n_particles=50,\n",
    "    #     dimensions=num_cuts,\n",
    "    #     options={'c1': 0.5, 'c2': 0.3, 'w': 0.9},\n",
    "    #     bounds=bounds\n",
    "    # )\n",
    "\n",
    "    # best_cost, best_pos = optimizer.optimize(partial(objective_function, seq=seq), iters=200)\n",
    "    # return best_cost, best_pos\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "222b207a-3501-46b7-9421-28b4512de3a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(array([  80,  160,  240,  320,  400,  480,  560,  640,  720,  800,  880,\n",
      "        960, 1040, 1120, 1200, 1280]), array([-322, -242, -162,  -82,   -2,   78,  158,  238,  318,  398,  478,\n",
      "        558,  638,  718,  798,  878]))\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "cannot unpack non-iterable NoneType object",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[12], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m cost, positions \u001b[38;5;241m=\u001b[39m optimize_sequence(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mgaaataattttgtttaactttaagaaggagatatacatATGCGCAAGCTTTCTCGTTGTGAGGACGGCTACGTCAAGATTCAGGGGATCTACATTTACTATAAAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCACGGTGGACCCGGCATGAGCCACGACTACTTGCTTTCCCTGACCGATTTAGCTGAGAAGGGGATTACGGTTTTATTCTATGACCAATTCGGCTGTGGACGTTCAGAAGAACCTGAGAAGGAGAAATTTACGATCGACTATGGAGTGGAAGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTATTCCTGATGGGATCGTCCTATGGTGGGGCTTTAGCTCTGGCCTACGCTGTTAAGTATCAAGCACATTTGAAGGGCCTGATCATCTCGGGCGGATTATCATCTGTGCCTCTGACAGTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACCGCAACGCGATCCGCAAGTATGGCGAAGTGGGGGATTATCAGAACCCTGAATACCAGGAGGCTGTGAACTATTTCTATCACCAGCATTTGTTACGCTCTGAAGACTGGCCCCCCGAGGTTTTGAAAAGCCTGGAGTACGCAGAAGAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAATCACTGGAACTATCCGCGACTGGGATATCACCGATAAGATCGGGATTATCAGTGTTCCTACTCTTATTACGGTCGGTGAGTTTGACGAGGTAACGCAAAATGTCGCAGAAGTTATCCACTCCAAAATCGATAATTCCCAGTTAATTGTATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAGTACAACCGCATTTTATTACAGTTCATCGAAAAGAACATTCTCGAGCACCACCACCACCACCACTGAgatccggc\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;241m17\u001b[39m)\n",
      "\u001b[0;31mTypeError\u001b[0m: cannot unpack non-iterable NoneType object"
     ]
    }
   ],
   "source": [
    "cost, positions = optimize_sequence('gaaataattttgtttaactttaagaaggagatatacatATGCGCAAGCTTTCTCGTTGTGAGGACGGCTACGTCAAGATTCAGGGGATCTACATTTACTATAAAGTGTGCAAAGCCGAGAATGAGAAAGCCAAGTTAATGACACTGCACGGTGGACCCGGCATGAGCCACGACTACTTGCTTTCCCTGACCGATTTAGCTGAGAAGGGGATTACGGTTTTATTCTATGACCAATTCGGCTGTGGACGTTCAGAAGAACCTGAGAAGGAGAAATTTACGATCGACTATGGAGTGGAAGAAGCAGAAGCGGTGAAAAAAAACATCTTTGGTGATGATAAGGTATTCCTGATGGGATCGTCCTATGGTGGGGCTTTAGCTCTGGCCTACGCTGTTAAGTATCAAGCACATTTGAAGGGCCTGATCATCTCGGGCGGATTATCATCTGTGCCTCTGACAGTAAAAGAAATGCAGCGCTTGATTGACGAGCTTCCAGAAAAGTACCGCAACGCGATCCGCAAGTATGGCGAAGTGGGGGATTATCAGAACCCTGAATACCAGGAGGCTGTGAACTATTTCTATCACCAGCATTTGTTACGCTCTGAAGACTGGCCCCCCGAGGTTTTGAAAAGCCTGGAGTACGCAGAAGAGCGCAATGTGTACCGCACAATGAACGGCCCCAATGAATTTACAATCACTGGAACTATCCGCGACTGGGATATCACCGATAAGATCGGGATTATCAGTGTTCCTACTCTTATTACGGTCGGTGAGTTTGACGAGGTAACGCAAAATGTCGCAGAAGTTATCCACTCCAAAATCGATAATTCCCAGTTAATTGTATTTAAAGCATGTAGCCACCTGACCATGTGGGAGGATCGTGATGAGTACAACCGCATTTTATTACAGTTCATCGAAAAGAACATTCTCGAGCACCACCACCACCACCACTGAgatccggc', 17)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "935d69d6-de4b-438f-9d95-ecd8ad3b9135",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[102,\n",
       " 214,\n",
       " 293,\n",
       " 319,\n",
       " 399,\n",
       " 403,\n",
       " 430,\n",
       " 519,\n",
       " 535,\n",
       " 615,\n",
       " 639,\n",
       " 668,\n",
       " 754,\n",
       " 772,\n",
       " 783,\n",
       " 862]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "positions.sort()\n",
    "positions = [int(p) for p in positions]\n",
    "oligo"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
