{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "04e33c47-4362-4ef3-9168-fb9bba96524c",
   "metadata": {},
   "source": [
    "# DNA weaver\n",
    "\n",
    "#### \n",
    "https://github.com/Edinburgh-Genome-Foundry/DnaWeaver/blob/master/examples/10kb_from_vendor_or_oligos.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a3400d1e-d6ed-42d7-870c-99487f0bb26d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dnaweaver in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (0.3.8)\n",
      "Requirement already satisfied: numpy in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (2.1.0)\n",
      "Requirement already satisfied: Biopython in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (1.84)\n",
      "Requirement already satisfied: proglog in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (0.1.10)\n",
      "Requirement already satisfied: networkx in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (3.3)\n",
      "Requirement already satisfied: flametree in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (0.1.12)\n",
      "Requirement already satisfied: dna-features-viewer in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (3.1.3)\n",
      "Requirement already satisfied: weasyprint in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (63.0)\n",
      "Requirement already satisfied: pandas in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (2.2.2)\n",
      "Requirement already satisfied: jinja2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (3.1.4)\n",
      "Requirement already satisfied: matplotlib>=3 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dna-features-viewer->dnaweaver) (3.9.2)\n",
      "Requirement already satisfied: packaging in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dna-features-viewer->dnaweaver) (24.1)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from jinja2->dnaweaver) (2.1.5)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnaweaver) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnaweaver) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnaweaver) (2024.1)\n",
      "Requirement already satisfied: tqdm in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from proglog->dnaweaver) (4.67.1)\n",
      "Requirement already satisfied: pydyf<0.12,>=0.11.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (0.11.0)\n",
      "Requirement already satisfied: cffi<2,>=0.6 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (1.17.0)\n",
      "Requirement already satisfied: tinyhtml5<3,>=2.0.0b1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (2.0.0)\n",
      "Requirement already satisfied: tinycss2<2,>=1.4.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (1.4.0)\n",
      "Requirement already satisfied: cssselect2<0.8,>=0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (0.7.0)\n",
      "Requirement already satisfied: Pyphen<0.16,>=0.9.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (0.15.0)\n",
      "Requirement already satisfied: Pillow<11,>=9.1.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (10.4.0)\n",
      "Requirement already satisfied: fonttools<5,>=4.0.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->dnaweaver) (4.53.1)\n",
      "Requirement already satisfied: pycparser in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cffi<2,>=0.6->weasyprint->dnaweaver) (2.22)\n",
      "Requirement already satisfied: webencodings in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cssselect2<0.8,>=0.1->weasyprint->dnaweaver) (0.5.1)\n",
      "Requirement already satisfied: zopfli>=0.1.4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->dnaweaver) (0.2.3.post1)\n",
      "Requirement already satisfied: brotli>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->dnaweaver) (1.1.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib>=3->dna-features-viewer->dnaweaver) (1.3.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib>=3->dna-features-viewer->dnaweaver) (0.12.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib>=3->dna-features-viewer->dnaweaver) (1.4.5)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib>=3->dna-features-viewer->dnaweaver) (3.1.4)\n",
      "Requirement already satisfied: six>=1.5 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas->dnaweaver) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "! pip install dnaweaver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aff1bbdb-e28c-4861-b244-d1399ca74cd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                                                                                                                       "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ordering plan (Gibson Assembly Station)::\n",
      "  0-100: From CheapDNA. - price 11.10 - lead_time 0.0\n",
      "  100-210: From Oligo Assembly Station - price 40.43 - lead_time 8.0\n",
      "  210-320: From CheapDNA. - price 13.10 - lead_time 0.0\n",
      "  320-430: From CheapDNA. - price 13.10 - lead_time 0.0\n",
      "  430-519: From CheapDNA. - price 9.90 - lead_time 0.0\n",
      "Price:87.63, total lead_time:8.0\n",
      "Finished in 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    }
   ],
   "source": [
    "import dnaweaver as dw\n",
    "import time\n",
    "\n",
    "cheap_dna_offer = dw.CommercialDnaOffer(\n",
    "    name=\"CheapDNA.\",\n",
    "    sequence_constraints=[\n",
    "        dw.NoPatternConstraint(enzyme=\"BsaI\"),\n",
    "        dw.SequenceLengthConstraint(max_length=4000),\n",
    "        dw.GcContentConstraint(min_gc=0.3, max_gc=0.7)\n",
    "    ],\n",
    "    pricing=dw.PerBasepairPricing(0.10),\n",
    ")\n",
    "\n",
    "oligo_dna_offer = dw.CommercialDnaOffer(\n",
    "    name=\"OliGoo\",\n",
    "    sequence_constraints=[\n",
    "        dw.GcContentConstraint(min_gc=0.3, max_gc=0.7),\n",
    "        dw.SequenceLengthConstraint(max_length=100),\n",
    "    ],\n",
    "    pricing=dw.PerBasepairPricing(0.07),\n",
    "    memoize=True\n",
    ")\n",
    "\n",
    "oligo_assembly_station = dw.DnaAssemblyStation(\n",
    "    name=\"Oligo Assembly Station\",\n",
    "    assembly_method=dw.OligoAssemblyMethod(\n",
    "        overhang_selector=dw.TmSegmentSelector(\n",
    "            min_size=15, max_size=25, min_tm=50, max_tm=70\n",
    "        ),\n",
    "        min_segment_length=40,\n",
    "        max_segment_length=200,\n",
    "        sequence_constraints=[dw.SequenceLengthConstraint(max_length=1500)],\n",
    "        duration=8,\n",
    "        cost=30,\n",
    "    ),\n",
    "    supplier=oligo_dna_offer,\n",
    "    coarse_grain=20,\n",
    "    a_star_factor=\"auto\",\n",
    "    memoize=True,\n",
    ")\n",
    "\n",
    "assembly_station = dw.DnaAssemblyStation(\n",
    "    name=\"Gibson Assembly Station\",\n",
    "    assembly_method=dw.GibsonAssemblyMethod(\n",
    "        overhang_selector=dw.TmSegmentSelector(min_tm=55, max_tm=70),\n",
    "        min_segment_length=80,\n",
    "        max_segment_length=120,\n",
    "    ),\n",
    "    supplier=[cheap_dna_offer, oligo_assembly_station],\n",
    "    logger=\"bar\",\n",
    "    coarse_grain=100,\n",
    "    fine_grain=10,\n",
    "    a_star_factor=\"auto\",\n",
    ")\n",
    "print(\"Looking for the best assembly plan...\")\n",
    "t0 = time.time()\n",
    "sequence = \"ATGACGACGCCGTGGTCGGGAGAGTTAATTGTGGATTTATTAAATCAGCGTCCTCGCGGACAAGTCGTGTTAGATGCTGCTAATGGGGTGGATTATGAGCTGCGCTCTCAAGAGACCGGACCAATTGACATCGACGCACACGGACCCGCTCGCTGCCTGACCGGAACGGGCTCCCGTACGACGCGTTGGATTCGCCAACATCAGCGCAATACTCGCGTTTTGGGCCTGTCGCATGaccctacagttgctgacgcgcgcgaattgttacgtttaggggctttggGCCTTTTACCTAAAAATGCCGAGCCAGCGGAAATCTTCCAAGCGTTAGATCAAGCTGCCGCCGACCACATTCACACGAATGACCTTGTGCGCGACGCACCCCGTACCCCCGCTCTTCCTGAAGAAGAGGAGGCAGAACGCGTCAACGCTTTGCCAGACCAGGGCGCGGCGCGCACGGGCGCTTTGGGAGCCCAAGGTTGGCCCCACCATACTCCTGACCAACCGGCGCGCTTAGGT\"\n",
    "quote = assembly_station.get_quote(sequence, with_assembly_plan=True)\n",
    "print(quote.assembly_step_summary())\n",
    "print(\"Finished in %.01d seconds\" % (time.time() - t0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "db55993c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 108 (l) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 112 (p) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 108 (l) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 112 (p) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n"
     ]
    }
   ],
   "source": [
    "assembly_plan_report = quote.to_assembly_plan_report()\n",
    "assembly_plan_report.write_full_report(\"report.zip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d35ff231-d452-4934-b3b3-74d1224c0a65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(assembly_plan_report.plan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "872d7316-952c-4a5a-a370-d04e68a37de9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGACGACGCCGTGGTCGGGAGAGTTAATTGTGGATTTATTAAATCAGCGTCCTCGCGGACAAGTCGTGTTAGATGCTGCTAATGGGGTGGATTATGAGCTGCGCTCTCAAGAGACCGGACCAATTGACATCGACGCACACGGACCCGCTCGCTGCCTGACCGGAACGGGCTCCCGTACGACGCGTTGGATTCGCCAACATCAGCGCAATACTCGCGTTTTGGGCCTGTCGCATGaccctacagttgctgacgcgcgcgaattgttacgtttaggggctttggGCCTTTTACCTAAAAATGCCGAGCCAGCGGAAATCTTCCAAGCGTTAGATCAAGCTGCCGCCGACCACATTCACACGAATGACCTTGTGCGCGACGCACCCCGTACCCCCGCTCTTCCTGAAGAAGAGGAGGCAGAACGCGTCAACGCTTTGCCAGACCAGGGCGCGGCGCGCACGGGCGCTTTGGGAGCCCAAGGTTGGCCCCACCATACTCCTGACCAACCGGCGCGCTTAGGT\n"
     ]
    }
   ],
   "source": [
    "print(assembly_plan_report.plan.sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "22337005-113e-489f-91fd-5ec2774d97be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S_00000 ATGACGACGCCGTGGTCGGGAGAGTTAATTGTGGATTTATTAAATCAGCGTCCTCGCGGACAAGTCGTGTTAGATGCTGCTAATGGGGTGGATTATGAGCTGCGCTCTCAA\n",
      "{'via': ['Gibson Assembly Station comparator']}\n",
      "S_00003 GATTATGAGCTGCGCTCTCAAGAGACCGGACCAATTGACATCGACGCACACGGACCCGCTCGCTGCCTGACCGGAACGGGCTCCCGTACGACGCGTTGGATTCGCCAACATCAGCGCAATACTCGCGTTTT\n",
      "{'via': ['Gibson Assembly Station comparator', 'Gibson Assembly Station comparator']}\n",
      "S_00004 TCAGCGCAATACTCGCGTTTTGGGCCTGTCGCATGaccctacagttgctgacgcgcgcgaattgttacgtttaggggctttggGCCTTTTACCTAAAAATGCCGAGCCAGCGGAAATCTTCCAAGCGTTAG\n",
      "{'via': ['Gibson Assembly Station comparator']}\n",
      "S_00005 GCGGAAATCTTCCAAGCGTTAGATCAAGCTGCCGCCGACCACATTCACACGAATGACCTTGTGCGCGACGCACCCCGTACCCCCGCTCTTCCTGAAGAAGAGGAGGCAGAACGCGTCAACGCTTTGCCAGA\n",
      "{'via': ['Gibson Assembly Station comparator']}\n",
      "S_00006 CGCGTCAACGCTTTGCCAGACCAGGGCGCGGCGCGCACGGGCGCTTTGGGAGCCCAAGGTTGGCCCCACCATACTCCTGACCAACCGGCGCGCTTAGGT\n",
      "{'via': ['Gibson Assembly Station comparator']}\n"
     ]
    }
   ],
   "source": [
    "for oligo in assembly_plan_report.plan.assembly_plan:\n",
    "    if oligo.accepted:\n",
    "        print(oligo.id, oligo.sequence)\n",
    "        print(oligo.metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15f5b1c4-3645-45bc-b096-5796a68c41b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Other installs\n",
    "\n",
    "brew install glib\n",
    " brew install blast\n",
    "brew install python pango libffi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "79a04bd9-c8ea-4b5f-a6cb-30d11558395a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pdf_reports in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (0.3.7)\n",
      "Requirement already satisfied: pypugjs in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (5.11.0)\n",
      "Requirement already satisfied: jinja2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (3.1.4)\n",
      "Requirement already satisfied: weasyprint in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (63.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (4.12.3)\n",
      "Requirement already satisfied: pandas in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (2.2.2)\n",
      "Requirement already satisfied: Markdown in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (3.7)\n",
      "Requirement already satisfied: backports.functools-lru-cache in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (2.0.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from beautifulsoup4->pdf_reports) (2.6)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from jinja2->pdf_reports) (2.1.5)\n",
      "Requirement already satisfied: numpy>=1.22.4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->pdf_reports) (2.1.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->pdf_reports) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->pdf_reports) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->pdf_reports) (2024.1)\n",
      "Requirement already satisfied: six in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pypugjs->pdf_reports) (1.16.0)\n",
      "Requirement already satisfied: charset-normalizer in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pypugjs->pdf_reports) (3.3.2)\n",
      "Requirement already satisfied: pydyf<0.12,>=0.11.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (0.11.0)\n",
      "Requirement already satisfied: cffi<2,>=0.6 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (1.17.0)\n",
      "Requirement already satisfied: tinyhtml5<3,>=2.0.0b1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (2.0.0)\n",
      "Requirement already satisfied: tinycss2<2,>=1.4.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (1.4.0)\n",
      "Requirement already satisfied: cssselect2<0.8,>=0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (0.7.0)\n",
      "Requirement already satisfied: Pyphen<0.16,>=0.9.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (0.15.0)\n",
      "Requirement already satisfied: Pillow<11,>=9.1.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (10.4.0)\n",
      "Requirement already satisfied: fonttools<5,>=4.0.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf_reports) (4.53.1)\n",
      "Requirement already satisfied: pycparser in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cffi<2,>=0.6->weasyprint->pdf_reports) (2.22)\n",
      "Requirement already satisfied: webencodings in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cssselect2<0.8,>=0.1->weasyprint->pdf_reports) (0.5.1)\n",
      "Requirement already satisfied: zopfli>=0.1.4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf_reports) (0.2.3.post1)\n",
      "Requirement already satisfied: brotli>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf_reports) (1.1.0)\n"
     ]
    }
   ],
   "source": [
    "!  pip install pdf_reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9b8b2ebf-1650-4970-80e8-824e9025f6d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dnachisel in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (3.2.12)\n",
      "Requirement already satisfied: numpy in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (2.1.0)\n",
      "Requirement already satisfied: Biopython in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (1.84)\n",
      "Requirement already satisfied: proglog in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (0.1.10)\n",
      "Requirement already satisfied: docopt in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (0.6.2)\n",
      "Requirement already satisfied: flametree in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (0.1.12)\n",
      "Requirement already satisfied: python-codon-tables in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (0.1.12)\n",
      "Requirement already satisfied: tqdm in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from proglog->dnachisel) (4.67.1)\n",
      "Requirement already satisfied: dnachisel[reports] in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (3.2.12)\n",
      "Requirement already satisfied: numpy in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (2.1.0)\n",
      "Requirement already satisfied: Biopython in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (1.84)\n",
      "Requirement already satisfied: proglog in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (0.1.10)\n",
      "Requirement already satisfied: docopt in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (0.6.2)\n",
      "Requirement already satisfied: flametree in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (0.1.12)\n",
      "Requirement already satisfied: python-codon-tables in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (0.1.12)\n",
      "Collecting pdf-reports (from dnachisel[reports])\n",
      "  Downloading pdf_reports-0.3.7-py3-none-any.whl.metadata (2.1 kB)\n",
      "Collecting sequenticon (from dnachisel[reports])\n",
      "  Downloading sequenticon-0.1.5-py3-none-any.whl.metadata (2.0 kB)\n",
      "Requirement already satisfied: matplotlib in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (3.9.2)\n",
      "Requirement already satisfied: dna-features-viewer in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (3.1.3)\n",
      "Requirement already satisfied: pandas in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (2.2.2)\n",
      "Requirement already satisfied: packaging in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dna-features-viewer->dnachisel[reports]) (24.1)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (1.3.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (4.53.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (1.4.5)\n",
      "Requirement already satisfied: pillow>=8 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (3.1.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnachisel[reports]) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnachisel[reports]) (2024.1)\n",
      "Collecting pypugjs (from pdf-reports->dnachisel[reports])\n",
      "  Downloading pypugjs-5.11.0-py2.py3-none-any.whl.metadata (7.4 kB)\n",
      "Requirement already satisfied: jinja2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf-reports->dnachisel[reports]) (3.1.4)\n",
      "Requirement already satisfied: weasyprint in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf-reports->dnachisel[reports]) (63.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf-reports->dnachisel[reports]) (4.12.3)\n",
      "Collecting Markdown (from pdf-reports->dnachisel[reports])\n",
      "  Using cached Markdown-3.7-py3-none-any.whl.metadata (7.0 kB)\n",
      "Collecting backports.functools-lru-cache (from pdf-reports->dnachisel[reports])\n",
      "  Downloading backports.functools_lru_cache-2.0.0-py2.py3-none-any.whl.metadata (3.5 kB)\n",
      "Requirement already satisfied: tqdm in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from proglog->dnachisel[reports]) (4.67.1)\n",
      "Collecting pydenticon (from sequenticon->dnachisel[reports])\n",
      "  Downloading pydenticon-0.3.1.tar.gz (26 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting snapgene-reader (from sequenticon->dnachisel[reports])\n",
      "  Downloading snapgene_reader-0.1.21-py3-none-any.whl.metadata (1.6 kB)\n",
      "Requirement already satisfied: six>=1.5 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib->dnachisel[reports]) (1.16.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from beautifulsoup4->pdf-reports->dnachisel[reports]) (2.6)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from jinja2->pdf-reports->dnachisel[reports]) (2.1.5)\n",
      "Requirement already satisfied: charset-normalizer in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pypugjs->pdf-reports->dnachisel[reports]) (3.3.2)\n",
      "Collecting xmltodict (from snapgene-reader->sequenticon->dnachisel[reports])\n",
      "  Using cached xmltodict-0.14.2-py2.py3-none-any.whl.metadata (8.0 kB)\n",
      "Collecting html2text (from snapgene-reader->sequenticon->dnachisel[reports])\n",
      "  Downloading html2text-2024.2.26.tar.gz (56 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: pydyf<0.12,>=0.11.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (0.11.0)\n",
      "Requirement already satisfied: cffi<2,>=0.6 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (1.17.0)\n",
      "Requirement already satisfied: tinyhtml5<3,>=2.0.0b1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (2.0.0)\n",
      "Requirement already satisfied: tinycss2<2,>=1.4.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (1.4.0)\n",
      "Requirement already satisfied: cssselect2<0.8,>=0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (0.7.0)\n",
      "Requirement already satisfied: Pyphen<0.16,>=0.9.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (0.15.0)\n",
      "Requirement already satisfied: pycparser in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cffi<2,>=0.6->weasyprint->pdf-reports->dnachisel[reports]) (2.22)\n",
      "Requirement already satisfied: webencodings in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cssselect2<0.8,>=0.1->weasyprint->pdf-reports->dnachisel[reports]) (0.5.1)\n",
      "Requirement already satisfied: zopfli>=0.1.4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf-reports->dnachisel[reports]) (0.2.3.post1)\n",
      "Requirement already satisfied: brotli>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf-reports->dnachisel[reports]) (1.1.0)\n",
      "Downloading pdf_reports-0.3.7-py3-none-any.whl (118 kB)\n",
      "Downloading sequenticon-0.1.5-py3-none-any.whl (13 kB)\n",
      "Downloading backports.functools_lru_cache-2.0.0-py2.py3-none-any.whl (6.7 kB)\n",
      "Using cached Markdown-3.7-py3-none-any.whl (106 kB)\n",
      "Downloading pypugjs-5.11.0-py2.py3-none-any.whl (36 kB)\n",
      "Downloading snapgene_reader-0.1.21-py3-none-any.whl (7.1 kB)\n",
      "Using cached xmltodict-0.14.2-py2.py3-none-any.whl (10.0 kB)\n",
      "Building wheels for collected packages: pydenticon, html2text\n",
      "  Building wheel for pydenticon (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for pydenticon: filename=pydenticon-0.3.1-py3-none-any.whl size=6515 sha256=49fced90976c18391018235a763034fd30175a74b14f315a67ac26e2a82deb87\n",
      "  Stored in directory: /Users/arianemora/Library/Caches/pip/wheels/6f/20/6d/926175bfcc1c4e585db37504fde7cdbb7c741a97b38c5c294b\n",
      "  Building wheel for html2text (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for html2text: filename=html2text-2024.2.26-py3-none-any.whl size=33110 sha256=0e7026d49ea1191420012ca83d5e909a2bdee4edd826f8a1b0c6ac2c1336d8d8\n",
      "  Stored in directory: /Users/arianemora/Library/Caches/pip/wheels/f3/96/6d/a7eba8f80d31cbd188a2787b81514d82fc5ae6943c44777659\n",
      "Successfully built pydenticon html2text\n",
      "Installing collected packages: xmltodict, pypugjs, pydenticon, Markdown, html2text, backports.functools-lru-cache, snapgene-reader, pdf-reports, sequenticon\n",
      "Successfully installed Markdown-3.7 backports.functools-lru-cache-2.0.0 html2text-2024.2.26 pdf-reports-0.3.7 pydenticon-0.3.1 pypugjs-5.11.0 sequenticon-0.1.5 snapgene-reader-0.1.21 xmltodict-0.14.2\n"
     ]
    }
   ],
   "source": [
    "! pip install dnachisel\n",
    "! pip install 'dnachisel[reports]'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e5702e46-03b8-424a-bb34-b7a8421f1135",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dnachisel import *\n",
    "\n",
    "# DEFINE THE OPTIMIZATION PROBLEM\n",
    "seq = 'MASPQLQMALDAFKTMGEKMAQAGNDVKALRAVMEEMSGFPSAGETKCTPVNAGGVPAEWISGPGAAD'\n",
    "seq = reverse_translate(seq)\n",
    "problem = DnaOptimizationProblem(\n",
    "    sequence=seq,\n",
    "    constraints=[\n",
    "        AvoidPattern(\"BsaI_site\"),\n",
    "        EnforceGCContent(mini=0.3, maxi=0.7, window=50),\n",
    "    ],\n",
    "    objectives=[CodonOptimize(species='e_coli', location=(500, 1400))]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2daeb8c5-1b13-46c0-87ba-dd5cbe76ab1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solving constraints\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "constraint:   0%|                                                                                                        | 0/2 [00:00<?, ?it/s, now=AvoidPattern[0-10000](pat...]\n",
      "location:   0%|                                                                                                                                  | 0/3 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                                                                          | 0/3 [00:00<?, ?it/s, now=4574-4580(-)]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                          | 0/4 [00:00<?, ?it/s, now=4574-4580(-)]\u001b[A\n",
      "location:   0%|                                                                                                                             | 0/4 [00:00<?, ?it/s, now=2043-2099]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                             | 0/4 [00:00<?, ?it/s, now=2043-2099]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now optimizing the sequence\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "objective:   0%|                                                                                                         | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[500-1400](e_c...]\n",
      "location:   0%|                                                                                                                           | 0/200 [00:00<?, ?it/s, now=2043-2099]\u001b[A\n",
      "location:   0%|                                                                                                                             | 0/200 [00:00<?, ?it/s, now=500-503]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                             | 0/200 [00:00<?, ?it/s, now=500-503]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Success! Generating report.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(True, 'Optimization successful.', None)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "problem.optimize_with_report(target=\"dna_chisel_report.zip\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5c504165-0201-4517-9d6e-ac5f15e6004c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "constraint:   0%|                                                                                                        | 0/2 [00:00<?, ?it/s, now=AvoidPattern[0-10000](pat...]\n",
      "location:   0%|                                                                                                                                  | 0/7 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                                                                          | 0/7 [00:00<?, ?it/s, now=1114-1120(+)]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                          | 0/1 [00:00<?, ?it/s, now=1114-1120(+)]\u001b[A\n",
      "location:   0%|                                                                                                                             | 0/1 [00:00<?, ?it/s, now=1746-1797]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                             | 0/1 [00:00<?, ?it/s, now=1746-1797]\u001b[A\n",
      "objective:   0%|                                                                                                         | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[500-1400](e_c...]\u001b[A\n",
      "location:   0%|                                                                                                                           | 0/201 [00:00<?, ?it/s, now=1746-1797]\u001b[A\n",
      "location:   0%|                                                                                                                             | 0/201 [00:00<?, ?it/s, now=515-518]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                             | 0/201 [00:00<?, ?it/s, now=515-518]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-10000](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-10000](mini:0.30, maxi:0.70, window:50)\n",
      "      │ Passed !\n",
      "✔PASS ┍ EnforceTranslation[500-1400(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:     -4.70\n",
      "     -4.70 ┍ MaximizeCAI[500-1400](e_coli) \n",
      "           │ Codon opt. on window 500-1400 scored -4.70E+00\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    }
   ],
   "source": [
    "# SOLVE THE CONSTRAINTS, OPTIMIZE WITH RESPECT TO THE OBJECTIVE\n",
    "\n",
    "problem.resolve_constraints()\n",
    "problem.optimize()\n",
    "\n",
    "# PRINT SUMMARIES TO CHECK THAT CONSTRAINTS PASS\n",
    "\n",
    "print(problem.constraints_text_summary())\n",
    "print(problem.objectives_text_summary())\n",
    "\n",
    "# GET THE FINAL SEQUENCE (AS STRING OR ANNOTATED BIOPYTHON RECORDS)\n",
    "\n",
    "final_sequence = problem.sequence  # string\n",
    "final_record = problem.to_record(with_sequence_edits=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "417d5ed3-6d41-4e9f-a6d3-ece6385170a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTACGATCCATCGCCCCTGTTAACATTACGGAATTCAACAGGTTCTCCGGGCCGAACAAAAATGCGGAATGAGGAATAGATAGGCTATGCACGGCCTCCTAGAGCTTCTTGGTTCTTCCTCCATCCTACCCTAGCTGTTGCACGTATAGACATCCGCGATAGTCGGCACGAGCTTACCTAGTGATGGACATTCCTCACAGCATACGCTAGAAATCAGACTCATTCATTACGTTCGATCACCTACTCTGATCGAAGTGAATGTGGGAGGAGGTCACGAATGCCCCCAATATCCTACCGCATAGAGATAGGTTGAGATACGAACATACCTGGCGAGTAATGCGATAGTTGAAGGCAGAGAGGACCGCACACCGCCGTCGCGGTGTGGGTGCTGAAACATCCGTCAGGGTGGACCTTCGGAAGAAACTTACGTTCTGCCTGTTGCGTCAGTCGTCGGCAAGTCGTTTAGTACATAGGGACACGGCAAGCCTACGCCTCTCCAGGATTTTGGCATGCATCGCGTGGGCCCGAAACGCATTACCACCGTGCTGTCTCGCAGCGCACCGTGCACTCTGGGCGAAAAACAGGCGCTGCTGACCAAATGGCGCTGCCAGCTGAGCCATGCGCGTCGCCATCCGCGTTTTAGCACCGGCAGCGGCCGTCGCAAATAACTGTATGGCCCGAAAGGCCGCCGCCAGCATAGCCAGGTGGGCACCGATTGCTGCTAAATTTAATTTCCGGTGTTTGCGCTGGATGTGATGGGCCTGCCGGCGTGGGAAAACAACGGCACCAGCAGCGTGTATAGCGTGCGCCTGCATGCGCTGTAACCGAGCCATACCGATAGCTTTGCGCGCCTGCGCATTGAACTGCAGGTGACCGGCGTGACCACCGGCTATGGCCGCTATGCGAGCACCAGCCTGCTGACCCTGACCTATCTGGGCGAAGCGATTGAAGTGGTGCTGCTGATGAAAGAAGGCAAAAGCATTGATGAACATTGCGCGATTCATCAGAGCGAACAGCTGAAAGCGGCGCGCGATTTTAGCCCGGCGGCGCGCACCTAACGCATTCTGCTGCTGGTGCGCCATTATACCCCGCCGGGCCAACGTCGCGTTGGCGAAGTGAGCATTCCGGGCAAAACCTAAAGCCGCGATTGCCTGGCGCTGATTATTGCGCCGCTGGGCGTGGTGGAAACCATTACCCCGCGCCTGGAATGCTTTGTGCGCAGCTAAGGCCCGGTGGTGAACGTGCGCTAAGCGTGGAGCGGCCCGTTGATTAGCACCGTGCATACCATTCTGATGAACTAAGTGAACTATGAACTGAAAAGCACCCATAAATAAACCAAAGGCGTGCAGAACAGCGTGGCGTTTTAAAGCCATATTACCCGCGTGTATGGCCTGAGCATTGTACGATATACCGGGCATCGCGTCTACAGAAAACGGATTTACAGGCGCGGAGGTCCTCAGCTACCCCGCGAAGCAACGAAGCGAATGTGGTGTACCCGAGGTGGTCTTTATATTGCTTCAGATTTCCTGGTTCCTTGCTCTAGGGAGCATCTGTATGTGATCCGTCAGCCTCCACACTCTTCCCGGGCATATAGAACACAGATAACCCAAGTCCGCCAGAGTAGGGAGCTACCTACCTAGCACAGACGGTCAGAGGTTCGGACTACAGGCGGGAGGTATCGCTACAACGGTTTGGCCTAGCTGACCGTTTCCGATCAAGAGGCGTGAGTTAGATACGCGTACTGTTTGGGCAGGGGGGGTACGGTCTTCGGCCGCGAATCATCTGCTGGCCACTGCGAATTGCGACTCCCATGCCCGGAAATGTATCTAAGTGGCGTTTAGCGCGGTATCTAAAGGCCAGTCAGCCGCTGGCTAAAAGAGGGTTCGCTCAACCATAGAGAAATCAGGCCTTCACCAAACTTGCAGAGACAGGACAGCTCGATTTTTTTTGGTGTCTGGGCATTGGTTTGTCGCAGACGGTACCGATGCTCAATTGGACCGAGCGCCAGCTGCACGATGTAACTATGTGCTCCAGACGTCTCATTACGCCCGTACTGTTCAGCATATTAGGCTAAATCACCCTGGAGGGAATATTCGTTACACGGCCGCGCGATAGAGCAGACGGCTAAAACTCCGGGCTCTGGGACTGTGACAGGTGGGCCGCGTAGATCAACACTGTAAGGGCGGTGCGTCTTCGGGCATCAGGAGGAACGTAGTCGCTGCACCAGACTTCGACTCCACTCCATTCCAAAGTAGCTTTAGTAAATTAATTAGAATAGGGCGGGAGTCTTACACTTGACTTTGAAGTTACCGATGGCAAAGCTATTCATTGCGGTTTCGCCAGTGGCATACCTGCCATAGCAATTTTAACCATAGATACTAAATCGTGCTAGATTGTTAGCAAGATCTGCGTCGGAATGCCTGCGTGTAAGGTTAGCCTATGCGTACCTCTTTGATCCCAGGGTCACGTTTCCGATGGCCGGATAATACCCCTAACGACAGACTCGCAAACGGCGCGCGAAGTATCTGACGAAGCCTGAGGAGTCACATGTGCCGTACGTATAATATTGGAGACATCGTGAGGCGAACTTGAGAATTGATCGCTGGGTATCCGTACATCCTAAACAAATGTCGGAATTGCCATATGCTCTAGGGTTATGACAACGGCGTTGCAGCCCGACTCGTAGCGCAGATGTAAAAAAGGGAATTTGCGTGATACCGTCTACACCGTGGCTTCACGAGGATTCTAAAGATCTACGTTTATGTAGTGTCTATGGCAGTGAGTGAGGTCCATGCTCACCGTCCCACCTTACACAACCCCTGAAATGTGTACGACTCCGCTCTAAAAGGGGTCGACAGCCAGATATCCAACGTTTCCGCGTGATTAAAGTGGATGACAGTTTTTACAATTAAACGAGAGTTTCGGACGCCTTGGGCGTGTCGCATCTGTCGAAGTCCTGATTTATTTTGGATTCGCGCCCTCGCTCCCTTCCCAGTTTTTCCGAAGAATACTAGTTCTGAGATTCTTATCCGGCGCCTAACGCCATCTCACTATGCTACCATTGCCCGTAGTAGGACTATGACCTTACCACAACGATATCAGAAATACGACGGATTATCAAACGAGTCAGCCGCTTGGAGCGGATATATCGATCAATGGCACCGTTCGGGCTTTTGACGACTTCTGGAACACTCTGCACTGGGGAGCCTAACTGAATTCCAGGTCGAGTATCGTTCCTCCAGAGGAGCCAGGCCCGTAGTTACATTCATTGGTCGTACCTGAGGCCGTACTGACACGAGGAGTAACAGGTCTTGCGATAGACCCATTAGGAATGGCCATAATTCACCGCTACTGACCCTGCTTTTCGATAGGATGTCTTCGGACTAGCACAGTCCGTTGTGTACTCGACTAATGCATGGAGGAGCTCGCATATCCCTGTCCGATGAGGTCTTCACATCGCATCGTATTACGATAAAGAGTAGTCTTTGACCGCCATGTAGTGGATCCCTGCAGTAGGCCCACGTGGGTTGGGCTAGCGTTCATTCGTATCTGGATGGTAATGGATTATCGAATTTATGATTAAGTTTGGCTGCTGGAAATGCCGTAGTCGTAGGCATTTCATCTTCATGCGAGGCGTCCCGAGCAAAGGAAAGTCCGGTGTCGGGGTCAACTGACTCAGAGGGTTCTAGCGGTGGGTATGCAGTAGAACCGACTTGGGCTACATACATGCTGCGTACAAACGGTGCGCACGCAGCACCGCGGTAGAGGTAGTCCCGTATGCTGCACATCTGGTTAAACGCATTCCTTCCAGATCGTTGGTACTCTTTAGTGACGGACTACTATCGCCTTTGGAGAACCGCTATATAGACTTGAACGCCTGTTGTTGTTGAGGAGAGTTTCGCATTTGCGAATAACCCCTGTTAGTATCGAGCGTACAGGTGCCACTGACGCTATCAAGAGCCAAGGAACGACCCAGTTTGCTCCAGGGGAGCAAACATCGCGAGAGTTGGCGGCGTTGCCTATAGCCACGGGGCACCGTTGATCAACGTCCTTCCGGATCTCTGAAAACATGGAATTCAGGAGAACGGCCGTCCATCACCCACTTGCGCGTTAGTGGCTATACAGTTTAGCGCGGGGACGTTTACGATGTACTATTTGACCAGCAGAAACTCTGAAGCCCATCAAACGCAGGTGGGCAGGCTCTAACGACCCTGACAGTACTCCCAATATGGATGACGGCTGTCCCTGTCCTGGGCTCTCGGTCCTCAAACACTGACGAAGTATCTGGGTACTAGTGGATGTGACAATACTTATGGGCGTAGCGTTTCTGGTCAAAAGCCTTACTACGCGTGATAGGCACGACCCACAGGAACCGCAGGCATCGTATGTTAGTAAGATAGAAACGCGGGACGGGCGCTTTGCATACCCACTTATGCAAAACCCTATGTTTGCTCGAAGCTCTCGGATTAGCGCGATATCGATGTCATATGTACTAAGACTCCTGGCTTTCCGCAACCGTTCCGCCCTTACCAAATACTGGTATTCCACTTGCCCTCTAGGAAAGATCTGATTTGGGCGCGGGTCAGATGGGCTGAAATAGCGACATTACAAAAGCCCACAGCCGAATACGGACTTACTTAGCTGGCAACCTCTGGGCTCGGGGGCGAGCGAGGGGCATAGTAACCCAAGAAACTGCAAAAGCTGCATTATTGACTAATTACAGACATATTAGATCCTTCTGGGAGGTCGCTGAGGAAAGGCATTCTAATCCTCGTAAAAGGTCGCTAGGTTTTTTGGGAACTAGAGAAAACAAGATGATTCTAGAACGGCTTGATGTACACAAGGTCGGGTGGTAACCGTTAGTTTAATGGGTAGTATAAACAACGTAGTTCGTGAGTGAAGACTTATTCCTCGTGGGTAACCAGGTTCGTCCATTTTTGTTGACTCAGCCCCATGTCCATACGACCTGGCCAGGGAGAACCGCGGTGCCAGCTACTTGTTGCAGTGGTTCCTCTCGATAAGAGTATTAGTTTCCTCCTCTCCGGGAGTATGAGGATGGAATTGAATTTCAGAGACGGCTCCAATCCTTTGGGTCTAGTGCTTTGCGCGCCTAATACGAGACGACCTCCGATCTGCTCAGTCGGGATTTGATATCTTAGTTTGCGCTAGGTATGTGTGACGGTTACACAATTAGGCAACGGAGCGGGGCACGCTAACCACACTAACTTTTGAAACCGTGGCTAAAAGCAAGAGGAATTACCATTAAGACCTCCCCCTCCGCAGTTTCAACTTGGGGGTTCATTGGAACAAGGATCTTGCGAAGCGTCAGTTGAGATATAATAGTCATGATTCTATTAGCGGAGGACTTGCGATGCCCGGTAATCACTACCCCAGCCTAAGGAAGATGGTCTAAAATTCAGCTAATCTCACATGGGCTACCTGCCCTACACAACCCTGACGCCGGGGAGTTTATACAAGCACGCCCATCGCTGTTGGTTGCCTCATGCCTCTTGTTGACCGGCACTCATGTTTCTTCGGTTTGGAGTAGGGGCCTTTGACACTCCTTAGGTTTATACCGTGAACCCTTTTGATATTGATAAGTGCTTGCAATGCCTGAATAATCATCCACAGTTTAGATATCAAAATCTCAACACAAGACTTCACTGTTGGAAGGGACTGTGGGCGTGCATCACAAGTAATGTGAAGTCTTCTGGTAGAGCAGCAGACCCTGTCGAAGCTTCGCGGCAAGTAGCCCGCATAGACCAGTGGACCTCCGTCCTCTATCAGGGTATGCGGTGTCCTGTGTATATTGAAGGGCACTCGCAGATTGTATGCGGCCGGGTCCCCTCTGTGAGTGTGCAGACGCCTACGGCGATGCACGAAGAAAAGCTTAGTGGATTGCAGTCCCTGCGTTTTGGTAGACTATGATCGCACACAAAGGATGCAGTAAGTCTGATATTGTCGAGCATGGATTGATTGCTACGTTATGTAATAACGGGTTGTTTTAATCGATACGGCCTCCAATGGGTGTTAAACCCCCCTCCTGTGTTGCTAAGTCTGGGCGATGTTCTAGTCTGCGGGCTTGCAACCGAGTTGAAATGTATATCAACGTGTAAGCTAGCGCGACCTATCCTACTCCCCCTGTAAATCGATTGACGTTATAATGCGGGTCACCTGGCGATCCAGGAGGTTCGTCTGCCCGGAAGAAGGATTTGTACAATGCGGTAAATGGCGTAAGCATGCCGTACGAGAGGTACGATACCCTTGAGAGGTGGGAGTAATGTGCTATGCTGGCTTCGCTTGAACAGCTGCTAGTTGCTACAATCCTAGTGACACCTATCGCGATGCCCGCTACGAGTACCTCCCTTAGTCTTCCCCGGTTCAGTTCAGACGTAAGCTTCTTGGCATCTCCAATTGCTGATATTCACATGACTAACAGTTGCGGGAGAAATTGGTTGCCGATGAAGTGCCCACCAAGTGGCCCGGAACAAACTAAAGTCTACTGCTTATGGGGCCCCCTGTGCTCTACCGGGTTTGTAGGTAAGTTCGTTGACATATCCAGCGATATGATTGGTATCGTAACAACTCGCTCGGCACCCAGCATCGCTACTGTGCTAACTTCGATGGAGCGCTGTAACGCGCCTTTTTATATGACCTGGCATTGATCTTTTACTCTTGAATATGACCACAGGACTTCACCAGAGAGATTCAGTGACGACAAAAGTGGCGGTTTTTTAAACTCGCGAATGACTCAGTATTAGAGTGTCCAGAAAGCCCTTGCTCTATATCCCGTTAGTCGTTCTACAATGAAGAGTAATAGGCATGTCGCATTTCCTGCCGACACTTTGACGACGCATTTATACGAAACACGTACCGAAAACTGCAAAAAGGTAACTGAACGTTTCCAAGGTAAGCAAGATTAAAACATAAAGATGCACGTAGAGAGGTCATCGAGGTGCCTGGCCGCAGTCAGCTGTCGGGCGCCTCATTTTTTATACCTTCCACTGCTCAAACTCCTCTGTGAAGCGTCTAAGATTTAGACAGCCTAAAGTGCTAATCACTGTAGTTTACAATGTTTAGTACACAAATCGTGGGATCGAGAGCCGATCGGTCTATTACGACAATCACTCAACGAAGATCGATATACCTAGTAGACTTGTCTGAGCACTTAGGGAGGATGTTTGACCGAGACAAGCGTTCGCCACCTTTCCATTCAATTTGCTGGTGGTTCGCTTGGATTATTGCTCTTCCCTCCTCCGTCGTACCAAGTTACTTTGTTCGCAGACATCAACCTAATCCCATCAGATTTTACTTGAGGCCCTGTCCGGTCAGCCAGCCGTAGTATCAAAGGACTCGAACGCATGTTGAAGTTGCCTCGAGCGCACCCTCCGCTTGCTCACGAACTCTGTTCTTGTACTTAGTTCTATGGGTAGGTTGTGCCTGTTATGAGCCTAAGAACCCAAGCGCCCCGGTGTCGGAAGACGAATAACTGAGCGAATACGTCCTACACAGTTAGGGAAACAACTATGCATATAATGACGCGGTGTGTACAGTGTGCGTGCAGTTTCTTCCAGTCTCTTTCGCTTCATCACGTACGTGTGCTTCAAGACCTAATTATAGTCGTTTGCGCACATCACCCCCCTCAAGAATCATGCGAATTGCATGAATGTGAAGGAGACAACCCCTCTAGGGTTGATCTATATGTGAGATTATAGCCGCCAAACGATTGTCTCAACACTACGGGTGCCAGAAGGGTAACACTCTCTTCTTCCGTCGACGTGCTCACACCTGAGTGCGGCACAGATATACTCATACATATCCTTCAATCGTGAGGTCTAGGGGCGAGCGCAGCGGACACCTCATCCAAGTGCAGATAGAAGCAGTCTTTCTACGGAAATTAGAATGCGCCGTCCAAATCTCCTCCTACGTTTTCCCCTCTACTCTTAGAGCGACGCAGGAGAAAGCTCAATGAATGTTAGCAAAGTTTCATTGGGTCAATCTCATATCATAAGAGAAAGTCTCGTAACCTCGACCCGTCTTGGAATGCATTCTTATCGGCAGCACCTAGTGTATCCAGCGGTTGGTTATACCAGACACACTAGCAAATGCAGTCACATGTAATCAGAGTAGGGTGAGAAATCACACTGCCTGCTAGAATGATCTTTCTTCCGGGCGTAGACTTGCAGAGCGTAGAATACATTATTACTCTCGTGGATAATCAGGATCACTACATTATCGGTACATGTCAATCTAGGCATCGATGGAGTGAAACCCTAGTCGGGATAGAGAATAGAAGGCTGTCACCTCATCTATCACGGTCTACGTACGAGACAGGGTCCCTCTGCATGCACAGAGCCAGTAATAGTACTCCGGGGAAGCCACCCCGTCCTGTGGGGTTATCTTTATCTCCCGTGCCAACTACAACTGCCTGGGGGCTCCGTGATACGCCATCGGTACTTAATAGGCGGCCAGTTACGAACATTTTAAAACCGCTAGGGGTACACCTTCAGCGCGGTAAAGACACCAGTATTCTGTTATTCACACTCAGAGTGCCCCGGAACTATTAGCGGGGCCCATGCGTCGGTTTGCACGCAGTCCGTCGCACTGTGTTCAACGGGGGCAGTTTACAACAGTTACTCTGCGTAGGGTAAGGCATAAACTCCCGTCCTGGCCATGTACGGGCTAAGATGGTCGCTTTTGGGATAATTATCCTCGTTCATGGAGTTAGATGGTCTACTATGCTTATGGGGCCTAGTATGTCTGTCCTAAGAATTGATTGATCTCGACCACATGGACATTACTATTAAATGAACATTCCCGTTGGACAGTCATTTGTGCTCGCAGTCCATACTATCCTAGAAGGGGGTAATAGCCACACTTTGTAAGGATAAATTGCGCACGGGGTGTCTCCATAGCCTAGTGGTGTCGATGAGGGTCTAAGTCCAAGTTTCCGTACCAGAGCCTGGATGGATATGCGGAATACCGGAGTTTGGCTTAGAGCTGTAGACCCTCACATCAGATTTCGAGCCTCGACTTGAGGTTCTTCCTGCATTCTAACGCTGTCCGCTTACCCTGATCAGTTAGCATGGCGGGGACGAAGTGAATACCCTGCCACACCTTGTTTCCTTTAGACACTATCGCTCGTACGCGACGCTAGACACAATGAACTACGCGGCGAAGGCTATATTATTATTACAAGCTGCGGTTGATGTAGTTCTATGGACTAACGGCATGACATGAAGAGGTTATTCACCAACCGAACGACATCCAGTACTAGTGACAGAGCAATTTCTTTCATGAGTGCCCCGTCCAGACAACGCTATAGACGGGGTGCTGGTCCTGAATATCATTACGAGCTCTTACGTAAAGCGCAACGAGAAGTATGCCTCGTGCTGTCTCCTCGTTGTACTAACATTATCTACGACATACCAAACTGTTCAAATAAATTCATGGTTGGAGGATCGGGCATTCATGTGTTATGCCTCCGTGAGGGTTCGAACTTAAGTTGGTCGTTAACCATTCTCGTGTCCGAGCATCATAATAGAATATGCCTCTTATACCAGTCAGGGATCCAACTTAGATGAACGACTCACACCCTCCAGACAAGTCTCACGAGTTGCTCACGCCCCGCACACGTTTTGACAACACGAGACATGGAGCTCCCAGCTTTGGACAAACTGATCGGTCCAAAGGCAGAATCAGACCATGACGGCACTTAAGTGAGGCTATTCGGGAACGGAACTAGATGGTGCTTCTACTTGCAGTCACGGTGTACGTGGAGAATCCCACCAATATCTATAGTAATCCACAAGCTTCTGTTCTAACCTAGATCCTCCATCAAGTTATTCCGAGGGATTGGGTAACCCTTCGGGTACAGTAACTGATGGTAGATCTCCATTCCGCGCAATTAGTGACGGCAGGAGAAATGTGGCTTCTTTTTCCATGTTAAAGCTGCAAGCCCAAGTGTACAAGTTGCTCGTATCCCCCCGATCTCTAATAA'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c45155cb-ff9e-41c6-a7b0-3ada05cfd936",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting PyGObject\n",
      "  Downloading pygobject-3.50.0.tar.gz (1.1 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.1/1.1 MB\u001b[0m \u001b[31m10.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Installing build dependencies ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[79 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Collecting meson-python>=0.12.1\n",
      "  \u001b[31m   \u001b[0m   Downloading meson_python-0.17.1-py3-none-any.whl.metadata (4.1 kB)\n",
      "  \u001b[31m   \u001b[0m Collecting pycairo>=1.16\n",
      "  \u001b[31m   \u001b[0m   Downloading pycairo-1.27.0.tar.gz (661 kB)\n",
      "  \u001b[31m   \u001b[0m \u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/661.5 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m661.5/661.5 kB\u001b[0m \u001b[31m5.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \u001b[?25h  Installing build dependencies: started\n",
      "  \u001b[31m   \u001b[0m   Installing build dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: started\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Installing backend dependencies: started\n",
      "  \u001b[31m   \u001b[0m   Installing backend dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): finished with status 'error'\n",
      "  \u001b[31m   \u001b[0m   \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   \u001b[31m×\u001b[0m \u001b[32mPreparing metadata \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mpyproject.toml\u001b[0m\u001b[1;32m)\u001b[0m did not run successfully.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m╰─>\u001b[0m \u001b[31m[50 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[36m\u001b[1m+ meson setup /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39 /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39/.mesonpy-db1o3mrk -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md -Dwheel=true -Dtests=false --native-file=/private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39/.mesonpy-db1o3mrk/meson-python-native-file.ini\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m The Meson build system\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Version: 1.6.0\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Source dir: /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Build dir: /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39/.mesonpy-db1o3mrk\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Build type: native build\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Project name: pycairo\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Project version: 1.27.0\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m C compiler for the host machine: cc (clang 15.0.0 \"Apple clang version 15.0.0 (clang-1500.3.9.4)\")\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m C linker for the host machine: cc ld64 1053.12\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Host machine cpu family: aarch64\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Host machine cpu: aarch64\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Program python3 found: YES (/Users/arianemora/miniconda3/envs/oligo/bin/python3.10)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wall: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Warray-bounds: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wcast-align: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wconversion: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wextra: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wformat=2: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wformat-nonliteral: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wformat-security: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wimplicit-function-declaration: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Winit-self: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Winline: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wmissing-format-attribute: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wmissing-noreturn: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wnested-externs: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wold-style-definition: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wpacked: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wpointer-arith: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wreturn-type: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wshadow: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wsign-compare: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wstrict-aliasing: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wundef: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wunused-but-set-variable: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wswitch-default: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wno-missing-field-initializers: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wno-unused-parameter: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -fno-strict-aliasing: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -fvisibility=hidden: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Did not find pkg-config by name 'pkg-config'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Found pkg-config: NO\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Did not find CMake 'cmake'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Found CMake: NO\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Run-time dependency cairo found: NO (tried framework)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m ../cairo/meson.build:31:12: ERROR: Dependency lookup for cairo with method 'pkgconfig' failed: Pkg-config for machine host machine not found. Giving up.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m A full log can be found at /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39/.mesonpy-db1o3mrk/meson-logs/meson-log.txt\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  \u001b[31m   \u001b[0m \u001b[1;31merror\u001b[0m: \u001b[1mmetadata-generation-failed\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m \u001b[31m×\u001b[0m Encountered error while generating package metadata.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m╰─>\u001b[0m See above for output.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m \u001b[1;35mnote\u001b[0m: This is an issue with the package mentioned above, not pip.\n",
      "  \u001b[31m   \u001b[0m \u001b[1;36mhint\u001b[0m: See above for details.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "\u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! pip install PyGObject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ef38f9d-dc30-4799-9d9f-d6bc1c6b3aff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "dd368813-dff0-4682-943c-1e7ec2d2a7c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dnaweaver as dw\n",
    "import time\n",
    "import dnachisel as dnachisel \n",
    "from sciutil import SciUtil\n",
    "from Bio.Seq import Seq\n",
    "from difflib import SequenceMatcher\n",
    "import primer3\n",
    "\n",
    "u = SciUtil()\n",
    "\n",
    "from primer3 import calcHairpin, calcHomodimer\n",
    "\n",
    "def check_secondary_structure(sequence):\n",
    "    \"\"\"\n",
    "    Check secondary structures like hairpins and homodimers in a given primer sequence.\n",
    "    \n",
    "    Args:\n",
    "        sequence (str): The DNA sequence of the primer to analyze.\n",
    "        \n",
    "    Returns:\n",
    "        dict: Results for hairpin and homodimer properties.\n",
    "    \"\"\"\n",
    "    # Check for hairpin structure\n",
    "    hairpin_result = calcHairpin(sequence)\n",
    "    hairpin_info = {\n",
    "        \"hairpin_found\": hairpin_result.structure_found,\n",
    "        \"hairpin_tm\": hairpin_result.tm,\n",
    "        \"hairpin_dg\": hairpin_result.dg,\n",
    "        \"hairpin_dh\": hairpin_result.dh,\n",
    "        \"hairpin_ds\": hairpin_result.ds,\n",
    "    }\n",
    "\n",
    "    # Check for homodimer structure\n",
    "    homodimer_result = calcHomodimer(sequence)\n",
    "    homodimer_info = {\n",
    "        \"homodimer_found\": homodimer_result.structure_found,\n",
    "        \"homodimer_tm\": homodimer_result.tm,\n",
    "        \"homodimer_dg\": homodimer_result.dg,\n",
    "        \"homodimer_dh\": homodimer_result.dh,\n",
    "        \"homodimer_ds\": homodimer_result.ds,\n",
    "    }\n",
    "\n",
    "    # Combine results\n",
    "    return {\"hairpin\": hairpin_info, \"homodimer\": homodimer_info}\n",
    "\n",
    "def build_oligos(seq_id: str, sequence: str, output_directory: str, min_gc=0.3, max_gc=0.7, min_tm=55, max_tm=70, min_segment_length=40, max_segment_length=100, max_length=1500):\n",
    "    \"\"\" Use DNAweaver to build oligos \"\"\"\n",
    "    # Here we use a comercial supplier but don't actually care. \n",
    "    cheap_dna_offer = dw.CommercialDnaOffer(\n",
    "        name=\"CheapDNA.\",\n",
    "        sequence_constraints=[\n",
    "            dw.NoPatternConstraint(enzyme=\"BsaI\"),\n",
    "            dw.SequenceLengthConstraint(max_length=4000),\n",
    "            dw.GcContentConstraint(min_gc=min_gc, max_gc=max_gc)\n",
    "        ],\n",
    "        pricing=dw.PerBasepairPricing(0.10),\n",
    "    )\n",
    "\n",
    "    oligo_dna_offer = dw.CommercialDnaOffer(\n",
    "        name=\"OliGoo\",\n",
    "        sequence_constraints=[\n",
    "            dw.GcContentConstraint(min_gc=min_gc, max_gc=max_gc),\n",
    "            dw.SequenceLengthConstraint(max_length=4000),\n",
    "        ],\n",
    "        pricing=dw.PerBasepairPricing(0.07),\n",
    "        memoize=True\n",
    "    )\n",
    "\n",
    "    oligo_assembly_station = dw.DnaAssemblyStation(\n",
    "        name=\"Oligo Assembly Station\",\n",
    "        assembly_method=dw.OligoAssemblyMethod(\n",
    "            overhang_selector=dw.TmSegmentSelector(\n",
    "                min_size=15, max_size=25, min_tm=min_tm, max_tm=max_tm\n",
    "            ),\n",
    "            min_segment_length=min_segment_length,\n",
    "            max_segment_length=max_segment_length,\n",
    "            sequence_constraints=[dw.SequenceLengthConstraint(max_length=4000)],\n",
    "            duration=8,\n",
    "            cost=30,\n",
    "        ),\n",
    "        supplier=oligo_dna_offer,\n",
    "        coarse_grain=20,\n",
    "        a_star_factor=\"auto\",\n",
    "        memoize=True,\n",
    "    )\n",
    "\n",
    "    assembly_station = dw.DnaAssemblyStation(\n",
    "        name=\"Gibson Assembly Station\",\n",
    "        assembly_method=dw.GibsonAssemblyMethod(\n",
    "            overhang_selector=dw.TmSegmentSelector(min_tm=min_tm, max_tm=max_tm),\n",
    "            min_segment_length=min_segment_length,\n",
    "            max_segment_length=max_segment_length + 20, # add a bit of a buffer\n",
    "        ),\n",
    "        supplier=[cheap_dna_offer, oligo_assembly_station],\n",
    "        logger=\"bar\",\n",
    "        coarse_grain=100,\n",
    "        fine_grain=10,\n",
    "        a_star_factor=\"auto\",\n",
    "    )\n",
    "    \n",
    "    print(\"Looking for the best assembly plan...\")\n",
    "    t0 = time.time()\n",
    "    quote = assembly_station.get_quote(sequence, with_assembly_plan=True)\n",
    "    assembly_plan_report = quote.to_assembly_plan_report()\n",
    "    #assembly_plan_report.write_full_report(f\"{output_directory}/oligo_assembly_plan_{seq_id}.zip\")\n",
    "    original_sequence = assembly_plan_report.plan.sequence\n",
    "    # Then get the sequence \n",
    "    rows = []\n",
    "    for oligo in assembly_plan_report.plan.assembly_plan:\n",
    "        # If this was chosen then choose it\n",
    "        if oligo.accepted:\n",
    "            rows.append([oligo.id, oligo.sequence, original_sequence])\n",
    "    return rows\n",
    "\n",
    "def get_oligos(df, protein_column, id_column, output_directory, forward_primer: str, reverse_primer: str, min_overlap=10, min_gc=0.3, \n",
    "               max_gc=0.7, min_tm=55, max_tm=70, min_segment_length=40, max_segment_length=100, max_length=1500):\n",
    "    \"\"\" Get the oligos for a dataframe \"\"\"\n",
    "    rows = []   \n",
    "    for seq_id, protein_sequence in df[[id_column, protein_column]].values:\n",
    "        # Add on the primers that the user has provided\n",
    "        optimzed_sequence = codon_optimize(protein_sequence, min_gc, max_gc)\n",
    "        if optimzed_sequence[:3] != \"ATG\":\n",
    "            u.dp([f\"Warning: {seq_id} does not start with a methionine. \", optimzed_sequence[:3]])\n",
    "            if 'ATG' not in forward_primer:\n",
    "                u.warn_p([f\"Warning: {seq_id} does not start with a methionine. AND you don't have a methonine in your primer!!\", forward_primer])\n",
    "                print(\"We expect the primer to be in 5 to 3 prime direction.\")\n",
    "        # ALso check the end and or the reverse primer check for the three ones\n",
    "        if optimzed_sequence[-3:] != \"TAA\" and optimzed_sequence[-3:] != \"TGA\" and optimzed_sequence[-3:] != \"TAG\":\n",
    "            u.dp([f\"Warning: {seq_id} does not end with a stop codon. \", optimzed_sequence[-3:]])\n",
    "            if 'TAA' not in reverse_primer and \"TGA\" not in reverse_primer and \"TAG\" not in reverse_primer:\n",
    "                u.warn_p([f\"Warning: {seq_id} does not end with a stop codon. AND you don't have a stop codon in your primer!!\", reverse_primer])\n",
    "                print(\"We expect the primer to be in 5 to 3 prime direction.\")\n",
    "        print(optimzed_sequence)\n",
    "        codon_optimized_sequence = forward_primer + optimzed_sequence + reverse_primer\n",
    "        # Check now some simple things like that there is \n",
    "        oligos = build_oligos(seq_id, codon_optimized_sequence, output_directory, min_gc, max_gc, min_tm, max_tm, min_segment_length, max_segment_length, max_length)\n",
    "        prev_oligo = None\n",
    "        for i, oligo in enumerate(oligos):\n",
    "            seq = oligo[1]\n",
    "            # CHeck that there is an overlap with the previous sequence and that it is not too short\n",
    "            # Also make sure we swap the directions of the oligos so they automatically anneal\n",
    "            # Also assert that the start is a methionine (and if not warn it... )\n",
    "            primer_overlap = None\n",
    "            primer_tm = None\n",
    "            primer_len = None\n",
    "            homodimer_tm = None\n",
    "            hairpin_tm = None\n",
    "            if prev_oligo:\n",
    "                # Get the overlap with the previous sequence\n",
    "                match = SequenceMatcher(None, prev_oligo, seq).find_longest_match()\n",
    "                primer_overlap = prev_oligo[match.a:match.a + match.size]\n",
    "                # Analyze the primer sequence\n",
    "                results = check_secondary_structure(primer_overlap)\n",
    "                homodimer_tm = results['homodimer']['homodimer_tm']\n",
    "                hairpin_tm = results['hairpin']['hairpin_tm']\n",
    "                primer_tm = primer3.bindings.calcTm(primer_overlap)\n",
    "                primer_len = len(primer_overlap)\n",
    "            prev_oligo = seq\n",
    "            if i % 2 == 0:\n",
    "                seq = str(Seq(seq).reverse_complement())\n",
    "            oligo_tm = primer3.bindings.calcTm(seq)\n",
    "            rows.append([seq_id, oligo[0], seq, len(seq), oligo_tm, primer_overlap, primer_tm, primer_len, homodimer_tm, hairpin_tm, oligo[2]])\n",
    "    oligo_df = pd.DataFrame(rows, columns=[\"id\", \"oligo_id\", \"oligo_sequence\", \"oligo_length\", \"oligo_tm\", \"primer_overlap_with_previous\", \"overlap_tm_5prime\", \"overlap_length\", \n",
    "                                            \"overlap_homodimer_tm\", \"overlap_hairpin_tm\", \"original_sequence\"])\n",
    "    return oligo_df\n",
    "\n",
    "def codon_optimize(protein_sequence: str, min_gc=0.3, max_gc=0.7):\n",
    "    \"\"\" Codon optimize the protein sequence using DNA chisel: https://github.com/Edinburgh-Genome-Foundry/DnaChisel\"\"\"\n",
    "    seq = dnachisel.reverse_translate(protein_sequence)\n",
    "    problem = dnachisel.DnaOptimizationProblem(\n",
    "        sequence=seq,\n",
    "        constraints=[\n",
    "            AvoidPattern(\"BsaI_site\"),\n",
    "            EnforceGCContent(mini=min_gc, maxi=max_gc, window=50),\n",
    "        ],\n",
    "        objectives=[CodonOptimize(species='e_coli', location=(0, len(seq)))]\n",
    "    )\n",
    "    # SOLVE THE CONSTRAINTS, OPTIMIZE WITH RESPECT TO THE OBJECTIVE\n",
    "    problem.resolve_constraints()\n",
    "    problem.optimize()\n",
    "\n",
    "    # PRINT SUMMARIES TO CHECK THAT CONSTRAINTS PASS\n",
    "    print(problem.constraints_text_summary())\n",
    "    print(problem.objectives_text_summary())\n",
    "\n",
    "    # GET THE FINAL SEQUENCE (AS STRING OR ANNOTATED BIOPYTHON RECORDS)\n",
    "    final_sequence = problem.sequence  # string\n",
    "    final_record = problem.to_record(with_sequence_edits=True)\n",
    "    return final_sequence\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "683b7fd8-91bc-4676-8ed4-f7e82fdc31f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "objective:   0%|                                                  | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-912](e_coli...]\n",
      "location:   0%|                                                                         | 0/238 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                          | 0/238 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-912](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-912](mini:0.25, maxi:0.65, window:50)\n",
      "      │ Passed !\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:         0\n",
      "✔        0 ┍ MaximizeCAI[0-912](e_coli) \n",
      "           │ Codon opt. on window 0-912 scored -0.00E+00\n",
      "\n",
      "\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m              Warning: Seq1 does not end with a stop codon. \tCGC\t               \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "ATGCGCACCTATTTTTTTGTGGCGGCGTTTGGCTTTGCGTTTGCGGCGGATTTTCCGGTGGTGGATCCGTTTGCGCAGCCGGATCATGAAGCGACCATGGATCGCTATTTTTTTCAGCGCCAGGATTTTCCGTATCGCTTTTATGCGGTGGATTTTGATTATGTGGATGGCCAGCGCCATTATCGCTTTTGGTTTGGCCGCCCGTTTCAGGCGCATCCGGCGGATGGCTATCCGGATGATTGGATGTTTGATGGCAACGCGGCGGTGGGCGCGTTTGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTTTTTGTGGCGATTGGCTATCGCACCCCGTTTCGCATTGATCGCGCGGGCCTGACCTTTGATTATACCCCGGCGTATCCGGGCCATGATGATCAGCGCGATCCGTTTAACGGCTTTCCGTATGGCGGCGCGGATGCGTTTTTTGATTTTTTTCGCGATGGCATGCGCCCGGCGGTGGATGCGCATGATCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCCATTATTATGGCGGCTTTTTTGTGTTTCATGCGTTTTTTACCCGCCCGGGCGAATTTGCGCGCTATGCGGCGGATTATCCGTATTTTTGGTGGCGCGATGGCGCGATTTTTGGCGAACGCGCGGGCTTTGAACAGCTGTTTCGCGGCAAACGCGCGGAATTTTTTTTTTGGCGCGGCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACCGGGCCATGCGATGGCGCGCTTTGTGGATGATTTTCGCCGCGATGATGGCTTTACCTTTGATTTTCAGCCGTTTGATGGCTTTGGCCATGGCGAAACCTTTGGCGCGTATTTTCGCTTTTTTTTTGCGCGCCCGGCGGTGGAACGCCAGCGC\n",
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n",
      "constraint:   0%|                                                 | 0/2 [00:00<?, ?it/s, now=AvoidPattern[0-954](patte...]\n",
      "location:   0%|                                                                           | 0/8 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                           | 0/8 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "objective:   0%|                                                  | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-954](e_coli...]\u001b[A\n",
      "location:   0%|                                                                         | 0/186 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "location:   0%|                                                                          | 0/186 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\n",
      "location:   0%|                                                                          | 0/186 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-954](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-954](mini:0.25, maxi:0.65, window:50)\n",
      "      │ Passed !\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:         0\n",
      "✔        0 ┍ MaximizeCAI[0-954](e_coli) \n",
      "           │ Codon opt. on window 0-954 scored -0.00E+00\n",
      "\n",
      "\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m              Warning: Seq2 does not end with a stop codon. \tGGC\t               \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "ATGCTGCTGTTTGATTTTCCGTTTGATCAGTTTCAGACCTATAAACCGGAAAAAGCGGCGCCGAAAGATTTTTGGGAATTTTGGAAATGCTATTTTGAAGAATTTGCGAAAGTGCAGTATGAACCGGATTTTCAGCCGGTGGATTATCCGGCGGATGGCGTGAAAGTGTATCGCTTTACCTATAAATATTTTGGCAACGCGCGCATTGCGGGCTGGTATGCGGTGCCGGATAAAGAAGTGCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTTCATGGCTATGCGACCTTTGGCATGTTTGTGCGCGGCCAGCAGTATTATGAAGATACCTATATTTATCCGCATGGCCATGCGTTTGGCTGGATGACCAAAGGCATTTTTGATAAAGATACCGATCATTTTCGCGGCGTGTATTTTGATGCGGTGCGCGCGTTTGAAGTGATTTATTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCTATCAGGGCGTGGGCTTTAACATTGATGCGGATGCGTTTTATGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATTTTTATAACTTTGAACGCGCGATTGATGTGGCGTTTGAACAGCCGTATTTTGAAATTAACTATTTTGTGCGCCGCAACGGCTATCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGTATTTTTTTGATATTCTGAACTTTGCGGATCGCGTGAAAGTGCCGGTGTTTATGTATATTGGCTTTATTGATAAAGTGACCCCGCCGACCACCGTGTTTGCGGGCTATAACCAGTTTGCGACCAAAAAAGAATTTAACGTGTATCGCTATTTTGGCGATGTGTATATTCCGGCGTTTCAGACCGAAAAATTTGCGTTTTTTAAACAGCATTTTAAAGGC\n",
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>oligo_id</th>\n",
       "      <th>oligo_sequence</th>\n",
       "      <th>oligo_length</th>\n",
       "      <th>oligo_tm</th>\n",
       "      <th>primer_overlap_with_previous</th>\n",
       "      <th>overlap_tm_5prime</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>overlap_homodimer_tm</th>\n",
       "      <th>overlap_hairpin_tm</th>\n",
       "      <th>original_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00000</td>\n",
       "      <td>CAAACGGATCCACCACCGGAAAATCCGCCGCAAACGCAAAGCCAAA...</td>\n",
       "      <td>119</td>\n",
       "      <td>74.669187</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00001</td>\n",
       "      <td>CGGTGGTGGATCCGTTTGCGCAGCCGGATCATGAAGCGACCATGGA...</td>\n",
       "      <td>128</td>\n",
       "      <td>82.149159</td>\n",
       "      <td>CGGTGGTGGATCCGTTTG</td>\n",
       "      <td>56.343909</td>\n",
       "      <td>18.0</td>\n",
       "      <td>7.388261</td>\n",
       "      <td>41.578087</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00002</td>\n",
       "      <td>GCGAAAGGTATATTCATCAAACGCGCCCACCGCCGCGTTGCCATCA...</td>\n",
       "      <td>129</td>\n",
       "      <td>85.231564</td>\n",
       "      <td>GATGGCCAGCGCCATTAT</td>\n",
       "      <td>55.811457</td>\n",
       "      <td>18.0</td>\n",
       "      <td>17.815718</td>\n",
       "      <td>68.998303</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00003</td>\n",
       "      <td>TGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.057453</td>\n",
       "      <td>TGATGAATATACCTTTCGC</td>\n",
       "      <td>47.775584</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-64.316560</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00004</td>\n",
       "      <td>AAACGGATCATGCGCATCCACCGCCGGGCGCATGCCATCGCGAAAA...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.703740</td>\n",
       "      <td>CCCCGGCGTATCCGGGCC</td>\n",
       "      <td>65.660649</td>\n",
       "      <td>18.0</td>\n",
       "      <td>30.462576</td>\n",
       "      <td>75.699579</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00005</td>\n",
       "      <td>GATGCGCATGATCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCC...</td>\n",
       "      <td>128</td>\n",
       "      <td>82.795446</td>\n",
       "      <td>GATGCGCATGATCCGTTT</td>\n",
       "      <td>54.144649</td>\n",
       "      <td>18.0</td>\n",
       "      <td>30.763247</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00006</td>\n",
       "      <td>ACGCCGGTTCCGCATAGCCGCGCCAAAAAAAAAATTCCGCGCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.057453</td>\n",
       "      <td>TGCGGCGGATTATCCGTA</td>\n",
       "      <td>56.451446</td>\n",
       "      <td>18.0</td>\n",
       "      <td>30.590793</td>\n",
       "      <td>59.177368</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00007</td>\n",
       "      <td>GCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACC...</td>\n",
       "      <td>128</td>\n",
       "      <td>83.764878</td>\n",
       "      <td>GCTATGCGGAACCGGCGT</td>\n",
       "      <td>61.126936</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.010537</td>\n",
       "      <td>40.188310</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00008</td>\n",
       "      <td>TTCCACCGCCGGGCGCGCAAAAAAAAAGCGAAAATACGCGCCAAAG...</td>\n",
       "      <td>78</td>\n",
       "      <td>81.265500</td>\n",
       "      <td>CAGCCGTTTGATGGCTTT</td>\n",
       "      <td>54.166830</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10.864156</td>\n",
       "      <td>58.661393</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00009</td>\n",
       "      <td>GCGCGCCCGGCGGTGGAACGCCAGCGCCTCGAGCACCACCACCACC...</td>\n",
       "      <td>73</td>\n",
       "      <td>80.600609</td>\n",
       "      <td>GCGCGCCCGGCGGTGGAA</td>\n",
       "      <td>69.523553</td>\n",
       "      <td>18.0</td>\n",
       "      <td>31.975298</td>\n",
       "      <td>63.355769</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00000</td>\n",
       "      <td>AAAAATCTTTCGGCGCCGCTTTTTCCGGTTTATAGGTCTGAAACTG...</td>\n",
       "      <td>119</td>\n",
       "      <td>71.540938</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00001</td>\n",
       "      <td>CGGCGCCGAAAGATTTTTGGGAATTTTGGAAATGCTATTTTGAAGA...</td>\n",
       "      <td>129</td>\n",
       "      <td>79.780705</td>\n",
       "      <td>CGGCGCCGAAAGATTTTT</td>\n",
       "      <td>55.017416</td>\n",
       "      <td>18.0</td>\n",
       "      <td>44.859279</td>\n",
       "      <td>42.830305</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00002</td>\n",
       "      <td>TTCGCCATCATAATACGCGTTATAGCCATGATATTTCACAATCGCC...</td>\n",
       "      <td>130</td>\n",
       "      <td>81.266757</td>\n",
       "      <td>AGTGTATCGCTTTACCTATA</td>\n",
       "      <td>48.313698</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-41.223309</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00003</td>\n",
       "      <td>CGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>80.856584</td>\n",
       "      <td>CGCGTATTATGATGGCGAA</td>\n",
       "      <td>53.868963</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10.800809</td>\n",
       "      <td>49.069751</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00004</td>\n",
       "      <td>TCATCCACTTCATCAAAATAATAAATCACTTCAAACGCGCGCACCG...</td>\n",
       "      <td>130</td>\n",
       "      <td>79.675896</td>\n",
       "      <td>ATCCGCATGGCCATGCGT</td>\n",
       "      <td>60.798906</td>\n",
       "      <td>18.0</td>\n",
       "      <td>57.093991</td>\n",
       "      <td>67.117879</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00005</td>\n",
       "      <td>TTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGC...</td>\n",
       "      <td>131</td>\n",
       "      <td>81.151404</td>\n",
       "      <td>TTATTTTGATGAAGTGGATGA</td>\n",
       "      <td>48.512171</td>\n",
       "      <td>21.0</td>\n",
       "      <td>-138.759018</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00006</td>\n",
       "      <td>GGGTTTTCATCGCCTGCACTTCGGTTTCCGGATAGCCGTTGCGGCG...</td>\n",
       "      <td>130</td>\n",
       "      <td>80.312240</td>\n",
       "      <td>GCGGATTATCCGTATTTTTATA</td>\n",
       "      <td>49.832919</td>\n",
       "      <td>22.0</td>\n",
       "      <td>30.590793</td>\n",
       "      <td>59.177368</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00007</td>\n",
       "      <td>TGCAGGCGATGAAAACCCTGTATTTTTTTGATATTCTGAACTTTGC...</td>\n",
       "      <td>128</td>\n",
       "      <td>80.210296</td>\n",
       "      <td>TGCAGGCGATGAAAACCC</td>\n",
       "      <td>56.217177</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-38.140845</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00008</td>\n",
       "      <td>GCCTTTAAAATGCTGTTTAAAAAACGCAAATTTTTCGGTCTGAAAC...</td>\n",
       "      <td>129</td>\n",
       "      <td>77.856873</td>\n",
       "      <td>ACCGTGTTTGCGGGCTAT</td>\n",
       "      <td>57.547578</td>\n",
       "      <td>18.0</td>\n",
       "      <td>9.356132</td>\n",
       "      <td>54.474060</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00009</td>\n",
       "      <td>TTAAACAGCATTTTAAAGGCCTCGAGCACCACCACCACCACCACTG...</td>\n",
       "      <td>66</td>\n",
       "      <td>71.478719</td>\n",
       "      <td>TTAAACAGCATTTTAAAGGC</td>\n",
       "      <td>48.268215</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6.193641</td>\n",
       "      <td>29.811277</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id oligo_id                                     oligo_sequence  \\\n",
       "0   Seq1  S_00000  CAAACGGATCCACCACCGGAAAATCCGCCGCAAACGCAAAGCCAAA...   \n",
       "1   Seq1  S_00001  CGGTGGTGGATCCGTTTGCGCAGCCGGATCATGAAGCGACCATGGA...   \n",
       "2   Seq1  S_00002  GCGAAAGGTATATTCATCAAACGCGCCCACCGCCGCGTTGCCATCA...   \n",
       "3   Seq1  S_00003  TGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTT...   \n",
       "4   Seq1  S_00004  AAACGGATCATGCGCATCCACCGCCGGGCGCATGCCATCGCGAAAA...   \n",
       "5   Seq1  S_00005  GATGCGCATGATCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCC...   \n",
       "6   Seq1  S_00006  ACGCCGGTTCCGCATAGCCGCGCCAAAAAAAAAATTCCGCGCGTTT...   \n",
       "7   Seq1  S_00007  GCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACC...   \n",
       "8   Seq1  S_00008  TTCCACCGCCGGGCGCGCAAAAAAAAAGCGAAAATACGCGCCAAAG...   \n",
       "9   Seq1  S_00009  GCGCGCCCGGCGGTGGAACGCCAGCGCCTCGAGCACCACCACCACC...   \n",
       "10  Seq2  S_00000  AAAAATCTTTCGGCGCCGCTTTTTCCGGTTTATAGGTCTGAAACTG...   \n",
       "11  Seq2  S_00001  CGGCGCCGAAAGATTTTTGGGAATTTTGGAAATGCTATTTTGAAGA...   \n",
       "12  Seq2  S_00002  TTCGCCATCATAATACGCGTTATAGCCATGATATTTCACAATCGCC...   \n",
       "13  Seq2  S_00003  CGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTT...   \n",
       "14  Seq2  S_00004  TCATCCACTTCATCAAAATAATAAATCACTTCAAACGCGCGCACCG...   \n",
       "15  Seq2  S_00005  TTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGC...   \n",
       "16  Seq2  S_00006  GGGTTTTCATCGCCTGCACTTCGGTTTCCGGATAGCCGTTGCGGCG...   \n",
       "17  Seq2  S_00007  TGCAGGCGATGAAAACCCTGTATTTTTTTGATATTCTGAACTTTGC...   \n",
       "18  Seq2  S_00008  GCCTTTAAAATGCTGTTTAAAAAACGCAAATTTTTCGGTCTGAAAC...   \n",
       "19  Seq2  S_00009  TTAAACAGCATTTTAAAGGCCTCGAGCACCACCACCACCACCACTG...   \n",
       "\n",
       "    oligo_length   oligo_tm primer_overlap_with_previous  overlap_tm_5prime  \\\n",
       "0            119  74.669187                         None                NaN   \n",
       "1            128  82.149159           CGGTGGTGGATCCGTTTG          56.343909   \n",
       "2            129  85.231564           GATGGCCAGCGCCATTAT          55.811457   \n",
       "3            128  85.057453          TGATGAATATACCTTTCGC          47.775584   \n",
       "4            128  85.703740           CCCCGGCGTATCCGGGCC          65.660649   \n",
       "5            128  82.795446           GATGCGCATGATCCGTTT          54.144649   \n",
       "6            128  85.057453           TGCGGCGGATTATCCGTA          56.451446   \n",
       "7            128  83.764878           GCTATGCGGAACCGGCGT          61.126936   \n",
       "8             78  81.265500           CAGCCGTTTGATGGCTTT          54.166830   \n",
       "9             73  80.600609           GCGCGCCCGGCGGTGGAA          69.523553   \n",
       "10           119  71.540938                         None                NaN   \n",
       "11           129  79.780705           CGGCGCCGAAAGATTTTT          55.017416   \n",
       "12           130  81.266757         AGTGTATCGCTTTACCTATA          48.313698   \n",
       "13           128  80.856584          CGCGTATTATGATGGCGAA          53.868963   \n",
       "14           130  79.675896           ATCCGCATGGCCATGCGT          60.798906   \n",
       "15           131  81.151404        TTATTTTGATGAAGTGGATGA          48.512171   \n",
       "16           130  80.312240       GCGGATTATCCGTATTTTTATA          49.832919   \n",
       "17           128  80.210296           TGCAGGCGATGAAAACCC          56.217177   \n",
       "18           129  77.856873           ACCGTGTTTGCGGGCTAT          57.547578   \n",
       "19            66  71.478719         TTAAACAGCATTTTAAAGGC          48.268215   \n",
       "\n",
       "    overlap_length  overlap_homodimer_tm  overlap_hairpin_tm  \\\n",
       "0              NaN                   NaN                 NaN   \n",
       "1             18.0              7.388261           41.578087   \n",
       "2             18.0             17.815718           68.998303   \n",
       "3             19.0            -64.316560            0.000000   \n",
       "4             18.0             30.462576           75.699579   \n",
       "5             18.0             30.763247            0.000000   \n",
       "6             18.0             30.590793           59.177368   \n",
       "7             18.0             15.010537           40.188310   \n",
       "8             18.0             10.864156           58.661393   \n",
       "9             18.0             31.975298           63.355769   \n",
       "10             NaN                   NaN                 NaN   \n",
       "11            18.0             44.859279           42.830305   \n",
       "12            20.0            -41.223309            0.000000   \n",
       "13            19.0             10.800809           49.069751   \n",
       "14            18.0             57.093991           67.117879   \n",
       "15            21.0           -138.759018            0.000000   \n",
       "16            22.0             30.590793           59.177368   \n",
       "17            18.0            -38.140845            0.000000   \n",
       "18            18.0              9.356132           54.474060   \n",
       "19            20.0              6.193641           29.811277   \n",
       "\n",
       "                                    original_sequence  \n",
       "0   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "1   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "2   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "3   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "4   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "5   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "6   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "7   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "8   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "9   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "10  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "11  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "12  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "13  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "14  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "15  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "16  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "17  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "18  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "19  cccctctagaaataattttgtttaactttaagaaggagatatacat...  "
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['id'] = ['Seq1', 'Seq2']\n",
    "min_gc = 0.25\n",
    "max_gc = 0.65\n",
    "min_tm = 10\n",
    "max_tm = 100\n",
    "min_segment_length = 40\n",
    "max_segment_length = 100\n",
    "max_length = 500\n",
    "\n",
    "df['seqs'] = ['MRTSLLVAALGLALAAALPGGAPLAQPDPEATMDRSLLQRQDLPYRFSAVDLDSVDGQRHYRLWLGRPLQAPPAAGYPVVWMLDGNAAVGALDESTLRRLADGDAPLLVAIGYRTPLRIDRAGRTFDYTPASPGQADQRDPLNGLPSGGADAFLDLLRDGMRPAVAAQAPLDTARQTLWGHSYGGLLVLHALFTRPGEFARYAAASPSLWWRDGAILGERAGLEQRLRGKRAELLLWRGSAEPASPRGSLKAEPGQAMARLVDDLRRVAGLTLDFQPLDGLGHGETLGASLRLLLARPAVERQR', \n",
    "             'MQLFDLPLDQLQTYKPEKTAPKDFSEFWKLSLEELAKVQAEPDLQPVDYPADGVKVYRLTYKSFGNARITGWYAVPDKEGPHPAIVKYHGYNASYDGEIHEMVNWALHGYATFGMLVRGQQSSEDTSISPHGHALGWMTKGILDKDTYYYRGVYLDAVRALEVISSFDEVDETRIGVTGGSQGGGLTIAAAALSDIPKAAVADYPYLSNFERAIDVALEQPYLEINSFFRRNGSPETEVQAMKTLSYFDIMNLADRVKVPVLMSIGLIDKVTPPSTVFAAYNHLETKKELKVYRYFGHEYIPAFQTEKLAFFKQHLKG']\n",
    "oligo_df = get_oligos(df, 'seqs', 'id', '.', 'cccctctagaaataattttgtttaactttaagaaggagatatacat', 'CTCGAGCACCACCACCACCACCACTGAgatccggctgctaacaaag', \n",
    "                     min_gc=min_gc, max_gc=max_gc, min_tm=min_tm, max_tm=max_tm, min_segment_length=min_segment_length, max_segment_length=max_segment_length)\n",
    "oligo_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "b6a9107d-3199-476c-aa3d-b803c7b7fc9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "objective:   0%|                                                  | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-912](e_coli...]\n",
      "location:   0%|                                                                         | 0/238 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                          | 0/238 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-912](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-912](mini:0.25, maxi:0.75, window:50)\n",
      "      │ Passed !\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:         0\n",
      "✔        0 ┍ MaximizeCAI[0-912](e_coli) \n",
      "           │ Codon opt. on window 0-912 scored -0.00E+00\n",
      "\n",
      "\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m              Warning: Seq1 does not end with a stop codon. \tCGC\t               \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[93m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[93mWarning: Seq1 does not end with a stop codon. AND you don't have a stop codon in your primer!!\tCTCGAGCACCACCACCACCACCACTGAgatccggctgctaacaaag\t\u001b[0m\n",
      "\u001b[93m--------------------------------------------------------------------------------\u001b[0m\n",
      "We expect the primer to be in 5 to 3 prime direction.\n",
      "ATGCGCACCTATTTTTTTGTGGCGGCGTTTGGCTTTGCGTTTGCGGCGGCGTTTCCGGGCGGCGCGCCGTTTGCGCAGCATGATCCGGAAGCGACCATGGATCGCTATTTTTTTCAGCGCCAGGATTTTCCGTATCGCTTTTATGCGGTGGATTTTGATTATGTGGATGGCCAGCGCCATTATCGCTTTTGGTTTGGCCGCCCGTTTCAGGCGCCGCCGGCGGCGGGCTATCCGGTGGATTGGATGTTTGATGGCAACGCGGCGGTGGGCGCGTTTGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTTTTTGTGGCGATTGGCTATCGCACCCCGTTTCGCATTGATCGCGCGGGCCGCACCTTTGATTATACCCCGGCGTATCCGGGCCAGGCGGATCAGCGCGATCCGTTTAACGGCTTTCCGTATGGCGGCGCGGATGCGTTTTTTGATTTTTTTCGCGATGGCATGCGCCCGGCGGTGGCGGCGCAGGCGCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCCATTATTATGGCGGCTTTTTTGTGTTTCATGCGTTTTTTACCCGCCCGGGCGAATTTGCGCGCTATGCGGCGGCGTATCCGTATTTTTGGTGGCGCGATGGCGCGATTTTTGGCGAACGCGCGGGCTTTGAACAGCGCTTTCGCGGCAAACGCGCGGAATTTTTTTTTTGGCGCGGCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACCGGGCCAGGCGATGGCGCGCTTTGTGGATGATTTTCGCCGCGTGGCGGGCTTTACCTTTGATTTTCAGCCGTTTGATGGCTTTGGCCATGGCGAAACCTTTGGCGCGTATTTTCGCTTTTTTTTTGCGCGCCCGGCGGTGGAACGCCAGCGC\n",
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n",
      "constraint:   0%|                                                 | 0/2 [00:00<?, ?it/s, now=AvoidPattern[0-954](patte...]\n",
      "location:   0%|                                                                           | 0/8 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                           | 0/8 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "objective:   0%|                                                  | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-954](e_coli...]\u001b[A\n",
      "location:   0%|                                                                         | 0/194 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "location:   0%|                                                                          | 0/194 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\n",
      "location:   0%|                                                                          | 0/194 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-954](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-954](mini:0.25, maxi:0.75, window:50)\n",
      "      │ Passed !\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:         0\n",
      "✔        0 ┍ MaximizeCAI[0-954](e_coli) \n",
      "           │ Codon opt. on window 0-954 scored -0.00E+00\n",
      "\n",
      "\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m              Warning: Seq2 does not end with a stop codon. \tGGC\t               \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[93m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[93mWarning: Seq2 does not end with a stop codon. AND you don't have a stop codon in your primer!!\tCTCGAGCACCACCACCACCACCACTGAgatccggctgctaacaaag\t\u001b[0m\n",
      "\u001b[93m--------------------------------------------------------------------------------\u001b[0m\n",
      "We expect the primer to be in 5 to 3 prime direction.\n",
      "ATGCAGTTTTTTGATTTTCCGTTTGATCCGCTGCAGACCTATAAACCGGAAGAAACCGCGCCGAAAGAATTTTATGAATTTGGCAAATGGTATTGGGAAGAATTTGCGAAAGTGCAGGCGGAACCGGATTTTCAGCCGGTGGATTATCCGGCGGATGGCGTGAAAGTGTATCGCGTGACCTATAAATATTTTGGCAACGCGCGCACCACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTTCATGGCTATGCGACCTTTGGCATGTTTGTGCGCGGCCAGCAGTATTATGAAGATACCTATATTTATCCGCATGGCCATGCGTTTGGCTGGATGACCAAAGGCATTTTTGGCAAAGATACCTATTATTATCGCGGCGTGTATTTTGATGCGGTGCGCGCGTTTGAAGTGATTTATTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCTATCAGGGCGGCGGCTTTACCATTGCGGCGGCGGCGTTTTATGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATTTTTATAACTTTGAACGCGCGATTGATGCGGCGTTTGAACAGCCGTATTTTGAAATTAACTATTTTGTGCGCCGCAACGGCTATCCGGAAAACGAAGTGCAGGCGATGCAGACCTTTTATTATTTTGAAATTATGAACTTTGCGGATTGGGTGGAAGTGCCGGTGTTTATGTGGATTGGCTTTATTGATAAAGTGACCCCGCCGTGGATTGTGTTTGCGGCGTATAACCATTTTGAAACCAGCAAAGGCTTTATGGTGTATCGCTATTATGGCCATGAATATATTCCGGATTTTCAGACCGATAAATTTCCGTTTTTTAAACCGCATTTTAAAGGC\n",
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>oligo_id</th>\n",
       "      <th>oligo_sequence</th>\n",
       "      <th>oligo_length</th>\n",
       "      <th>oligo_tm</th>\n",
       "      <th>primer_overlap_with_previous</th>\n",
       "      <th>overlap_tm_5prime</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>overlap_homodimer_tm</th>\n",
       "      <th>overlap_hairpin_tm</th>\n",
       "      <th>original_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00000</td>\n",
       "      <td>CAAACGGCGCGCCGCCCGGAAACGCCGCCGCAAACGCAAAGCCAAA...</td>\n",
       "      <td>119</td>\n",
       "      <td>76.754686</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00001</td>\n",
       "      <td>CGGGCGGCGCGCCGTTTGCGCAGCATGATCCGGAAGCGACCATGGA...</td>\n",
       "      <td>128</td>\n",
       "      <td>83.441734</td>\n",
       "      <td>CGGGCGGCGCGCCGTTTG</td>\n",
       "      <td>68.658228</td>\n",
       "      <td>18.0</td>\n",
       "      <td>56.011620</td>\n",
       "      <td>68.396815</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00002</td>\n",
       "      <td>GCGAAAGGTATATTCATCAAACGCGCCCACCGCCGCGTTGCCATCA...</td>\n",
       "      <td>129</td>\n",
       "      <td>86.834758</td>\n",
       "      <td>GATGGCCAGCGCCATTAT</td>\n",
       "      <td>55.811457</td>\n",
       "      <td>18.0</td>\n",
       "      <td>17.815718</td>\n",
       "      <td>68.998303</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00003</td>\n",
       "      <td>TGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.380596</td>\n",
       "      <td>TGATGAATATACCTTTCGC</td>\n",
       "      <td>47.775584</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-64.316560</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00004</td>\n",
       "      <td>AAACGGCGCCTGCGCCGCCACCGCCGGGCGCATGCCATCGCGAAAA...</td>\n",
       "      <td>128</td>\n",
       "      <td>88.288890</td>\n",
       "      <td>CCCCGGCGTATCCGGGCC</td>\n",
       "      <td>65.660649</td>\n",
       "      <td>18.0</td>\n",
       "      <td>30.462576</td>\n",
       "      <td>75.699579</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00005</td>\n",
       "      <td>GCGGCGCAGGCGCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCC...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.057453</td>\n",
       "      <td>GCGGCGCAGGCGCCGTTT</td>\n",
       "      <td>68.094164</td>\n",
       "      <td>18.0</td>\n",
       "      <td>53.361828</td>\n",
       "      <td>83.411766</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00006</td>\n",
       "      <td>ACGCCGGTTCCGCATAGCCGCGCCAAAAAAAAAATTCCGCGCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>86.026884</td>\n",
       "      <td>TGCGGCGGCGTATCCGTA</td>\n",
       "      <td>62.137239</td>\n",
       "      <td>18.0</td>\n",
       "      <td>32.349272</td>\n",
       "      <td>58.411563</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00007</td>\n",
       "      <td>GCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACC...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.057453</td>\n",
       "      <td>GCTATGCGGAACCGGCGT</td>\n",
       "      <td>61.126936</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.010537</td>\n",
       "      <td>40.188310</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00008</td>\n",
       "      <td>TTCCACCGCCGGGCGCGCAAAAAAAAAGCGAAAATACGCGCCAAAG...</td>\n",
       "      <td>78</td>\n",
       "      <td>81.265500</td>\n",
       "      <td>CAGCCGTTTGATGGCTTT</td>\n",
       "      <td>54.166830</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10.864156</td>\n",
       "      <td>58.661393</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00009</td>\n",
       "      <td>GCGCGCCCGGCGGTGGAACGCCAGCGCCTCGAGCACCACCACCACC...</td>\n",
       "      <td>73</td>\n",
       "      <td>80.600609</td>\n",
       "      <td>GCGCGCCCGGCGGTGGAA</td>\n",
       "      <td>69.523553</td>\n",
       "      <td>18.0</td>\n",
       "      <td>31.975298</td>\n",
       "      <td>63.355769</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00000</td>\n",
       "      <td>TAAAATTCTTTCGGCGCGGTTTCTTCCGGTTTATAGGTCTGCAGCG...</td>\n",
       "      <td>120</td>\n",
       "      <td>71.840745</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00001</td>\n",
       "      <td>CCGCGCCGAAAGAATTTTATGAATTTGGCAAATGGTATTGGGAAGA...</td>\n",
       "      <td>129</td>\n",
       "      <td>81.704538</td>\n",
       "      <td>CCGCGCCGAAAGAATTTTA</td>\n",
       "      <td>54.887527</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-2.987647</td>\n",
       "      <td>39.271088</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00002</td>\n",
       "      <td>TTCGCCATCATAATACGCGTTATAGCCATGATATTTCACAATCGCC...</td>\n",
       "      <td>129</td>\n",
       "      <td>82.666454</td>\n",
       "      <td>GTGTATCGCGTGACCTATA</td>\n",
       "      <td>51.807772</td>\n",
       "      <td>19.0</td>\n",
       "      <td>5.768584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00003</td>\n",
       "      <td>CGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>80.856584</td>\n",
       "      <td>CGCGTATTATGATGGCGAA</td>\n",
       "      <td>53.868963</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10.800809</td>\n",
       "      <td>49.069751</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00004</td>\n",
       "      <td>TCATCCACTTCATCAAAATAATAAATCACTTCAAACGCGCGCACCG...</td>\n",
       "      <td>130</td>\n",
       "      <td>79.675896</td>\n",
       "      <td>ATCCGCATGGCCATGCGT</td>\n",
       "      <td>60.798906</td>\n",
       "      <td>18.0</td>\n",
       "      <td>57.093991</td>\n",
       "      <td>67.117879</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00005</td>\n",
       "      <td>TTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGC...</td>\n",
       "      <td>131</td>\n",
       "      <td>83.045865</td>\n",
       "      <td>TTATTTTGATGAAGTGGATGA</td>\n",
       "      <td>48.512171</td>\n",
       "      <td>21.0</td>\n",
       "      <td>-138.759018</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00006</td>\n",
       "      <td>AGGTCTGCATCGCCTGCACTTCGTTTTCCGGATAGCCGTTGCGGCG...</td>\n",
       "      <td>130</td>\n",
       "      <td>80.630413</td>\n",
       "      <td>GCGGATTATCCGTATTTTTATA</td>\n",
       "      <td>49.832919</td>\n",
       "      <td>22.0</td>\n",
       "      <td>30.590793</td>\n",
       "      <td>59.177368</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00007</td>\n",
       "      <td>TGCAGGCGATGCAGACCTTTTATTATTTTGAAATTATGAACTTTGC...</td>\n",
       "      <td>129</td>\n",
       "      <td>79.780705</td>\n",
       "      <td>TGCAGGCGATGCAGACCT</td>\n",
       "      <td>59.627099</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10.850945</td>\n",
       "      <td>59.622681</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00008</td>\n",
       "      <td>GCCTTTAAAATGCGGTTTAAAAAACGGAAATTTATCGGTCTGAAAA...</td>\n",
       "      <td>129</td>\n",
       "      <td>76.894956</td>\n",
       "      <td>ATTGTGTTTGCGGCGTATA</td>\n",
       "      <td>53.466680</td>\n",
       "      <td>19.0</td>\n",
       "      <td>9.367922</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00009</td>\n",
       "      <td>TAAACCGCATTTTAAAGGCCTCGAGCACCACCACCACCACCACTGA...</td>\n",
       "      <td>65</td>\n",
       "      <td>72.196960</td>\n",
       "      <td>TAAACCGCATTTTAAAGGC</td>\n",
       "      <td>49.857060</td>\n",
       "      <td>19.0</td>\n",
       "      <td>4.109652</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id oligo_id                                     oligo_sequence  \\\n",
       "0   Seq1  S_00000  CAAACGGCGCGCCGCCCGGAAACGCCGCCGCAAACGCAAAGCCAAA...   \n",
       "1   Seq1  S_00001  CGGGCGGCGCGCCGTTTGCGCAGCATGATCCGGAAGCGACCATGGA...   \n",
       "2   Seq1  S_00002  GCGAAAGGTATATTCATCAAACGCGCCCACCGCCGCGTTGCCATCA...   \n",
       "3   Seq1  S_00003  TGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTT...   \n",
       "4   Seq1  S_00004  AAACGGCGCCTGCGCCGCCACCGCCGGGCGCATGCCATCGCGAAAA...   \n",
       "5   Seq1  S_00005  GCGGCGCAGGCGCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCC...   \n",
       "6   Seq1  S_00006  ACGCCGGTTCCGCATAGCCGCGCCAAAAAAAAAATTCCGCGCGTTT...   \n",
       "7   Seq1  S_00007  GCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACC...   \n",
       "8   Seq1  S_00008  TTCCACCGCCGGGCGCGCAAAAAAAAAGCGAAAATACGCGCCAAAG...   \n",
       "9   Seq1  S_00009  GCGCGCCCGGCGGTGGAACGCCAGCGCCTCGAGCACCACCACCACC...   \n",
       "10  Seq2  S_00000  TAAAATTCTTTCGGCGCGGTTTCTTCCGGTTTATAGGTCTGCAGCG...   \n",
       "11  Seq2  S_00001  CCGCGCCGAAAGAATTTTATGAATTTGGCAAATGGTATTGGGAAGA...   \n",
       "12  Seq2  S_00002  TTCGCCATCATAATACGCGTTATAGCCATGATATTTCACAATCGCC...   \n",
       "13  Seq2  S_00003  CGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTT...   \n",
       "14  Seq2  S_00004  TCATCCACTTCATCAAAATAATAAATCACTTCAAACGCGCGCACCG...   \n",
       "15  Seq2  S_00005  TTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGC...   \n",
       "16  Seq2  S_00006  AGGTCTGCATCGCCTGCACTTCGTTTTCCGGATAGCCGTTGCGGCG...   \n",
       "17  Seq2  S_00007  TGCAGGCGATGCAGACCTTTTATTATTTTGAAATTATGAACTTTGC...   \n",
       "18  Seq2  S_00008  GCCTTTAAAATGCGGTTTAAAAAACGGAAATTTATCGGTCTGAAAA...   \n",
       "19  Seq2  S_00009  TAAACCGCATTTTAAAGGCCTCGAGCACCACCACCACCACCACTGA...   \n",
       "\n",
       "    oligo_length   oligo_tm primer_overlap_with_previous  overlap_tm_5prime  \\\n",
       "0            119  76.754686                         None                NaN   \n",
       "1            128  83.441734           CGGGCGGCGCGCCGTTTG          68.658228   \n",
       "2            129  86.834758           GATGGCCAGCGCCATTAT          55.811457   \n",
       "3            128  85.380596          TGATGAATATACCTTTCGC          47.775584   \n",
       "4            128  88.288890           CCCCGGCGTATCCGGGCC          65.660649   \n",
       "5            128  85.057453           GCGGCGCAGGCGCCGTTT          68.094164   \n",
       "6            128  86.026884           TGCGGCGGCGTATCCGTA          62.137239   \n",
       "7            128  85.057453           GCTATGCGGAACCGGCGT          61.126936   \n",
       "8             78  81.265500           CAGCCGTTTGATGGCTTT          54.166830   \n",
       "9             73  80.600609           GCGCGCCCGGCGGTGGAA          69.523553   \n",
       "10           120  71.840745                         None                NaN   \n",
       "11           129  81.704538          CCGCGCCGAAAGAATTTTA          54.887527   \n",
       "12           129  82.666454          GTGTATCGCGTGACCTATA          51.807772   \n",
       "13           128  80.856584          CGCGTATTATGATGGCGAA          53.868963   \n",
       "14           130  79.675896           ATCCGCATGGCCATGCGT          60.798906   \n",
       "15           131  83.045865        TTATTTTGATGAAGTGGATGA          48.512171   \n",
       "16           130  80.630413       GCGGATTATCCGTATTTTTATA          49.832919   \n",
       "17           129  79.780705           TGCAGGCGATGCAGACCT          59.627099   \n",
       "18           129  76.894956          ATTGTGTTTGCGGCGTATA          53.466680   \n",
       "19            65  72.196960          TAAACCGCATTTTAAAGGC          49.857060   \n",
       "\n",
       "    overlap_length  overlap_homodimer_tm  overlap_hairpin_tm  \\\n",
       "0              NaN                   NaN                 NaN   \n",
       "1             18.0             56.011620           68.396815   \n",
       "2             18.0             17.815718           68.998303   \n",
       "3             19.0            -64.316560            0.000000   \n",
       "4             18.0             30.462576           75.699579   \n",
       "5             18.0             53.361828           83.411766   \n",
       "6             18.0             32.349272           58.411563   \n",
       "7             18.0             15.010537           40.188310   \n",
       "8             18.0             10.864156           58.661393   \n",
       "9             18.0             31.975298           63.355769   \n",
       "10             NaN                   NaN                 NaN   \n",
       "11            19.0             -2.987647           39.271088   \n",
       "12            19.0              5.768584            0.000000   \n",
       "13            19.0             10.800809           49.069751   \n",
       "14            18.0             57.093991           67.117879   \n",
       "15            21.0           -138.759018            0.000000   \n",
       "16            22.0             30.590793           59.177368   \n",
       "17            18.0             10.850945           59.622681   \n",
       "18            19.0              9.367922            0.000000   \n",
       "19            19.0              4.109652            0.000000   \n",
       "\n",
       "                                    original_sequence  \n",
       "0   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "1   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "2   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "3   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "4   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "5   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "6   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "7   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "8   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "9   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "10  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "11  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "12  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "13  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "14  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "15  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "16  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "17  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "18  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "19  cccctctagaaataattttgtttaactttaagaaggagatatacat...  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import random\n",
    "import numpy \n",
    "SEED = 128\n",
    "random.seed(SEED)\n",
    "numpy.random.seed(SEED)  \n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['id'] = ['Seq1', 'Seq2']\n",
    "min_gc = 0.25\n",
    "max_gc = 0.75\n",
    "min_tm = 10\n",
    "max_tm = 100\n",
    "min_segment_length = 40\n",
    "max_segment_length = 100\n",
    "max_length = 500\n",
    "\n",
    "df['seqs'] = ['MRTSLLVAALGLALAAALPGGAPLAQPDPEATMDRSLLQRQDLPYRFSAVDLDSVDGQRHYRLWLGRPLQAPPAAGYPVVWMLDGNAAVGALDESTLRRLADGDAPLLVAIGYRTPLRIDRAGRTFDYTPASPGQADQRDPLNGLPSGGADAFLDLLRDGMRPAVAAQAPLDTARQTLWGHSYGGLLVLHALFTRPGEFARYAAASPSLWWRDGAILGERAGLEQRLRGKRAELLLWRGSAEPASPRGSLKAEPGQAMARLVDDLRRVAGLTLDFQPLDGLGHGETLGASLRLLLARPAVERQR', \n",
    "             'MQLFDLPLDQLQTYKPEKTAPKDFSEFWKLSLEELAKVQAEPDLQPVDYPADGVKVYRLTYKSFGNARITGWYAVPDKEGPHPAIVKYHGYNASYDGEIHEMVNWALHGYATFGMLVRGQQSSEDTSISPHGHALGWMTKGILDKDTYYYRGVYLDAVRALEVISSFDEVDETRIGVTGGSQGGGLTIAAAALSDIPKAAVADYPYLSNFERAIDVALEQPYLEINSFFRRNGSPETEVQAMKTLSYFDIMNLADRVKVPVLMSIGLIDKVTPPSTVFAAYNHLETKKELKVYRYFGHEYIPAFQTEKLAFFKQHLKG']\n",
    "oligo_df = get_oligos(df, 'seqs', 'id', '.', 'cccctctagaaataattttgtttaactttaagaaggagatatacat', 'CTCGAGCACCACCACCACCACCACTGAgatccggctgctaacaaag', \n",
    "                     min_gc=min_gc, max_gc=max_gc, min_tm=min_tm, max_tm=max_tm, min_segment_length=min_segment_length, max_segment_length=max_segment_length)\n",
    "oligo_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
