{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "04e33c47-4362-4ef3-9168-fb9bba96524c",
   "metadata": {},
   "source": [
    "# DNA weaver\n",
    "\n",
    "#### \n",
    "https://github.com/Edinburgh-Genome-Foundry/DnaWeaver/blob/master/examples/10kb_from_vendor_or_oligos.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a3400d1e-d6ed-42d7-870c-99487f0bb26d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dnaweaver in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (0.3.8)\n",
      "Requirement already satisfied: numpy in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (2.1.0)\n",
      "Requirement already satisfied: Biopython in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (1.84)\n",
      "Requirement already satisfied: proglog in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (0.1.10)\n",
      "Requirement already satisfied: networkx in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (3.3)\n",
      "Requirement already satisfied: flametree in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (0.1.12)\n",
      "Requirement already satisfied: dna-features-viewer in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (3.1.3)\n",
      "Requirement already satisfied: weasyprint in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (63.0)\n",
      "Requirement already satisfied: pandas in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (2.2.2)\n",
      "Requirement already satisfied: jinja2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnaweaver) (3.1.4)\n",
      "Requirement already satisfied: matplotlib>=3 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dna-features-viewer->dnaweaver) (3.9.2)\n",
      "Requirement already satisfied: packaging in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dna-features-viewer->dnaweaver) (24.1)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from jinja2->dnaweaver) (2.1.5)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnaweaver) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnaweaver) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnaweaver) (2024.1)\n",
      "Requirement already satisfied: tqdm in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from proglog->dnaweaver) (4.67.1)\n",
      "Requirement already satisfied: pydyf<0.12,>=0.11.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (0.11.0)\n",
      "Requirement already satisfied: cffi<2,>=0.6 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (1.17.0)\n",
      "Requirement already satisfied: tinyhtml5<3,>=2.0.0b1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (2.0.0)\n",
      "Requirement already satisfied: tinycss2<2,>=1.4.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (1.4.0)\n",
      "Requirement already satisfied: cssselect2<0.8,>=0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (0.7.0)\n",
      "Requirement already satisfied: Pyphen<0.16,>=0.9.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (0.15.0)\n",
      "Requirement already satisfied: Pillow<11,>=9.1.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->dnaweaver) (10.4.0)\n",
      "Requirement already satisfied: fonttools<5,>=4.0.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->dnaweaver) (4.53.1)\n",
      "Requirement already satisfied: pycparser in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cffi<2,>=0.6->weasyprint->dnaweaver) (2.22)\n",
      "Requirement already satisfied: webencodings in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cssselect2<0.8,>=0.1->weasyprint->dnaweaver) (0.5.1)\n",
      "Requirement already satisfied: zopfli>=0.1.4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->dnaweaver) (0.2.3.post1)\n",
      "Requirement already satisfied: brotli>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->dnaweaver) (1.1.0)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib>=3->dna-features-viewer->dnaweaver) (1.3.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib>=3->dna-features-viewer->dnaweaver) (0.12.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib>=3->dna-features-viewer->dnaweaver) (1.4.5)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib>=3->dna-features-viewer->dnaweaver) (3.1.4)\n",
      "Requirement already satisfied: six>=1.5 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from python-dateutil>=2.8.2->pandas->dnaweaver) (1.16.0)\n"
     ]
    }
   ],
   "source": [
    "! pip install dnaweaver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "aff1bbdb-e28c-4861-b244-d1399ca74cd2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                                                                                                                       "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ordering plan (Gibson Assembly Station)::\n",
      "  0-100: From CheapDNA. - price 11.10 - lead_time 0.0\n",
      "  100-210: From Oligo Assembly Station - price 40.43 - lead_time 8.0\n",
      "  210-320: From CheapDNA. - price 13.10 - lead_time 0.0\n",
      "  320-430: From CheapDNA. - price 13.10 - lead_time 0.0\n",
      "  430-519: From CheapDNA. - price 9.90 - lead_time 0.0\n",
      "Price:87.63, total lead_time:8.0\n",
      "Finished in 0 seconds\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    }
   ],
   "source": [
    "import dnaweaver as dw\n",
    "import time\n",
    "\n",
    "cheap_dna_offer = dw.CommercialDnaOffer(\n",
    "    name=\"CheapDNA.\",\n",
    "    sequence_constraints=[\n",
    "        dw.NoPatternConstraint(enzyme=\"BsaI\"),\n",
    "        dw.SequenceLengthConstraint(max_length=4000),\n",
    "        dw.GcContentConstraint(min_gc=0.3, max_gc=0.7)\n",
    "    ],\n",
    "    pricing=dw.PerBasepairPricing(0.10),\n",
    ")\n",
    "\n",
    "oligo_dna_offer = dw.CommercialDnaOffer(\n",
    "    name=\"OliGoo\",\n",
    "    sequence_constraints=[\n",
    "        dw.GcContentConstraint(min_gc=0.3, max_gc=0.7),\n",
    "        dw.SequenceLengthConstraint(max_length=100),\n",
    "    ],\n",
    "    pricing=dw.PerBasepairPricing(0.07),\n",
    "    memoize=True\n",
    ")\n",
    "\n",
    "oligo_assembly_station = dw.DnaAssemblyStation(\n",
    "    name=\"Oligo Assembly Station\",\n",
    "    assembly_method=dw.OligoAssemblyMethod(\n",
    "        overhang_selector=dw.TmSegmentSelector(\n",
    "            min_size=15, max_size=25, min_tm=50, max_tm=70\n",
    "        ),\n",
    "        min_segment_length=40,\n",
    "        max_segment_length=200,\n",
    "        sequence_constraints=[dw.SequenceLengthConstraint(max_length=1500)],\n",
    "        duration=8,\n",
    "        cost=30,\n",
    "    ),\n",
    "    supplier=oligo_dna_offer,\n",
    "    coarse_grain=20,\n",
    "    a_star_factor=\"auto\",\n",
    "    memoize=True,\n",
    ")\n",
    "\n",
    "assembly_station = dw.DnaAssemblyStation(\n",
    "    name=\"Gibson Assembly Station\",\n",
    "    assembly_method=dw.GibsonAssemblyMethod(\n",
    "        overhang_selector=dw.TmSegmentSelector(min_tm=55, max_tm=70),\n",
    "        min_segment_length=80,\n",
    "        max_segment_length=120,\n",
    "    ),\n",
    "    supplier=[cheap_dna_offer, oligo_assembly_station],\n",
    "    logger=\"bar\",\n",
    "    coarse_grain=100,\n",
    "    fine_grain=10,\n",
    "    a_star_factor=\"auto\",\n",
    ")\n",
    "print(\"Looking for the best assembly plan...\")\n",
    "t0 = time.time()\n",
    "sequence = \"ATGACGACGCCGTGGTCGGGAGAGTTAATTGTGGATTTATTAAATCAGCGTCCTCGCGGACAAGTCGTGTTAGATGCTGCTAATGGGGTGGATTATGAGCTGCGCTCTCAAGAGACCGGACCAATTGACATCGACGCACACGGACCCGCTCGCTGCCTGACCGGAACGGGCTCCCGTACGACGCGTTGGATTCGCCAACATCAGCGCAATACTCGCGTTTTGGGCCTGTCGCATGaccctacagttgctgacgcgcgcgaattgttacgtttaggggctttggGCCTTTTACCTAAAAATGCCGAGCCAGCGGAAATCTTCCAAGCGTTAGATCAAGCTGCCGCCGACCACATTCACACGAATGACCTTGTGCGCGACGCACCCCGTACCCCCGCTCTTCCTGAAGAAGAGGAGGCAGAACGCGTCAACGCTTTGCCAGACCAGGGCGCGGCGCGCACGGGCGCTTTGGGAGCCCAAGGTTGGCCCCACCATACTCCTGACCAACCGGCGCGCTTAGGT\"\n",
    "quote = assembly_station.get_quote(sequence, with_assembly_plan=True)\n",
    "print(quote.assembly_step_summary())\n",
    "print(\"Finished in %.01d seconds\" % (time.time() - t0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "db55993c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 108 (l) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 112 (p) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 108 (l) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 112 (p) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n"
     ]
    }
   ],
   "source": [
    "assembly_plan_report = quote.to_assembly_plan_report()\n",
    "assembly_plan_report.write_full_report(\"report.zip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d35ff231-d452-4934-b3b3-74d1224c0a65",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(assembly_plan_report.plan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "872d7316-952c-4a5a-a370-d04e68a37de9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATGACGACGCCGTGGTCGGGAGAGTTAATTGTGGATTTATTAAATCAGCGTCCTCGCGGACAAGTCGTGTTAGATGCTGCTAATGGGGTGGATTATGAGCTGCGCTCTCAAGAGACCGGACCAATTGACATCGACGCACACGGACCCGCTCGCTGCCTGACCGGAACGGGCTCCCGTACGACGCGTTGGATTCGCCAACATCAGCGCAATACTCGCGTTTTGGGCCTGTCGCATGaccctacagttgctgacgcgcgcgaattgttacgtttaggggctttggGCCTTTTACCTAAAAATGCCGAGCCAGCGGAAATCTTCCAAGCGTTAGATCAAGCTGCCGCCGACCACATTCACACGAATGACCTTGTGCGCGACGCACCCCGTACCCCCGCTCTTCCTGAAGAAGAGGAGGCAGAACGCGTCAACGCTTTGCCAGACCAGGGCGCGGCGCGCACGGGCGCTTTGGGAGCCCAAGGTTGGCCCCACCATACTCCTGACCAACCGGCGCGCTTAGGT\n"
     ]
    }
   ],
   "source": [
    "print(assembly_plan_report.plan.sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "22337005-113e-489f-91fd-5ec2774d97be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S_00000 ATGACGACGCCGTGGTCGGGAGAGTTAATTGTGGATTTATTAAATCAGCGTCCTCGCGGACAAGTCGTGTTAGATGCTGCTAATGGGGTGGATTATGAGCTGCGCTCTCAA\n",
      "{'via': ['Gibson Assembly Station comparator']}\n",
      "S_00003 GATTATGAGCTGCGCTCTCAAGAGACCGGACCAATTGACATCGACGCACACGGACCCGCTCGCTGCCTGACCGGAACGGGCTCCCGTACGACGCGTTGGATTCGCCAACATCAGCGCAATACTCGCGTTTT\n",
      "{'via': ['Gibson Assembly Station comparator', 'Gibson Assembly Station comparator']}\n",
      "S_00004 TCAGCGCAATACTCGCGTTTTGGGCCTGTCGCATGaccctacagttgctgacgcgcgcgaattgttacgtttaggggctttggGCCTTTTACCTAAAAATGCCGAGCCAGCGGAAATCTTCCAAGCGTTAG\n",
      "{'via': ['Gibson Assembly Station comparator']}\n",
      "S_00005 GCGGAAATCTTCCAAGCGTTAGATCAAGCTGCCGCCGACCACATTCACACGAATGACCTTGTGCGCGACGCACCCCGTACCCCCGCTCTTCCTGAAGAAGAGGAGGCAGAACGCGTCAACGCTTTGCCAGA\n",
      "{'via': ['Gibson Assembly Station comparator']}\n",
      "S_00006 CGCGTCAACGCTTTGCCAGACCAGGGCGCGGCGCGCACGGGCGCTTTGGGAGCCCAAGGTTGGCCCCACCATACTCCTGACCAACCGGCGCGCTTAGGT\n",
      "{'via': ['Gibson Assembly Station comparator']}\n"
     ]
    }
   ],
   "source": [
    "for oligo in assembly_plan_report.plan.assembly_plan:\n",
    "    if oligo.accepted:\n",
    "        print(oligo.id, oligo.sequence)\n",
    "        print(oligo.metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15f5b1c4-3645-45bc-b096-5796a68c41b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Other installs\n",
    "\n",
    "brew install glib\n",
    " brew install blast\n",
    "brew install python pango libffi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "79a04bd9-c8ea-4b5f-a6cb-30d11558395a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pdf_reports in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (0.3.7)\n",
      "Requirement already satisfied: pypugjs in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (5.11.0)\n",
      "Requirement already satisfied: jinja2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (3.1.4)\n",
      "Requirement already satisfied: weasyprint in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (63.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (4.12.3)\n",
      "Requirement already satisfied: pandas in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (2.2.2)\n",
      "Requirement already satisfied: Markdown in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (3.7)\n",
      "Requirement already satisfied: backports.functools-lru-cache in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf_reports) (2.0.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from beautifulsoup4->pdf_reports) (2.6)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from jinja2->pdf_reports) (2.1.5)\n",
      "Requirement already satisfied: numpy>=1.22.4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->pdf_reports) (2.1.0)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->pdf_reports) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->pdf_reports) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->pdf_reports) (2024.1)\n",
      "Requirement already satisfied: six in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pypugjs->pdf_reports) (1.16.0)\n",
      "Requirement already satisfied: charset-normalizer in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pypugjs->pdf_reports) (3.3.2)\n",
      "Requirement already satisfied: pydyf<0.12,>=0.11.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (0.11.0)\n",
      "Requirement already satisfied: cffi<2,>=0.6 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (1.17.0)\n",
      "Requirement already satisfied: tinyhtml5<3,>=2.0.0b1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (2.0.0)\n",
      "Requirement already satisfied: tinycss2<2,>=1.4.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (1.4.0)\n",
      "Requirement already satisfied: cssselect2<0.8,>=0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (0.7.0)\n",
      "Requirement already satisfied: Pyphen<0.16,>=0.9.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (0.15.0)\n",
      "Requirement already satisfied: Pillow<11,>=9.1.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf_reports) (10.4.0)\n",
      "Requirement already satisfied: fonttools<5,>=4.0.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf_reports) (4.53.1)\n",
      "Requirement already satisfied: pycparser in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cffi<2,>=0.6->weasyprint->pdf_reports) (2.22)\n",
      "Requirement already satisfied: webencodings in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cssselect2<0.8,>=0.1->weasyprint->pdf_reports) (0.5.1)\n",
      "Requirement already satisfied: zopfli>=0.1.4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf_reports) (0.2.3.post1)\n",
      "Requirement already satisfied: brotli>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf_reports) (1.1.0)\n"
     ]
    }
   ],
   "source": [
    "!  pip install pdf_reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9b8b2ebf-1650-4970-80e8-824e9025f6d2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: dnachisel in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (3.2.12)\n",
      "Requirement already satisfied: numpy in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (2.1.0)\n",
      "Requirement already satisfied: Biopython in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (1.84)\n",
      "Requirement already satisfied: proglog in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (0.1.10)\n",
      "Requirement already satisfied: docopt in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (0.6.2)\n",
      "Requirement already satisfied: flametree in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (0.1.12)\n",
      "Requirement already satisfied: python-codon-tables in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel) (0.1.12)\n",
      "Requirement already satisfied: tqdm in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from proglog->dnachisel) (4.67.1)\n",
      "Requirement already satisfied: dnachisel[reports] in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (3.2.12)\n",
      "Requirement already satisfied: numpy in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (2.1.0)\n",
      "Requirement already satisfied: Biopython in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (1.84)\n",
      "Requirement already satisfied: proglog in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (0.1.10)\n",
      "Requirement already satisfied: docopt in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (0.6.2)\n",
      "Requirement already satisfied: flametree in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (0.1.12)\n",
      "Requirement already satisfied: python-codon-tables in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (0.1.12)\n",
      "Collecting pdf-reports (from dnachisel[reports])\n",
      "  Downloading pdf_reports-0.3.7-py3-none-any.whl.metadata (2.1 kB)\n",
      "Collecting sequenticon (from dnachisel[reports])\n",
      "  Downloading sequenticon-0.1.5-py3-none-any.whl.metadata (2.0 kB)\n",
      "Requirement already satisfied: matplotlib in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (3.9.2)\n",
      "Requirement already satisfied: dna-features-viewer in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (3.1.3)\n",
      "Requirement already satisfied: pandas in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dnachisel[reports]) (2.2.2)\n",
      "Requirement already satisfied: packaging in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from dna-features-viewer->dnachisel[reports]) (24.1)\n",
      "Requirement already satisfied: contourpy>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (1.3.0)\n",
      "Requirement already satisfied: cycler>=0.10 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (0.12.1)\n",
      "Requirement already satisfied: fonttools>=4.22.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (4.53.1)\n",
      "Requirement already satisfied: kiwisolver>=1.3.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (1.4.5)\n",
      "Requirement already satisfied: pillow>=8 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (10.4.0)\n",
      "Requirement already satisfied: pyparsing>=2.3.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (3.1.4)\n",
      "Requirement already satisfied: python-dateutil>=2.7 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from matplotlib->dnachisel[reports]) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnachisel[reports]) (2024.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pandas->dnachisel[reports]) (2024.1)\n",
      "Collecting pypugjs (from pdf-reports->dnachisel[reports])\n",
      "  Downloading pypugjs-5.11.0-py2.py3-none-any.whl.metadata (7.4 kB)\n",
      "Requirement already satisfied: jinja2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf-reports->dnachisel[reports]) (3.1.4)\n",
      "Requirement already satisfied: weasyprint in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf-reports->dnachisel[reports]) (63.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pdf-reports->dnachisel[reports]) (4.12.3)\n",
      "Collecting Markdown (from pdf-reports->dnachisel[reports])\n",
      "  Using cached Markdown-3.7-py3-none-any.whl.metadata (7.0 kB)\n",
      "Collecting backports.functools-lru-cache (from pdf-reports->dnachisel[reports])\n",
      "  Downloading backports.functools_lru_cache-2.0.0-py2.py3-none-any.whl.metadata (3.5 kB)\n",
      "Requirement already satisfied: tqdm in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from proglog->dnachisel[reports]) (4.67.1)\n",
      "Collecting pydenticon (from sequenticon->dnachisel[reports])\n",
      "  Downloading pydenticon-0.3.1.tar.gz (26 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting snapgene-reader (from sequenticon->dnachisel[reports])\n",
      "  Downloading snapgene_reader-0.1.21-py3-none-any.whl.metadata (1.6 kB)\n",
      "Requirement already satisfied: six>=1.5 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from python-dateutil>=2.7->matplotlib->dnachisel[reports]) (1.16.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from beautifulsoup4->pdf-reports->dnachisel[reports]) (2.6)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from jinja2->pdf-reports->dnachisel[reports]) (2.1.5)\n",
      "Requirement already satisfied: charset-normalizer in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from pypugjs->pdf-reports->dnachisel[reports]) (3.3.2)\n",
      "Collecting xmltodict (from snapgene-reader->sequenticon->dnachisel[reports])\n",
      "  Using cached xmltodict-0.14.2-py2.py3-none-any.whl.metadata (8.0 kB)\n",
      "Collecting html2text (from snapgene-reader->sequenticon->dnachisel[reports])\n",
      "  Downloading html2text-2024.2.26.tar.gz (56 kB)\n",
      "  Preparing metadata (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25hRequirement already satisfied: pydyf<0.12,>=0.11.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (0.11.0)\n",
      "Requirement already satisfied: cffi<2,>=0.6 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (1.17.0)\n",
      "Requirement already satisfied: tinyhtml5<3,>=2.0.0b1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (2.0.0)\n",
      "Requirement already satisfied: tinycss2<2,>=1.4.0 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (1.4.0)\n",
      "Requirement already satisfied: cssselect2<0.8,>=0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (0.7.0)\n",
      "Requirement already satisfied: Pyphen<0.16,>=0.9.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from weasyprint->pdf-reports->dnachisel[reports]) (0.15.0)\n",
      "Requirement already satisfied: pycparser in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cffi<2,>=0.6->weasyprint->pdf-reports->dnachisel[reports]) (2.22)\n",
      "Requirement already satisfied: webencodings in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from cssselect2<0.8,>=0.1->weasyprint->pdf-reports->dnachisel[reports]) (0.5.1)\n",
      "Requirement already satisfied: zopfli>=0.1.4 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf-reports->dnachisel[reports]) (0.2.3.post1)\n",
      "Requirement already satisfied: brotli>=1.0.1 in /Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages (from fonttools[woff]<5,>=4.0.0->weasyprint->pdf-reports->dnachisel[reports]) (1.1.0)\n",
      "Downloading pdf_reports-0.3.7-py3-none-any.whl (118 kB)\n",
      "Downloading sequenticon-0.1.5-py3-none-any.whl (13 kB)\n",
      "Downloading backports.functools_lru_cache-2.0.0-py2.py3-none-any.whl (6.7 kB)\n",
      "Using cached Markdown-3.7-py3-none-any.whl (106 kB)\n",
      "Downloading pypugjs-5.11.0-py2.py3-none-any.whl (36 kB)\n",
      "Downloading snapgene_reader-0.1.21-py3-none-any.whl (7.1 kB)\n",
      "Using cached xmltodict-0.14.2-py2.py3-none-any.whl (10.0 kB)\n",
      "Building wheels for collected packages: pydenticon, html2text\n",
      "  Building wheel for pydenticon (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for pydenticon: filename=pydenticon-0.3.1-py3-none-any.whl size=6515 sha256=49fced90976c18391018235a763034fd30175a74b14f315a67ac26e2a82deb87\n",
      "  Stored in directory: /Users/arianemora/Library/Caches/pip/wheels/6f/20/6d/926175bfcc1c4e585db37504fde7cdbb7c741a97b38c5c294b\n",
      "  Building wheel for html2text (setup.py) ... \u001b[?25ldone\n",
      "\u001b[?25h  Created wheel for html2text: filename=html2text-2024.2.26-py3-none-any.whl size=33110 sha256=0e7026d49ea1191420012ca83d5e909a2bdee4edd826f8a1b0c6ac2c1336d8d8\n",
      "  Stored in directory: /Users/arianemora/Library/Caches/pip/wheels/f3/96/6d/a7eba8f80d31cbd188a2787b81514d82fc5ae6943c44777659\n",
      "Successfully built pydenticon html2text\n",
      "Installing collected packages: xmltodict, pypugjs, pydenticon, Markdown, html2text, backports.functools-lru-cache, snapgene-reader, pdf-reports, sequenticon\n",
      "Successfully installed Markdown-3.7 backports.functools-lru-cache-2.0.0 html2text-2024.2.26 pdf-reports-0.3.7 pydenticon-0.3.1 pypugjs-5.11.0 sequenticon-0.1.5 snapgene-reader-0.1.21 xmltodict-0.14.2\n"
     ]
    }
   ],
   "source": [
    "! pip install dnachisel\n",
    "! pip install 'dnachisel[reports]'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e5702e46-03b8-424a-bb34-b7a8421f1135",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dnachisel import *\n",
    "\n",
    "# DEFINE THE OPTIMIZATION PROBLEM\n",
    "seq = 'MASPQLQMALDAFKTMGEKMAQAGNDVKALRAVMEEMSGFPSAGETKCTPVNAGGVPAEWISGPGAAD'\n",
    "seq = reverse_translate(seq)\n",
    "problem = DnaOptimizationProblem(\n",
    "    sequence=seq,\n",
    "    constraints=[\n",
    "        AvoidPattern(\"BsaI_site\"),\n",
    "        EnforceGCContent(mini=0.3, maxi=0.7, window=50),\n",
    "    ],\n",
    "    objectives=[CodonOptimize(species='e_coli', location=(500, 1400))]\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2daeb8c5-1b13-46c0-87ba-dd5cbe76ab1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Solving constraints\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "constraint:   0%|                                                                                                        | 0/2 [00:00<?, ?it/s, now=AvoidPattern[0-10000](pat...]\n",
      "location:   0%|                                                                                                                                  | 0/3 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                                                                          | 0/3 [00:00<?, ?it/s, now=4574-4580(-)]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                          | 0/4 [00:00<?, ?it/s, now=4574-4580(-)]\u001b[A\n",
      "location:   0%|                                                                                                                             | 0/4 [00:00<?, ?it/s, now=2043-2099]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                             | 0/4 [00:00<?, ?it/s, now=2043-2099]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now optimizing the sequence\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "objective:   0%|                                                                                                         | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[500-1400](e_c...]\n",
      "location:   0%|                                                                                                                           | 0/200 [00:00<?, ?it/s, now=2043-2099]\u001b[A\n",
      "location:   0%|                                                                                                                             | 0/200 [00:00<?, ?it/s, now=500-503]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                             | 0/200 [00:00<?, ?it/s, now=500-503]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Success! Generating report.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(True, 'Optimization successful.', None)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "problem.optimize_with_report(target=\"dna_chisel_report.zip\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5c504165-0201-4517-9d6e-ac5f15e6004c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "constraint:   0%|                                                                                                        | 0/2 [00:00<?, ?it/s, now=AvoidPattern[0-10000](pat...]\n",
      "location:   0%|                                                                                                                                  | 0/7 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                                                                          | 0/7 [00:00<?, ?it/s, now=1114-1120(+)]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                          | 0/1 [00:00<?, ?it/s, now=1114-1120(+)]\u001b[A\n",
      "location:   0%|                                                                                                                             | 0/1 [00:00<?, ?it/s, now=1746-1797]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                             | 0/1 [00:00<?, ?it/s, now=1746-1797]\u001b[A\n",
      "objective:   0%|                                                                                                         | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[500-1400](e_c...]\u001b[A\n",
      "location:   0%|                                                                                                                           | 0/201 [00:00<?, ?it/s, now=1746-1797]\u001b[A\n",
      "location:   0%|                                                                                                                             | 0/201 [00:00<?, ?it/s, now=515-518]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                                             | 0/201 [00:00<?, ?it/s, now=515-518]\u001b[A\n",
      "                                                                                                                                                                                 \u001b[A"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-10000](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-10000](mini:0.30, maxi:0.70, window:50)\n",
      "      │ Passed !\n",
      "✔PASS ┍ EnforceTranslation[500-1400(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:     -4.70\n",
      "     -4.70 ┍ MaximizeCAI[500-1400](e_coli) \n",
      "           │ Codon opt. on window 500-1400 scored -4.70E+00\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    }
   ],
   "source": [
    "# SOLVE THE CONSTRAINTS, OPTIMIZE WITH RESPECT TO THE OBJECTIVE\n",
    "\n",
    "problem.resolve_constraints()\n",
    "problem.optimize()\n",
    "\n",
    "# PRINT SUMMARIES TO CHECK THAT CONSTRAINTS PASS\n",
    "\n",
    "print(problem.constraints_text_summary())\n",
    "print(problem.objectives_text_summary())\n",
    "\n",
    "# GET THE FINAL SEQUENCE (AS STRING OR ANNOTATED BIOPYTHON RECORDS)\n",
    "\n",
    "final_sequence = problem.sequence  # string\n",
    "final_record = problem.to_record(with_sequence_edits=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "417d5ed3-6d41-4e9f-a6d3-ece6385170a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTACGATCCATCGCCCCTGTTAACATTACGGAATTCAACAGGTTCTCCGGGCCGAACAAAAATGCGGAATGAGGAATAGATAGGCTATGCACGGCCTCCTAGAGCTTCTTGGTTCTTCCTCCATCCTACCCTAGCTGTTGCACGTATAGACATCCGCGATAGTCGGCACGAGCTTACCTAGTGATGGACATTCCTCACAGCATACGCTAGAAATCAGACTCATTCATTACGTTCGATCACCTACTCTGATCGAAGTGAATGTGGGAGGAGGTCACGAATGCCCCCAATATCCTACCGCATAGAGATAGGTTGAGATACGAACATACCTGGCGAGTAATGCGATAGTTGAAGGCAGAGAGGACCGCACACCGCCGTCGCGGTGTGGGTGCTGAAACATCCGTCAGGGTGGACCTTCGGAAGAAACTTACGTTCTGCCTGTTGCGTCAGTCGTCGGCAAGTCGTTTAGTACATAGGGACACGGCAAGCCTACGCCTCTCCAGGATTTTGGCATGCATCGCGTGGGCCCGAAACGCATTACCACCGTGCTGTCTCGCAGCGCACCGTGCACTCTGGGCGAAAAACAGGCGCTGCTGACCAAATGGCGCTGCCAGCTGAGCCATGCGCGTCGCCATCCGCGTTTTAGCACCGGCAGCGGCCGTCGCAAATAACTGTATGGCCCGAAAGGCCGCCGCCAGCATAGCCAGGTGGGCACCGATTGCTGCTAAATTTAATTTCCGGTGTTTGCGCTGGATGTGATGGGCCTGCCGGCGTGGGAAAACAACGGCACCAGCAGCGTGTATAGCGTGCGCCTGCATGCGCTGTAACCGAGCCATACCGATAGCTTTGCGCGCCTGCGCATTGAACTGCAGGTGACCGGCGTGACCACCGGCTATGGCCGCTATGCGAGCACCAGCCTGCTGACCCTGACCTATCTGGGCGAAGCGATTGAAGTGGTGCTGCTGATGAAAGAAGGCAAAAGCATTGATGAACATTGCGCGATTCATCAGAGCGAACAGCTGAAAGCGGCGCGCGATTTTAGCCCGGCGGCGCGCACCTAACGCATTCTGCTGCTGGTGCGCCATTATACCCCGCCGGGCCAACGTCGCGTTGGCGAAGTGAGCATTCCGGGCAAAACCTAAAGCCGCGATTGCCTGGCGCTGATTATTGCGCCGCTGGGCGTGGTGGAAACCATTACCCCGCGCCTGGAATGCTTTGTGCGCAGCTAAGGCCCGGTGGTGAACGTGCGCTAAGCGTGGAGCGGCCCGTTGATTAGCACCGTGCATACCATTCTGATGAACTAAGTGAACTATGAACTGAAAAGCACCCATAAATAAACCAAAGGCGTGCAGAACAGCGTGGCGTTTTAAAGCCATATTACCCGCGTGTATGGCCTGAGCATTGTACGATATACCGGGCATCGCGTCTACAGAAAACGGATTTACAGGCGCGGAGGTCCTCAGCTACCCCGCGAAGCAACGAAGCGAATGTGGTGTACCCGAGGTGGTCTTTATATTGCTTCAGATTTCCTGGTTCCTTGCTCTAGGGAGCATCTGTATGTGATCCGTCAGCCTCCACACTCTTCCCGGGCATATAGAACACAGATAACCCAAGTCCGCCAGAGTAGGGAGCTACCTACCTAGCACAGACGGTCAGAGGTTCGGACTACAGGCGGGAGGTATCGCTACAACGGTTTGGCCTAGCTGACCGTTTCCGATCAAGAGGCGTGAGTTAGATACGCGTACTGTTTGGGCAGGGGGGGTACGGTCTTCGGCCGCGAATCATCTGCTGGCCACTGCGAATTGCGACTCCCATGCCCGGAAATGTATCTAAGTGGCGTTTAGCGCGGTATCTAAAGGCCAGTCAGCCGCTGGCTAAAAGAGGGTTCGCTCAACCATAGAGAAATCAGGCCTTCACCAAACTTGCAGAGACAGGACAGCTCGATTTTTTTTGGTGTCTGGGCATTGGTTTGTCGCAGACGGTACCGATGCTCAATTGGACCGAGCGCCAGCTGCACGATGTAACTATGTGCTCCAGACGTCTCATTACGCCCGTACTGTTCAGCATATTAGGCTAAATCACCCTGGAGGGAATATTCGTTACACGGCCGCGCGATAGAGCAGACGGCTAAAACTCCGGGCTCTGGGACTGTGACAGGTGGGCCGCGTAGATCAACACTGTAAGGGCGGTGCGTCTTCGGGCATCAGGAGGAACGTAGTCGCTGCACCAGACTTCGACTCCACTCCATTCCAAAGTAGCTTTAGTAAATTAATTAGAATAGGGCGGGAGTCTTACACTTGACTTTGAAGTTACCGATGGCAAAGCTATTCATTGCGGTTTCGCCAGTGGCATACCTGCCATAGCAATTTTAACCATAGATACTAAATCGTGCTAGATTGTTAGCAAGATCTGCGTCGGAATGCCTGCGTGTAAGGTTAGCCTATGCGTACCTCTTTGATCCCAGGGTCACGTTTCCGATGGCCGGATAATACCCCTAACGACAGACTCGCAAACGGCGCGCGAAGTATCTGACGAAGCCTGAGGAGTCACATGTGCCGTACGTATAATATTGGAGACATCGTGAGGCGAACTTGAGAATTGATCGCTGGGTATCCGTACATCCTAAACAAATGTCGGAATTGCCATATGCTCTAGGGTTATGACAACGGCGTTGCAGCCCGACTCGTAGCGCAGATGTAAAAAAGGGAATTTGCGTGATACCGTCTACACCGTGGCTTCACGAGGATTCTAAAGATCTACGTTTATGTAGTGTCTATGGCAGTGAGTGAGGTCCATGCTCACCGTCCCACCTTACACAACCCCTGAAATGTGTACGACTCCGCTCTAAAAGGGGTCGACAGCCAGATATCCAACGTTTCCGCGTGATTAAAGTGGATGACAGTTTTTACAATTAAACGAGAGTTTCGGACGCCTTGGGCGTGTCGCATCTGTCGAAGTCCTGATTTATTTTGGATTCGCGCCCTCGCTCCCTTCCCAGTTTTTCCGAAGAATACTAGTTCTGAGATTCTTATCCGGCGCCTAACGCCATCTCACTATGCTACCATTGCCCGTAGTAGGACTATGACCTTACCACAACGATATCAGAAATACGACGGATTATCAAACGAGTCAGCCGCTTGGAGCGGATATATCGATCAATGGCACCGTTCGGGCTTTTGACGACTTCTGGAACACTCTGCACTGGGGAGCCTAACTGAATTCCAGGTCGAGTATCGTTCCTCCAGAGGAGCCAGGCCCGTAGTTACATTCATTGGTCGTACCTGAGGCCGTACTGACACGAGGAGTAACAGGTCTTGCGATAGACCCATTAGGAATGGCCATAATTCACCGCTACTGACCCTGCTTTTCGATAGGATGTCTTCGGACTAGCACAGTCCGTTGTGTACTCGACTAATGCATGGAGGAGCTCGCATATCCCTGTCCGATGAGGTCTTCACATCGCATCGTATTACGATAAAGAGTAGTCTTTGACCGCCATGTAGTGGATCCCTGCAGTAGGCCCACGTGGGTTGGGCTAGCGTTCATTCGTATCTGGATGGTAATGGATTATCGAATTTATGATTAAGTTTGGCTGCTGGAAATGCCGTAGTCGTAGGCATTTCATCTTCATGCGAGGCGTCCCGAGCAAAGGAAAGTCCGGTGTCGGGGTCAACTGACTCAGAGGGTTCTAGCGGTGGGTATGCAGTAGAACCGACTTGGGCTACATACATGCTGCGTACAAACGGTGCGCACGCAGCACCGCGGTAGAGGTAGTCCCGTATGCTGCACATCTGGTTAAACGCATTCCTTCCAGATCGTTGGTACTCTTTAGTGACGGACTACTATCGCCTTTGGAGAACCGCTATATAGACTTGAACGCCTGTTGTTGTTGAGGAGAGTTTCGCATTTGCGAATAACCCCTGTTAGTATCGAGCGTACAGGTGCCACTGACGCTATCAAGAGCCAAGGAACGACCCAGTTTGCTCCAGGGGAGCAAACATCGCGAGAGTTGGCGGCGTTGCCTATAGCCACGGGGCACCGTTGATCAACGTCCTTCCGGATCTCTGAAAACATGGAATTCAGGAGAACGGCCGTCCATCACCCACTTGCGCGTTAGTGGCTATACAGTTTAGCGCGGGGACGTTTACGATGTACTATTTGACCAGCAGAAACTCTGAAGCCCATCAAACGCAGGTGGGCAGGCTCTAACGACCCTGACAGTACTCCCAATATGGATGACGGCTGTCCCTGTCCTGGGCTCTCGGTCCTCAAACACTGACGAAGTATCTGGGTACTAGTGGATGTGACAATACTTATGGGCGTAGCGTTTCTGGTCAAAAGCCTTACTACGCGTGATAGGCACGACCCACAGGAACCGCAGGCATCGTATGTTAGTAAGATAGAAACGCGGGACGGGCGCTTTGCATACCCACTTATGCAAAACCCTATGTTTGCTCGAAGCTCTCGGATTAGCGCGATATCGATGTCATATGTACTAAGACTCCTGGCTTTCCGCAACCGTTCCGCCCTTACCAAATACTGGTATTCCACTTGCCCTCTAGGAAAGATCTGATTTGGGCGCGGGTCAGATGGGCTGAAATAGCGACATTACAAAAGCCCACAGCCGAATACGGACTTACTTAGCTGGCAACCTCTGGGCTCGGGGGCGAGCGAGGGGCATAGTAACCCAAGAAACTGCAAAAGCTGCATTATTGACTAATTACAGACATATTAGATCCTTCTGGGAGGTCGCTGAGGAAAGGCATTCTAATCCTCGTAAAAGGTCGCTAGGTTTTTTGGGAACTAGAGAAAACAAGATGATTCTAGAACGGCTTGATGTACACAAGGTCGGGTGGTAACCGTTAGTTTAATGGGTAGTATAAACAACGTAGTTCGTGAGTGAAGACTTATTCCTCGTGGGTAACCAGGTTCGTCCATTTTTGTTGACTCAGCCCCATGTCCATACGACCTGGCCAGGGAGAACCGCGGTGCCAGCTACTTGTTGCAGTGGTTCCTCTCGATAAGAGTATTAGTTTCCTCCTCTCCGGGAGTATGAGGATGGAATTGAATTTCAGAGACGGCTCCAATCCTTTGGGTCTAGTGCTTTGCGCGCCTAATACGAGACGACCTCCGATCTGCTCAGTCGGGATTTGATATCTTAGTTTGCGCTAGGTATGTGTGACGGTTACACAATTAGGCAACGGAGCGGGGCACGCTAACCACACTAACTTTTGAAACCGTGGCTAAAAGCAAGAGGAATTACCATTAAGACCTCCCCCTCCGCAGTTTCAACTTGGGGGTTCATTGGAACAAGGATCTTGCGAAGCGTCAGTTGAGATATAATAGTCATGATTCTATTAGCGGAGGACTTGCGATGCCCGGTAATCACTACCCCAGCCTAAGGAAGATGGTCTAAAATTCAGCTAATCTCACATGGGCTACCTGCCCTACACAACCCTGACGCCGGGGAGTTTATACAAGCACGCCCATCGCTGTTGGTTGCCTCATGCCTCTTGTTGACCGGCACTCATGTTTCTTCGGTTTGGAGTAGGGGCCTTTGACACTCCTTAGGTTTATACCGTGAACCCTTTTGATATTGATAAGTGCTTGCAATGCCTGAATAATCATCCACAGTTTAGATATCAAAATCTCAACACAAGACTTCACTGTTGGAAGGGACTGTGGGCGTGCATCACAAGTAATGTGAAGTCTTCTGGTAGAGCAGCAGACCCTGTCGAAGCTTCGCGGCAAGTAGCCCGCATAGACCAGTGGACCTCCGTCCTCTATCAGGGTATGCGGTGTCCTGTGTATATTGAAGGGCACTCGCAGATTGTATGCGGCCGGGTCCCCTCTGTGAGTGTGCAGACGCCTACGGCGATGCACGAAGAAAAGCTTAGTGGATTGCAGTCCCTGCGTTTTGGTAGACTATGATCGCACACAAAGGATGCAGTAAGTCTGATATTGTCGAGCATGGATTGATTGCTACGTTATGTAATAACGGGTTGTTTTAATCGATACGGCCTCCAATGGGTGTTAAACCCCCCTCCTGTGTTGCTAAGTCTGGGCGATGTTCTAGTCTGCGGGCTTGCAACCGAGTTGAAATGTATATCAACGTGTAAGCTAGCGCGACCTATCCTACTCCCCCTGTAAATCGATTGACGTTATAATGCGGGTCACCTGGCGATCCAGGAGGTTCGTCTGCCCGGAAGAAGGATTTGTACAATGCGGTAAATGGCGTAAGCATGCCGTACGAGAGGTACGATACCCTTGAGAGGTGGGAGTAATGTGCTATGCTGGCTTCGCTTGAACAGCTGCTAGTTGCTACAATCCTAGTGACACCTATCGCGATGCCCGCTACGAGTACCTCCCTTAGTCTTCCCCGGTTCAGTTCAGACGTAAGCTTCTTGGCATCTCCAATTGCTGATATTCACATGACTAACAGTTGCGGGAGAAATTGGTTGCCGATGAAGTGCCCACCAAGTGGCCCGGAACAAACTAAAGTCTACTGCTTATGGGGCCCCCTGTGCTCTACCGGGTTTGTAGGTAAGTTCGTTGACATATCCAGCGATATGATTGGTATCGTAACAACTCGCTCGGCACCCAGCATCGCTACTGTGCTAACTTCGATGGAGCGCTGTAACGCGCCTTTTTATATGACCTGGCATTGATCTTTTACTCTTGAATATGACCACAGGACTTCACCAGAGAGATTCAGTGACGACAAAAGTGGCGGTTTTTTAAACTCGCGAATGACTCAGTATTAGAGTGTCCAGAAAGCCCTTGCTCTATATCCCGTTAGTCGTTCTACAATGAAGAGTAATAGGCATGTCGCATTTCCTGCCGACACTTTGACGACGCATTTATACGAAACACGTACCGAAAACTGCAAAAAGGTAACTGAACGTTTCCAAGGTAAGCAAGATTAAAACATAAAGATGCACGTAGAGAGGTCATCGAGGTGCCTGGCCGCAGTCAGCTGTCGGGCGCCTCATTTTTTATACCTTCCACTGCTCAAACTCCTCTGTGAAGCGTCTAAGATTTAGACAGCCTAAAGTGCTAATCACTGTAGTTTACAATGTTTAGTACACAAATCGTGGGATCGAGAGCCGATCGGTCTATTACGACAATCACTCAACGAAGATCGATATACCTAGTAGACTTGTCTGAGCACTTAGGGAGGATGTTTGACCGAGACAAGCGTTCGCCACCTTTCCATTCAATTTGCTGGTGGTTCGCTTGGATTATTGCTCTTCCCTCCTCCGTCGTACCAAGTTACTTTGTTCGCAGACATCAACCTAATCCCATCAGATTTTACTTGAGGCCCTGTCCGGTCAGCCAGCCGTAGTATCAAAGGACTCGAACGCATGTTGAAGTTGCCTCGAGCGCACCCTCCGCTTGCTCACGAACTCTGTTCTTGTACTTAGTTCTATGGGTAGGTTGTGCCTGTTATGAGCCTAAGAACCCAAGCGCCCCGGTGTCGGAAGACGAATAACTGAGCGAATACGTCCTACACAGTTAGGGAAACAACTATGCATATAATGACGCGGTGTGTACAGTGTGCGTGCAGTTTCTTCCAGTCTCTTTCGCTTCATCACGTACGTGTGCTTCAAGACCTAATTATAGTCGTTTGCGCACATCACCCCCCTCAAGAATCATGCGAATTGCATGAATGTGAAGGAGACAACCCCTCTAGGGTTGATCTATATGTGAGATTATAGCCGCCAAACGATTGTCTCAACACTACGGGTGCCAGAAGGGTAACACTCTCTTCTTCCGTCGACGTGCTCACACCTGAGTGCGGCACAGATATACTCATACATATCCTTCAATCGTGAGGTCTAGGGGCGAGCGCAGCGGACACCTCATCCAAGTGCAGATAGAAGCAGTCTTTCTACGGAAATTAGAATGCGCCGTCCAAATCTCCTCCTACGTTTTCCCCTCTACTCTTAGAGCGACGCAGGAGAAAGCTCAATGAATGTTAGCAAAGTTTCATTGGGTCAATCTCATATCATAAGAGAAAGTCTCGTAACCTCGACCCGTCTTGGAATGCATTCTTATCGGCAGCACCTAGTGTATCCAGCGGTTGGTTATACCAGACACACTAGCAAATGCAGTCACATGTAATCAGAGTAGGGTGAGAAATCACACTGCCTGCTAGAATGATCTTTCTTCCGGGCGTAGACTTGCAGAGCGTAGAATACATTATTACTCTCGTGGATAATCAGGATCACTACATTATCGGTACATGTCAATCTAGGCATCGATGGAGTGAAACCCTAGTCGGGATAGAGAATAGAAGGCTGTCACCTCATCTATCACGGTCTACGTACGAGACAGGGTCCCTCTGCATGCACAGAGCCAGTAATAGTACTCCGGGGAAGCCACCCCGTCCTGTGGGGTTATCTTTATCTCCCGTGCCAACTACAACTGCCTGGGGGCTCCGTGATACGCCATCGGTACTTAATAGGCGGCCAGTTACGAACATTTTAAAACCGCTAGGGGTACACCTTCAGCGCGGTAAAGACACCAGTATTCTGTTATTCACACTCAGAGTGCCCCGGAACTATTAGCGGGGCCCATGCGTCGGTTTGCACGCAGTCCGTCGCACTGTGTTCAACGGGGGCAGTTTACAACAGTTACTCTGCGTAGGGTAAGGCATAAACTCCCGTCCTGGCCATGTACGGGCTAAGATGGTCGCTTTTGGGATAATTATCCTCGTTCATGGAGTTAGATGGTCTACTATGCTTATGGGGCCTAGTATGTCTGTCCTAAGAATTGATTGATCTCGACCACATGGACATTACTATTAAATGAACATTCCCGTTGGACAGTCATTTGTGCTCGCAGTCCATACTATCCTAGAAGGGGGTAATAGCCACACTTTGTAAGGATAAATTGCGCACGGGGTGTCTCCATAGCCTAGTGGTGTCGATGAGGGTCTAAGTCCAAGTTTCCGTACCAGAGCCTGGATGGATATGCGGAATACCGGAGTTTGGCTTAGAGCTGTAGACCCTCACATCAGATTTCGAGCCTCGACTTGAGGTTCTTCCTGCATTCTAACGCTGTCCGCTTACCCTGATCAGTTAGCATGGCGGGGACGAAGTGAATACCCTGCCACACCTTGTTTCCTTTAGACACTATCGCTCGTACGCGACGCTAGACACAATGAACTACGCGGCGAAGGCTATATTATTATTACAAGCTGCGGTTGATGTAGTTCTATGGACTAACGGCATGACATGAAGAGGTTATTCACCAACCGAACGACATCCAGTACTAGTGACAGAGCAATTTCTTTCATGAGTGCCCCGTCCAGACAACGCTATAGACGGGGTGCTGGTCCTGAATATCATTACGAGCTCTTACGTAAAGCGCAACGAGAAGTATGCCTCGTGCTGTCTCCTCGTTGTACTAACATTATCTACGACATACCAAACTGTTCAAATAAATTCATGGTTGGAGGATCGGGCATTCATGTGTTATGCCTCCGTGAGGGTTCGAACTTAAGTTGGTCGTTAACCATTCTCGTGTCCGAGCATCATAATAGAATATGCCTCTTATACCAGTCAGGGATCCAACTTAGATGAACGACTCACACCCTCCAGACAAGTCTCACGAGTTGCTCACGCCCCGCACACGTTTTGACAACACGAGACATGGAGCTCCCAGCTTTGGACAAACTGATCGGTCCAAAGGCAGAATCAGACCATGACGGCACTTAAGTGAGGCTATTCGGGAACGGAACTAGATGGTGCTTCTACTTGCAGTCACGGTGTACGTGGAGAATCCCACCAATATCTATAGTAATCCACAAGCTTCTGTTCTAACCTAGATCCTCCATCAAGTTATTCCGAGGGATTGGGTAACCCTTCGGGTACAGTAACTGATGGTAGATCTCCATTCCGCGCAATTAGTGACGGCAGGAGAAATGTGGCTTCTTTTTCCATGTTAAAGCTGCAAGCCCAAGTGTACAAGTTGCTCGTATCCCCCCGATCTCTAATAA'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c45155cb-ff9e-41c6-a7b0-3ada05cfd936",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting PyGObject\n",
      "  Downloading pygobject-3.50.0.tar.gz (1.1 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.1/1.1 MB\u001b[0m \u001b[31m10.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25h  Installing build dependencies ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[79 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Collecting meson-python>=0.12.1\n",
      "  \u001b[31m   \u001b[0m   Downloading meson_python-0.17.1-py3-none-any.whl.metadata (4.1 kB)\n",
      "  \u001b[31m   \u001b[0m Collecting pycairo>=1.16\n",
      "  \u001b[31m   \u001b[0m   Downloading pycairo-1.27.0.tar.gz (661 kB)\n",
      "  \u001b[31m   \u001b[0m \u001b[?25l     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m0.0/661.5 kB\u001b[0m \u001b[31m?\u001b[0m eta \u001b[36m-:--:--\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m661.5/661.5 kB\u001b[0m \u001b[31m5.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \u001b[?25h  Installing build dependencies: started\n",
      "  \u001b[31m   \u001b[0m   Installing build dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: started\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Installing backend dependencies: started\n",
      "  \u001b[31m   \u001b[0m   Installing backend dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): finished with status 'error'\n",
      "  \u001b[31m   \u001b[0m   \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   \u001b[31m×\u001b[0m \u001b[32mPreparing metadata \u001b[0m\u001b[1;32m(\u001b[0m\u001b[32mpyproject.toml\u001b[0m\u001b[1;32m)\u001b[0m did not run successfully.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m╰─>\u001b[0m \u001b[31m[50 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[36m\u001b[1m+ meson setup /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39 /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39/.mesonpy-db1o3mrk -Dbuildtype=release -Db_ndebug=if-release -Db_vscrt=md -Dwheel=true -Dtests=false --native-file=/private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39/.mesonpy-db1o3mrk/meson-python-native-file.ini\u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m The Meson build system\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Version: 1.6.0\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Source dir: /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Build dir: /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39/.mesonpy-db1o3mrk\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Build type: native build\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Project name: pycairo\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Project version: 1.27.0\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m C compiler for the host machine: cc (clang 15.0.0 \"Apple clang version 15.0.0 (clang-1500.3.9.4)\")\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m C linker for the host machine: cc ld64 1053.12\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Host machine cpu family: aarch64\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Host machine cpu: aarch64\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Program python3 found: YES (/Users/arianemora/miniconda3/envs/oligo/bin/python3.10)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wall: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Warray-bounds: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wcast-align: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wconversion: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wextra: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wformat=2: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wformat-nonliteral: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wformat-security: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wimplicit-function-declaration: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Winit-self: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Winline: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wmissing-format-attribute: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wmissing-noreturn: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wnested-externs: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wold-style-definition: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wpacked: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wpointer-arith: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wreturn-type: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wshadow: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wsign-compare: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wstrict-aliasing: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wundef: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wunused-but-set-variable: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wswitch-default: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wno-missing-field-initializers: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -Wno-unused-parameter: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -fno-strict-aliasing: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Compiler for C supports arguments -fvisibility=hidden: YES\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Did not find pkg-config by name 'pkg-config'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Found pkg-config: NO\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Did not find CMake 'cmake'\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Found CMake: NO\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m Run-time dependency cairo found: NO (tried framework)\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m ../cairo/meson.build:31:12: ERROR: Dependency lookup for cairo with method 'pkgconfig' failed: Pkg-config for machine host machine not found. Giving up.\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m A full log can be found at /private/var/folders/sj/4wqsfdtd6093v9746b0t3mq40000gn/T/pip-install-v27hw42l/pycairo_598ca31f8cfd4a7a8b72b9c59e4cad39/.mesonpy-db1o3mrk/meson-logs/meson-log.txt\n",
      "  \u001b[31m   \u001b[0m   \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  \u001b[31m   \u001b[0m \u001b[1;31merror\u001b[0m: \u001b[1mmetadata-generation-failed\u001b[0m\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m \u001b[31m×\u001b[0m Encountered error while generating package metadata.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m╰─>\u001b[0m See above for output.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m \u001b[1;35mnote\u001b[0m: This is an issue with the package mentioned above, not pip.\n",
      "  \u001b[31m   \u001b[0m \u001b[1;36mhint\u001b[0m: See above for details.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "\u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[?25h"
     ]
    }
   ],
   "source": [
    "! pip install PyGObject"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "daf8e9c1-1816-4940-bf07-5508cb7fbd3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: Copy_of_Base_pET22b+_ba\n",
      "Name: Copy_of_Base_pET22b+_ba\n",
      "Number of features: 16\n",
      "/molecule_type=ds-DNA\n",
      "/topology=circular\n",
      "/data_file_division=   \n",
      "/date=18-JAN-2025\n",
      "Seq('ACGCGCCCTGTAGCGGCGCATTAAGCGCGGCGGGTGTGGTGGTTACGCGCAGCG...GGG')\n",
      "Updated GenBank file saved as modified_pet22b.gb\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import pandas as pd\n",
    "\n",
    "def insert_features_from_oligos(genbank_file, output_file, oligo_df):\n",
    "    \"\"\"\n",
    "    Insert features into a GenBank file based on oligo_df.\n",
    "\n",
    "    Args:\n",
    "        genbank_file (str): Path to the original GenBank file.\n",
    "        output_file (str): Path to save the updated GenBank file.\n",
    "        oligo_df (pd.DataFrame): DataFrame with oligo information. \n",
    "                                 Expected columns: ['id', 'oligo_id', 'oligo_sequence', 'oligo_length', ...].\n",
    "    \"\"\"\n",
    "    # Read the original GenBank file\n",
    "    record = SeqIO.read(genbank_file, \"genbank\")\n",
    "    print(record)\n",
    "    # Iterate through the oligo DataFrame to add features\n",
    "    for _, row in oligo_df.iterrows():\n",
    "        start = record.seq.find(row['oligo_sequence'].upper())\n",
    "        if start == -1:\n",
    "            print(f\"Warning: Oligo sequence {row['oligo_id']} not found in the GenBank sequence.\")\n",
    "            continue\n",
    "        \n",
    "        end = start + len(row['oligo_sequence'])\n",
    "        feature = SeqFeature(\n",
    "            location=FeatureLocation(start, end, strand=row['strand']),\n",
    "            type=\"misc_feature\",\n",
    "            qualifiers={\n",
    "                \"label\": f\"{row['oligo_id']} {'(reverse)' if row['strand'] == -1 else '(forward)'}\",\n",
    "                \"note\": f\"Length: {len(row['oligo_sequence'])}, TM: {row['oligo_tm']}\"\n",
    "            }\n",
    "        )\n",
    "        record.features.append(feature)\n",
    "    \n",
    "    # Save the updated GenBank file\n",
    "    SeqIO.write(record, output_file, \"genbank\")\n",
    "    print(f\"Updated GenBank file saved as {output_file}\")\n",
    "\n",
    "# Example usage\n",
    "if __name__ == \"__main__\":\n",
    "    # Input/output paths\n",
    "    genbank_file = \"copy-of-base-pet22b-base-anm.gb\"\n",
    "    output_file = \"modified_pet22b.gb\"\n",
    "    \n",
    "    # Mock oligo DataFrame (replace with real output from get_oligos)\n",
    "    oligo_data = [\n",
    "        {\"id\": \"seq1\", \"oligo_id\": \"oligo1\", \"strand\":-1, \"oligo_sequence\": \"tccggcggatggcgtgaaagattatcgcttttattatcagtatgtgggcaacgcgcgcatgaccc\", \"oligo_tm\": 65.5},\n",
    "        {\"id\": \"seq1\", \"oligo_id\": \"oligo2\", \"strand\":1, \"oligo_sequence\": \"attatgatggcgaac\",  \"oligo_tm\": 67.0},\n",
    "    ]\n",
    "    oligo_df = pd.DataFrame(oligo_data)\n",
    "    \n",
    "    # Insert features\n",
    "    insert_features_from_oligos(genbank_file, output_file, oligo_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ef38f9d-dc30-4799-9d9f-d6bc1c6b3aff",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "def insert_sequence_with_translation(input_file, output_file, insert_position, new_sequence, translation_label, reverse=False):\n",
    "    \"\"\"\n",
    "    Insert a new sequence at a specific position in a GenBank file, add a translation annotation,\n",
    "    and adjust existing annotations to avoid overlap.\n",
    "\n",
    "    Args:\n",
    "        input_file (str): Path to the original GenBank file.\n",
    "        output_file (str): Path to save the modified GenBank file.\n",
    "        insert_position (int): Position to insert the new sequence (0-based).\n",
    "        new_sequence (str): The DNA sequence to insert.\n",
    "        translation_label (str): Label for the translation feature.\n",
    "        reverse (bool): Whether the feature should be on the reverse strand.\n",
    "    \"\"\"\n",
    "    # Read the original GenBank file\n",
    "    record = SeqIO.read(input_file, \"genbank\")\n",
    "    \n",
    "    # Insert the new sequence at the specified position\n",
    "    if reverse:\n",
    "        new_sequence = str(Seq(new_sequence).reverse_complement())  # Reverse complement the sequence if needed\n",
    "    record.seq = record.seq[:insert_position] + Seq(new_sequence) + record.seq[insert_position:]\n",
    "    \n",
    "    # Adjust existing annotations to avoid overlap\n",
    "    inserted_length = len(new_sequence)\n",
    "    for feature in record.features:\n",
    "        if feature.location.start >= insert_position:\n",
    "            feature.location = FeatureLocation(\n",
    "                feature.location.start + inserted_length,\n",
    "                feature.location.end + inserted_length,\n",
    "                strand=feature.location.strand\n",
    "            )\n",
    "    \n",
    "    # Create the feature label\n",
    "    strand_label = \" (reverse)\" if reverse else \" (forward)\"\n",
    "    full_label = translation_label + strand_label\n",
    "    \n",
    "    # Add a feature for the inserted sequence\n",
    "    start = insert_position\n",
    "    end = insert_position + len(new_sequence)\n",
    "    feature = SeqFeature(\n",
    "        location=FeatureLocation(start, end, strand=-1 if reverse else 1),\n",
    "        type=\"CDS\",  # CDS type for coding sequences\n",
    "        qualifiers={\n",
    "            \"label\": full_label,\n",
    "            \"translation\": Seq(new_sequence).translate(table=11)  # Translation for the sequence\n",
    "        }\n",
    "    )\n",
    "    record.features.append(feature)\n",
    "    \n",
    "    # Save the modified GenBank file\n",
    "    SeqIO.write(record, output_file, \"genbank\")\n",
    "    print(f\"Modified GenBank file saved as {output_file}\")\n",
    "\n",
    "# Example usage\n",
    "if __name__ == \"__main__\":\n",
    "    input_genbank_file = \"base-pet22b-base-anm.gb\"\n",
    "    output_genbank_file = \"modified_with_translation.gb\"\n",
    "    \n",
    "    insert_position = 5193  # Position to insert the sequence (0-based)\n",
    "    new_sequence = \"ATGCGCACCTATTTTTTTGTGGCGGCGTTTGGCTTTGCGTTTGCGGCGGATTTTCCGGTGGTGGATCCGTTTGCGCAGCCGGATCATGAAGCGACCATGGATCGCTATTTTTTTCAGCGCCAGGATTTTCCGTATCGCTTTTATGCGGTGGATTTTGATTATGTGGATGGCCAGCGCCATTATCGCTTTTGGTTTGGCCGCCCGTTTCAGGCGCATCCGGCGGATGGCTATCCGGATGATTGGATGTTTGATGGCAACGCGGCGGTGGGCGCGTTTGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTTTTTGTGGCGATTGGCTATCGCACCCCGTTTCGCATTGATCGCGCGGGCCTGACCTTTGATTATACCCCGGCGTATCCGGGCCATGATGATCAGCGCGATCCGTTTAACGGCTTTCCGTATGGCGGCGCGGATGCGTTTTTTGATTTTTTTCGCGATGGCATGCGCCCGGCGGTGGATGCGCATGATCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCCATTATTATGGCGGCTTTTTTGTGTTTCATGCGTTTTTTACCCGCCCGGGCGAATTTGCGCGCTATGCGGCGGATTATCCGTATTTTTGGTGGCGCGATGGCGCGATTTTTGGCGAACGCGCGGGCTTTGAACAGCTGTTTCGCGGCAAACGCGCGGAATTTTTTTTTTGGCGCGGCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACCGGGCCATGCGATGGCGCGCTTTGTGGATGATTTTCGCCGCGATGATGGCTTTACCTTTGATTTTCAGCCGTTTGATGGCTTTGGCCATGGCGAAACCTTTGGCGCGTATTTTCGCTTTTTTTTTGCGCGCCCGGCGGTGGAACGCCAGCGC\"  # New DNA sequence to insert\n",
    "    translation_label = \"Insert_protein\"\n",
    "    reverse = False  # Set to True for reverse feature, False for forward\n",
    "\n",
    "    insert_sequence_with_translation(input_genbank_file, output_genbank_file, insert_position, new_sequence, translation_label, reverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "dd368813-dff0-4682-943c-1e7ec2d2a7c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dnaweaver as dw\n",
    "import time\n",
    "import dnachisel as dnachisel \n",
    "from sciutil import SciUtil\n",
    "from Bio.Seq import Seq\n",
    "from difflib import SequenceMatcher\n",
    "import primer3\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "import pandas as pd\n",
    "import math\n",
    "from dnachisel import *\n",
    "\n",
    "u = SciUtil()\n",
    "\n",
    "from primer3 import calcHairpin, calcHomodimer\n",
    "\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "\n",
    "def insert_sequence_with_translation(input_file, output_file, insert_position, new_sequence, translation_label, reverse=False):\n",
    "    \"\"\"\n",
    "    Insert a new sequence at a specific position in a GenBank file, add a translation annotation,\n",
    "    and adjust existing annotations to avoid overlap.\n",
    "\n",
    "    Args:\n",
    "        input_file (str): Path to the original GenBank file.\n",
    "        output_file (str): Path to save the modified GenBank file.\n",
    "        insert_position (int): Position to insert the new sequence (0-based).\n",
    "        new_sequence (str): The DNA sequence to insert.\n",
    "        translation_label (str): Label for the translation feature.\n",
    "        reverse (bool): Whether the feature should be on the reverse strand.\n",
    "    \"\"\"\n",
    "    # Read the original GenBank file\n",
    "    record = SeqIO.read(input_file, \"genbank\")\n",
    "    \n",
    "    # Insert the new sequence at the specified position\n",
    "    if reverse:\n",
    "        new_sequence = str(Seq(new_sequence).reverse_complement())  # Reverse complement the sequence if needed\n",
    "    record.seq = record.seq[:insert_position] + Seq(new_sequence) + record.seq[insert_position:]\n",
    "    \n",
    "    # Adjust existing annotations to avoid overlap\n",
    "    inserted_length = len(new_sequence)\n",
    "    for feature in record.features:\n",
    "        if feature.location.start >= insert_position:\n",
    "            feature.location = FeatureLocation(\n",
    "                feature.location.start + inserted_length,\n",
    "                feature.location.end + inserted_length,\n",
    "                strand=feature.location.strand\n",
    "            )\n",
    "    \n",
    "    # Create the feature label\n",
    "    strand_label = \" (reverse)\" if reverse else \" (forward)\"\n",
    "    full_label = translation_label + strand_label\n",
    "    \n",
    "    # Add a feature for the inserted sequence\n",
    "    start = insert_position\n",
    "    end = insert_position + len(new_sequence)\n",
    "    feature = SeqFeature(\n",
    "        location=FeatureLocation(start, end, strand=-1 if reverse else 1),\n",
    "        type=\"CDS\",  # CDS type for coding sequences\n",
    "        qualifiers={\n",
    "            \"label\": full_label,\n",
    "            \"translation\": Seq(new_sequence).translate(table=11)  # Translation for the sequence\n",
    "        }\n",
    "    )\n",
    "    record.features.append(feature)\n",
    "    # Save the modified GenBank file\n",
    "    if output_file:\n",
    "        SeqIO.write(record, output_file, \"genbank\")\n",
    "        print(f\"Updated GenBank file saved as {output_file}\")\n",
    "    \n",
    "    return record\n",
    "\n",
    "def insert_features_from_oligos(record, seq_id, seq, strand, tm, output_file):\n",
    "    \"\"\"\n",
    "    Insert features into a GenBank file based on oligo_df.\n",
    "\n",
    "    Args:\n",
    "        genbank_file (str): Path to the original GenBank file.\n",
    "        output_file (str): Path to save the updated GenBank file.\n",
    "        oligo_df (pd.DataFrame): DataFrame with oligo information. \n",
    "                                 Expected columns: ['id', 'oligo_id', 'oligo_sequence', 'oligo_length', ...].\n",
    "    \"\"\"\n",
    "    # Iterate through the oligo DataFrame to add features\n",
    "    start = record.seq.find(seq.upper())\n",
    "    if start == -1:\n",
    "        print(f\"Warning: Oligo sequence {seq_id} not found in the GenBank sequence.\")\n",
    "    \n",
    "    end = start + len(seq)\n",
    "    feature = SeqFeature(\n",
    "        location=FeatureLocation(start, end, strand=strand),\n",
    "        type=\"misc_feature\",\n",
    "        qualifiers={\n",
    "            \"label\": f\"{seq_id} {'(reverse)' if strand == -1 else '(forward)'}\",\n",
    "            \"note\": f\"Length: {len(seq)}, TM: {tm}\"\n",
    "        }\n",
    "    )\n",
    "    record.features.append(feature)\n",
    "    \n",
    "    # Save the updated GenBank file\n",
    "    if output_file:\n",
    "        SeqIO.write(record, output_file, \"genbank\")\n",
    "        print(f\"Updated GenBank file saved as {output_file}\")\n",
    "    return record\n",
    "    \n",
    "def check_secondary_structure(sequence, temp=55):\n",
    "    \"\"\"\n",
    "    Check secondary structures like hairpins and homodimers in a given primer sequence.\n",
    "    \n",
    "    Args:\n",
    "        sequence (str): The DNA sequence of the primer to analyze.\n",
    "        \n",
    "    Returns:\n",
    "        dict: Results for hairpin and homodimer properties.\n",
    "    \"\"\"\n",
    "    # Check for hairpin structure\n",
    "    hairpin_result = calcHairpin(sequence, temp_c=temp)\n",
    "    hairpin_info = {\n",
    "        \"hairpin_found\": hairpin_result.structure_found,\n",
    "        \"hairpin_tm\": hairpin_result.tm,\n",
    "        \"hairpin_dg\": hairpin_result.dg/1000,\n",
    "        \"hairpin_dh\": hairpin_result.dh/1000,\n",
    "        \"hairpin_ds\": hairpin_result.ds,\n",
    "    }\n",
    "\n",
    "    # Check for homodimer structure\n",
    "    homodimer_result = calcHomodimer(sequence, temp_c=temp)\n",
    "    homodimer_info = {\n",
    "        \"homodimer_found\": homodimer_result.structure_found,\n",
    "        \"homodimer_tm\": homodimer_result.tm,\n",
    "        \"homodimer_dg\": homodimer_result.dg/1000,\n",
    "        \"homodimer_dh\": homodimer_result.dh/1000,\n",
    "        \"homodimer_ds\": homodimer_result.ds,\n",
    "    }\n",
    "\n",
    "    # Combine results\n",
    "    return {\"hairpin\": hairpin_info, \"homodimer\": homodimer_info}\n",
    "\n",
    "def build_oligos(seq_id: str, sequence: str, output_directory: str, min_gc=0.3, max_gc=0.7, min_tm=55, max_tm=70, min_segment_length=40, max_segment_length=100, max_length=1500):\n",
    "    \"\"\" Use DNAweaver to build oligos \"\"\"\n",
    "    # Here we use a comercial supplier but don't actually care. \n",
    "    cheap_dna_offer = dw.CommercialDnaOffer(\n",
    "        name=\"CheapDNA.\",\n",
    "        sequence_constraints=[\n",
    "            dw.NoPatternConstraint(enzyme=\"BsaI\"),\n",
    "            dw.SequenceLengthConstraint(max_length=4000),\n",
    "            dw.GcContentConstraint(min_gc=min_gc, max_gc=max_gc)\n",
    "        ],\n",
    "        pricing=dw.PerBasepairPricing(0.10),\n",
    "    )\n",
    "\n",
    "    oligo_dna_offer = dw.CommercialDnaOffer(\n",
    "        name=\"OliGoo\",\n",
    "        sequence_constraints=[\n",
    "            dw.GcContentConstraint(min_gc=min_gc, max_gc=max_gc),\n",
    "            dw.SequenceLengthConstraint(max_length=4000),\n",
    "        ],\n",
    "        pricing=dw.PerBasepairPricing(0.07),\n",
    "        memoize=True\n",
    "    )\n",
    "\n",
    "    oligo_assembly_station = dw.DnaAssemblyStation(\n",
    "        name=\"Oligo Assembly Station\",\n",
    "        assembly_method=dw.OligoAssemblyMethod(\n",
    "            overhang_selector=dw.TmSegmentSelector(\n",
    "                min_size=15, max_size=25, min_tm=min_tm, max_tm=max_tm\n",
    "            ),\n",
    "            min_segment_length=min_segment_length,\n",
    "            max_segment_length=max_segment_length,\n",
    "            sequence_constraints=[dw.SequenceLengthConstraint(max_length=4000)],\n",
    "            duration=8,\n",
    "            cost=30,\n",
    "        ),\n",
    "        supplier=oligo_dna_offer,\n",
    "        coarse_grain=20,\n",
    "        a_star_factor=\"auto\",\n",
    "        memoize=True,\n",
    "    )\n",
    "\n",
    "    assembly_station = dw.DnaAssemblyStation(\n",
    "        name=\"Gibson Assembly Station\",\n",
    "        assembly_method=dw.GibsonAssemblyMethod(\n",
    "            overhang_selector=dw.TmSegmentSelector(min_tm=min_tm, max_tm=max_tm),\n",
    "            min_segment_length=min_segment_length,\n",
    "            max_segment_length=max_segment_length + 20, # add a bit of a buffer\n",
    "        ),\n",
    "        supplier=[cheap_dna_offer, oligo_assembly_station],\n",
    "        logger=\"bar\",\n",
    "        coarse_grain=100,\n",
    "        fine_grain=10,\n",
    "        a_star_factor=\"auto\",\n",
    "    )\n",
    "    \n",
    "    print(\"Looking for the best assembly plan...\")\n",
    "    t0 = time.time()\n",
    "    quote = assembly_station.get_quote(sequence, with_assembly_plan=True)\n",
    "    assembly_plan_report = quote.to_assembly_plan_report()\n",
    "    assembly_plan_report.write_full_report(f\"{output_directory}/oligo_assembly_plan_{seq_id}.zip\")\n",
    "    original_sequence = assembly_plan_report.plan.sequence\n",
    "    # Then get the sequence \n",
    "    rows = []\n",
    "    for oligo in assembly_plan_report.plan.assembly_plan:\n",
    "        # If this was chosen then choose it\n",
    "        if oligo.accepted:\n",
    "            rows.append([oligo.id, oligo.sequence, original_sequence])\n",
    "    return rows\n",
    "\n",
    "def get_oligos(df, protein_column, id_column, output_directory, forward_primer: str, reverse_primer: str, sequence_end: str, min_overlap=10, min_gc=0.3, \n",
    "               max_gc=0.7, min_tm=55, max_tm=70, min_segment_length=40, max_segment_length=100, max_length=1500, genbank_file=None, \n",
    "               insert_position=0, simple=False):\n",
    "    \"\"\" Get the oligos for a dataframe:\n",
    "    sequence_end is the end of the sequence i.e. TAA, TGA, etc or a histag \n",
    "    \"\"\"\n",
    "    rows = []   \n",
    "    for seq_id, protein_sequence in df[[id_column, protein_column]].values:\n",
    "        # Add on the primers that the user has provided\n",
    "        optimzed_sequence = codon_optimize(protein_sequence, min_gc, max_gc) + sequence_end\n",
    "        if genbank_file:\n",
    "            # Add in the optimzed sequence\n",
    "            translation_label = f\"Insert_{seq_id}\"\n",
    "            reverse = False  # Set to True for reverse feature, False for forward\n",
    "            record = insert_sequence_with_translation(genbank_file, None, insert_position, optimzed_sequence, translation_label, reverse)\n",
    "            \n",
    "        if optimzed_sequence[:3] != \"ATG\":\n",
    "            u.dp([f\"Warning: {seq_id} does not start with a methionine. \", optimzed_sequence[:3]])\n",
    "            if 'ATG' not in forward_primer:\n",
    "                u.warn_p([f\"Warning: {seq_id} does not start with a methionine. AND you don't have a methonine in your primer!!\", forward_primer])\n",
    "                print(\"We expect the primer to be in 5 to 3 prime direction.\")\n",
    "        # ALso check the end and or the reverse primer check for the three ones\n",
    "        if optimzed_sequence[-3:] != \"TAA\" and optimzed_sequence[-3:] != \"TGA\" and optimzed_sequence[-3:] != \"TAG\":\n",
    "            u.dp([f\"Warning: {seq_id} does not end with a stop codon. \", optimzed_sequence[-3:]])\n",
    "            if 'TAA' not in reverse_primer and \"TGA\" not in reverse_primer and \"TAG\" not in reverse_primer:\n",
    "                u.warn_p([f\"Warning: {seq_id} does not end with a stop codon. AND you don't have a stop codon in your primer!!\", reverse_primer])\n",
    "                print(\"We expect the primer to be in 5 to 3 prime direction.\")\n",
    "        codon_optimized_sequence = forward_primer + optimzed_sequence + reverse_primer\n",
    "        try:\n",
    "        # Check now some simple things like that there is \n",
    "            if simple:\n",
    "                oligos = build_simple_oligos(seq_id, codon_optimized_sequence, min_segment_length, max_segment_length)\n",
    "            else:\n",
    "                oligos = build_oligos(seq_id, codon_optimized_sequence, output_directory, min_gc, max_gc, min_tm, max_tm, min_segment_length, max_segment_length, max_length)\n",
    "        except Exception as e:\n",
    "            u.warn_p([f\"Warning: {seq_id} did not have any oligos built. \", e])\n",
    "            oligos = []\n",
    "        prev_oligo = None\n",
    "        # If a genbank file was provided also just add in the new sequnece\n",
    "        if len(oligos) > 0:\n",
    "            for i, oligo in enumerate(oligos):\n",
    "                seq = oligo[1]\n",
    "                # CHeck that there is an overlap with the previous sequence and that it is not too short\n",
    "                # Also make sure we swap the directions of the oligos so they automatically anneal\n",
    "                # Also assert that the start is a methionine (and if not warn it... )\n",
    "                primer_overlap = None\n",
    "                primer_tm = None\n",
    "                primer_len = None\n",
    "                homodimer_tm = None\n",
    "                hairpin_tm = None\n",
    "                if prev_oligo:\n",
    "                    # Get the overlap with the previous sequence\n",
    "                    match = SequenceMatcher(None, prev_oligo, seq).find_longest_match()\n",
    "                    primer_overlap = prev_oligo[match.a:match.a + match.size]\n",
    "                    # Analyze the primer sequence\n",
    "                    results = check_secondary_structure(primer_overlap)\n",
    "                    homodimer_tm = results['homodimer']['homodimer_dg']\n",
    "                    hairpin_tm = results['hairpin']['hairpin_dg']\n",
    "                    primer_tm = primer3.bindings.calcTm(primer_overlap)\n",
    "                    primer_len = len(primer_overlap)\n",
    "\n",
    "                prev_oligo = seq\n",
    "                orig_seq = seq\n",
    "                strand = 1\n",
    "                if i % 2 == 0:\n",
    "                    seq = str(Seq(seq).reverse_complement())\n",
    "                    strand = -1\n",
    "                oligo_tm = primer3.bindings.calcTm(seq)\n",
    "                if genbank_file:\n",
    "                    insert_features_from_oligos(record, f\"{seq_id}_oligo_{i}\", orig_seq, strand, oligo_tm, None)\n",
    "                rows.append([seq_id, oligo[0], seq, len(seq), oligo_tm, primer_overlap, primer_tm, primer_len, homodimer_tm, hairpin_tm, oligo[2]])\n",
    "            if genbank_file:\n",
    "                output_file = genbank_file.replace('.', f'_{seq_id}.')\n",
    "                record.name = seq_id\n",
    "                SeqIO.write(record, f'{output_directory}/{output_file}', \"genbank\")\n",
    "        else:\n",
    "            u.warn_p([f\"Warning: {seq_id} did not have any oligos built. \", optimzed_sequence])\n",
    "            rows.append([seq_id, None, optimzed_sequence, len(optimzed_sequence), None, None, None, None, None, None, None])\n",
    "        \n",
    "    oligo_df = pd.DataFrame(rows, columns=[\"id\", \"oligo_id\", \"oligo_sequence\", \"oligo_length\", \"oligo_tm\", \"primer_overlap_with_previous\", \"overlap_tm_5prime\", \"overlap_length\", \n",
    "                                            \"overlap_homodimer_dg\", \"overlap_hairpin_dg\", \"original_sequence\"])\n",
    "    \n",
    "        \n",
    "    return oligo_df\n",
    "\n",
    "\n",
    "def build_simple_oligos(seq_id: str, sequence: str, min_segment_length=40, max_segment_length=100, overlap_len=18):\n",
    "    # Basically get the splits size\n",
    "    seq_len = len(sequence)\n",
    "    num_splits = int(math.ceil(seq_len / max_segment_length))\n",
    "    # Make sure it is even\n",
    "    if num_splits % 2 != 0:\n",
    "        num_splits += 1\n",
    "    \n",
    "    remainder = seq_len % num_splits\n",
    "    # Check if there is a remainder, if so we need to add one to the splits and then share the new remainder \n",
    "    if remainder > 0:\n",
    "        num_splits += 1\n",
    "        remainder = seq_len % num_splits\n",
    "    \n",
    "    prev_overlap = ''\n",
    "    # Now we want to go through the new part length and while remainder is greater then 0 we distribute this across the splits\n",
    "    max_part_len = math.ceil(seq_len/num_splits)\n",
    "    \n",
    "    split_counts = {}\n",
    "    for i in range(0, num_splits + 1):\n",
    "        split_counts[i] = max_part_len\n",
    "    # Now distribute the remainder\n",
    "    split_count = 0\n",
    "    for i in range(0, remainder):\n",
    "        split_counts[split_count] += 1\n",
    "        split_count += 1\n",
    "        # Iterate through this again\n",
    "        if split_count == num_splits:\n",
    "            split_count = 0\n",
    "    prev_cut = 0\n",
    "    rows = []\n",
    "    print(split_counts)\n",
    "    for i in split_counts:\n",
    "        part_len = split_counts[i]\n",
    "        cut = prev_cut + part_len\n",
    "        oligo = sequence[prev_cut:cut]\n",
    "        print(prev_cut, part_len, cut, oligo)\n",
    "        rows.append([f'{seq_id}_{i}', oligo, sequence, prev_cut, cut, part_len])\n",
    "        prev_cut = cut - overlap_len\n",
    "    return rows\n",
    "\n",
    "\n",
    "def codon_optimize(protein_sequence: str, min_gc=0.3, max_gc=0.7):\n",
    "    \"\"\" Codon optimize the protein sequence using DNA chisel: https://github.com/Edinburgh-Genome-Foundry/DnaChisel\"\"\"\n",
    "    seq = dnachisel.reverse_translate(protein_sequence)\n",
    "    problem = dnachisel.DnaOptimizationProblem(\n",
    "        sequence=seq,\n",
    "        constraints=[\n",
    "            AvoidPattern(\"BsaI_site\"),\n",
    "            EnforceGCContent(mini=min_gc, maxi=max_gc, window=50),\n",
    "            EnforceTranslation(location=(0, len(seq))), \n",
    "            AvoidStopCodons(\n",
    "                location=(0, len(seq)-3)) # Let there be stop codons in the last bit\n",
    "        ],\n",
    "        objectives=[CodonOptimize(species='e_coli', location=(0, len(seq)))]\n",
    "    )\n",
    "    # SOLVE THE CONSTRAINTS, OPTIMIZE WITH RESPECT TO THE OBJECTIVE\n",
    "    problem.resolve_constraints()\n",
    "    problem.optimize()\n",
    "\n",
    "    # PRINT SUMMARIES TO CHECK THAT CONSTRAINTS PASS\n",
    "    print(problem.constraints_text_summary())\n",
    "    print(problem.objectives_text_summary())\n",
    "\n",
    "    # GET THE FINAL SEQUENCE (AS STRING OR ANNOTATED BIOPYTHON RECORDS)\n",
    "    final_sequence = problem.sequence  # string\n",
    "    final_record = problem.to_record(with_sequence_edits=True)\n",
    "    print(protein_sequence)\n",
    "    print(final_sequence)\n",
    "    return final_sequence\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "id": "683b7fd8-91bc-4676-8ed4-f7e82fdc31f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "objective:   0%|                                                                         | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-912](e_coli...]\n",
      "location:   0%|                                                                                                | 0/238 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                                                 | 0/238 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                                                 \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-912](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-912](mini:0.25, maxi:0.65, window:50)\n",
      "      │ Passed !\n",
      "✔PASS ┍ EnforceTranslation[0-912(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "✔PASS ┍ AvoidStopCodons[0-909]\n",
      "      │ All OK.\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:    -54.13\n",
      "    -54.13 ┍ MaximizeCAI[0-912](e_coli) \n",
      "           │ Codon opt. on window 0-912 scored -5.41E+01\n",
      "\n",
      "\n",
      "MRTSLLVAALGLALAAALPGGAPLAQPDPEATMDRSLLQRQDLPYRFSAVDLDSVDGQRHYRLWLGRPLQAPPAAGYPVVWMLDGNAAVGALDESTLRRLADGDAPLLVAIGYRTPLRIDRAGRTFDYTPASPGQADQRDPLNGLPSGGADAFLDLLRDGMRPAVAAQAPLDTARQTLWGHSYGGLLVLHALFTRPGEFARYAAASPSLWWRDGAILGERAGLEQRLRGKRAELLLWRGSAEPASPRGSLKAEPGQAMARLVDDLRRVAGLTLDFQPLDGLGHGETLGASLRLLLARPAVERQR\n",
      "ATGCGCACCAGCCTGCTGGTGGCGGCGCTGGGTTTAGCATTAGCAGCAGCATTACCGGGCGGTGCACCGTTAGCGCAGCCAGATCCAGAAGCGACTATGGATCGCAGCCTGCTGCAGCGCCAGGATCTGCCGTATCGCTTTAGCGCAGTGGATCTGGATAGCGTGGATGGCCAGCGCCATTATCGCCTGTGGCTGGGTCGTCCGTTACAAGCACCACCGGCAGCAGGTTATCCGGTGGTGTGGATGCTGGATGGCAACGCAGCGGTGGGTGCGTTAGATGAAAGCACCCTGCGCCGCTTAGCGGATGGCGATGCGCCGCTGTTAGTTGCGATTGGTTATCGCACCCCGCTGCGCATTGATCGCGCGGGCCGTACCTTTGATTATACCCCGGCAAGCCCGGGTCAAGCAGATCAGCGCGATCCGCTGAACGGTTTACCGAGCGGTGGCGCAGATGCGTTTCTGGATCTGCTGCGCGATGGTATGCGCCCAGCGGTTGCAGCACAAGCACCGTTAGATACCGCGCGCCAGACTTTATGGGGCCATAGCTATGGCGGCCTGCTGGTGTTACATGCGCTGTTTACCCGCCCGGGCGAATTTGCGCGCTATGCAGCAGCATCTCCGAGCTTATGGTGGCGCGATGGCGCGATTCTGGGCGAACGTGCAGGTCTGGAACAACGTCTGCGCGGCAAACGCGCAGAACTGCTGCTGTGGCGTGGTTCTGCAGAACCAGCGTCTCCACGCGGCAGCTTAAAAGCGGAACCGGGCCAAGCAATGGCACGCCTGGTGGATGATCTGCGTCGTGTTGCAGGCCTGACCCTGGATTTTCAGCCGCTGGATGGCCTGGGCCATGGTGAAACCCTGGGCGCATCTTTACGCTTACTGCTGGCGCGCCCGGCAGTTGAACGTCAGCGC\n",
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 108 (l) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 112 (p) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 108 (l) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 112 (p) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated GenBank file saved as base-pet22b-base-anm_Seq1.gb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "constraint:   0%|                                                                        | 0/3 [00:00<?, ?it/s, now=AvoidPattern[0-954](patte...]\n",
      "location:   0%|                                                                                                  | 0/8 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                                                  | 0/8 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "objective:   0%|                                                                         | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-954](e_coli...]\u001b[A\n",
      "location:   0%|                                                                                                | 0/202 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "location:   0%|                                                                                                 | 0/202 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                 | 0/202 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                                                 \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-954](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-954](mini:0.25, maxi:0.65, window:50)\n",
      "      │ Passed !\n",
      "✔PASS ┍ EnforceTranslation[0-954(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "✔PASS ┍ AvoidStopCodons[0-951]\n",
      "      │ All OK.\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:     -6.31\n",
      "     -6.31 ┍ MaximizeCAI[0-954](e_coli) \n",
      "           │ Codon opt. on window 0-954 scored -6.31E+00\n",
      "\n",
      "\n",
      "MQLFDLPLDQLQTYKPEKTAPKDFSEFWKLSLEELAKVQAEPDLQPVDYPADGVKVYRLTYKSFGNARITGWYAVPDKEGPHPAIVKYHGYNASYDGEIHEMVNWALHGYATFGMLVRGQQSSEDTSISPHGHALGWMTKGILDKDTYYYRGVYLDAVRALEVISSFDEVDETRIGVTGGSQGGGLTIAAAALSDIPKAAVADYPYLSNFERAIDVALEQPYLEINSFFRRNGSPETEVQAMKTLSYFDIMNLADRVKVPVLMSIGLIDKVTPPSTVFAAYNHLETKKELKVYRYFGHEYIPAFQTEKLAFFKQHLKG\n",
      "ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC\n",
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                                                                                 \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Updated GenBank file saved as base-pet22b-base-anm_Seq2.gb\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 108 (l) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 112 (p) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 108 (l) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/dnaweaver/AssemblyPlanReport/mixins/PlotsMixin/PlotsMixin.py:18: UserWarning: Glyph 112 (p) missing from font(s) FontAwesome.\n",
      "  ax.figure.savefig(target.open(\"wb\"), format=\"pdf\", bbox_inches=\"tight\")\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>oligo_id</th>\n",
       "      <th>oligo_sequence</th>\n",
       "      <th>oligo_length</th>\n",
       "      <th>oligo_tm</th>\n",
       "      <th>primer_overlap_with_previous</th>\n",
       "      <th>overlap_tm_5prime</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>overlap_homodimer_dg</th>\n",
       "      <th>overlap_hairpin_dg</th>\n",
       "      <th>original_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00000</td>\n",
       "      <td>CTAACGGTGCACCGCCCGGTAATGCTGCTGCTAATGCTAAACCCAG...</td>\n",
       "      <td>119</td>\n",
       "      <td>77.102269</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00001</td>\n",
       "      <td>CGGGCGGTGCACCGTTAGCGCAGCCAGATCCAGAAGCGACTATGGA...</td>\n",
       "      <td>128</td>\n",
       "      <td>86.673171</td>\n",
       "      <td>CGGGCGGTGCACCGTTAG</td>\n",
       "      <td>61.719427</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-9.460174</td>\n",
       "      <td>-0.339302</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00002</td>\n",
       "      <td>CGCAGGGTGCTTTCATCTAACGCACCCACCGCTGCGTTGCCATCCA...</td>\n",
       "      <td>128</td>\n",
       "      <td>86.996315</td>\n",
       "      <td>GATGGCCAGCGCCATTAT</td>\n",
       "      <td>55.811457</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.719598</td>\n",
       "      <td>-1.247846</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00003</td>\n",
       "      <td>AGATGAAAGCACCCTGCGCCGCTTAGCGGATGGCGATGCGCCGCTG...</td>\n",
       "      <td>128</td>\n",
       "      <td>86.996315</td>\n",
       "      <td>AGATGAAAGCACCCTGCG</td>\n",
       "      <td>55.588807</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.312299</td>\n",
       "      <td>0.316358</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00004</td>\n",
       "      <td>TAACGGTGCTTGTGCTGCAACCGCTGGGCGCATACCATCGCGCAGC...</td>\n",
       "      <td>128</td>\n",
       "      <td>87.642603</td>\n",
       "      <td>CCCCGGCAAGCCCGGGTC</td>\n",
       "      <td>66.182834</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.525976</td>\n",
       "      <td>-2.580547</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00005</td>\n",
       "      <td>GCAGCACAAGCACCGTTAGATACCGCGCGCCAGACTTTATGGGGCC...</td>\n",
       "      <td>128</td>\n",
       "      <td>86.350028</td>\n",
       "      <td>GCAGCACAAGCACCGTTA</td>\n",
       "      <td>56.557603</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-0.875108</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00006</td>\n",
       "      <td>ACGCTGGTTCTGCAGAACCACGCCACAGCAGCAGTTCTGCGCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>86.996315</td>\n",
       "      <td>TGCAGCAGCATCTCCGAG</td>\n",
       "      <td>57.731341</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.097456</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00007</td>\n",
       "      <td>GTTCTGCAGAACCAGCGTCTCCACGCGGCAGCTTAAAAGCGGAACC...</td>\n",
       "      <td>128</td>\n",
       "      <td>86.673171</td>\n",
       "      <td>GTTCTGCAGAACCAGCGT</td>\n",
       "      <td>55.847940</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-6.742123</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00008</td>\n",
       "      <td>TTCAACTGCCGGGCGCGCCAGCAGTAAGCGTAAAGATGCGCCCAGG...</td>\n",
       "      <td>78</td>\n",
       "      <td>84.977510</td>\n",
       "      <td>CAGCCGCTGGATGGCCTG</td>\n",
       "      <td>61.498122</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.630546</td>\n",
       "      <td>0.039317</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00009</td>\n",
       "      <td>GCGCGCCCGGCAGTTGAACGTCAGCGCCTCGAGCACCACCACCACC...</td>\n",
       "      <td>73</td>\n",
       "      <td>78.900785</td>\n",
       "      <td>GCGCGCCCGGCAGTTGAA</td>\n",
       "      <td>64.895379</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-5.908807</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00000</td>\n",
       "      <td>TAAAATCTTTCGGCGCGGTTTTTTCCGGTTTATAGGTCTGCAGCTG...</td>\n",
       "      <td>119</td>\n",
       "      <td>73.278854</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00001</td>\n",
       "      <td>CCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGA...</td>\n",
       "      <td>129</td>\n",
       "      <td>83.628370</td>\n",
       "      <td>CCGCGCCGAAAGATTTTA</td>\n",
       "      <td>54.090460</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.281008</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00002</td>\n",
       "      <td>TCGCCATCATAGCTCGCGTTATAGCCATGATATTTCACAATCGCCG...</td>\n",
       "      <td>128</td>\n",
       "      <td>83.441734</td>\n",
       "      <td>GTGTATCGCCTGACCTATA</td>\n",
       "      <td>51.006043</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-0.325379</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00003</td>\n",
       "      <td>CGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTG...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.057453</td>\n",
       "      <td>CGCGAGCTATGATGGCGA</td>\n",
       "      <td>57.967543</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.056448</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00004</td>\n",
       "      <td>CATCCACTTCATCAAAGCTGCTAATCACTTCCAGCGCGCGCACCGC...</td>\n",
       "      <td>129</td>\n",
       "      <td>83.628370</td>\n",
       "      <td>GCCCGCATGGCCATGCGC</td>\n",
       "      <td>66.122712</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-12.838022</td>\n",
       "      <td>-1.587316</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00005</td>\n",
       "      <td>AGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCA...</td>\n",
       "      <td>129</td>\n",
       "      <td>84.910925</td>\n",
       "      <td>AGCTTTGATGAAGTGGATG</td>\n",
       "      <td>50.750810</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-1.013548</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00006</td>\n",
       "      <td>GGGTTTTCATCGCCTGCACTTCGGTTTCCGGGCTGCCGTTGCGGCG...</td>\n",
       "      <td>128</td>\n",
       "      <td>83.118590</td>\n",
       "      <td>GGATTATCCGTATCTGAGC</td>\n",
       "      <td>49.967947</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-0.797296</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00007</td>\n",
       "      <td>TGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGC...</td>\n",
       "      <td>128</td>\n",
       "      <td>83.118590</td>\n",
       "      <td>TGCAGGCGATGAAAACCC</td>\n",
       "      <td>56.217177</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.037659</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00008</td>\n",
       "      <td>GCCTTTCAGATGCTGTTTAAAAAACGCCAGTTTTTCGGTCTGAAAC...</td>\n",
       "      <td>129</td>\n",
       "      <td>78.818789</td>\n",
       "      <td>ACCGTGTTTGCGGCGTAT</td>\n",
       "      <td>58.255093</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.467608</td>\n",
       "      <td>0.035670</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00009</td>\n",
       "      <td>TAAACAGCATCTGAAAGGCCTCGAGCACCACCACCACCACCACTGA...</td>\n",
       "      <td>65</td>\n",
       "      <td>72.833305</td>\n",
       "      <td>TAAACAGCATCTGAAAGGC</td>\n",
       "      <td>51.093793</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-0.061540</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id oligo_id                                     oligo_sequence  \\\n",
       "0   Seq1  S_00000  CTAACGGTGCACCGCCCGGTAATGCTGCTGCTAATGCTAAACCCAG...   \n",
       "1   Seq1  S_00001  CGGGCGGTGCACCGTTAGCGCAGCCAGATCCAGAAGCGACTATGGA...   \n",
       "2   Seq1  S_00002  CGCAGGGTGCTTTCATCTAACGCACCCACCGCTGCGTTGCCATCCA...   \n",
       "3   Seq1  S_00003  AGATGAAAGCACCCTGCGCCGCTTAGCGGATGGCGATGCGCCGCTG...   \n",
       "4   Seq1  S_00004  TAACGGTGCTTGTGCTGCAACCGCTGGGCGCATACCATCGCGCAGC...   \n",
       "5   Seq1  S_00005  GCAGCACAAGCACCGTTAGATACCGCGCGCCAGACTTTATGGGGCC...   \n",
       "6   Seq1  S_00006  ACGCTGGTTCTGCAGAACCACGCCACAGCAGCAGTTCTGCGCGTTT...   \n",
       "7   Seq1  S_00007  GTTCTGCAGAACCAGCGTCTCCACGCGGCAGCTTAAAAGCGGAACC...   \n",
       "8   Seq1  S_00008  TTCAACTGCCGGGCGCGCCAGCAGTAAGCGTAAAGATGCGCCCAGG...   \n",
       "9   Seq1  S_00009  GCGCGCCCGGCAGTTGAACGTCAGCGCCTCGAGCACCACCACCACC...   \n",
       "10  Seq2  S_00000  TAAAATCTTTCGGCGCGGTTTTTTCCGGTTTATAGGTCTGCAGCTG...   \n",
       "11  Seq2  S_00001  CCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGA...   \n",
       "12  Seq2  S_00002  TCGCCATCATAGCTCGCGTTATAGCCATGATATTTCACAATCGCCG...   \n",
       "13  Seq2  S_00003  CGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTG...   \n",
       "14  Seq2  S_00004  CATCCACTTCATCAAAGCTGCTAATCACTTCCAGCGCGCGCACCGC...   \n",
       "15  Seq2  S_00005  AGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCA...   \n",
       "16  Seq2  S_00006  GGGTTTTCATCGCCTGCACTTCGGTTTCCGGGCTGCCGTTGCGGCG...   \n",
       "17  Seq2  S_00007  TGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGC...   \n",
       "18  Seq2  S_00008  GCCTTTCAGATGCTGTTTAAAAAACGCCAGTTTTTCGGTCTGAAAC...   \n",
       "19  Seq2  S_00009  TAAACAGCATCTGAAAGGCCTCGAGCACCACCACCACCACCACTGA...   \n",
       "\n",
       "    oligo_length   oligo_tm primer_overlap_with_previous  overlap_tm_5prime  \\\n",
       "0            119  77.102269                         None                NaN   \n",
       "1            128  86.673171           CGGGCGGTGCACCGTTAG          61.719427   \n",
       "2            128  86.996315           GATGGCCAGCGCCATTAT          55.811457   \n",
       "3            128  86.996315           AGATGAAAGCACCCTGCG          55.588807   \n",
       "4            128  87.642603           CCCCGGCAAGCCCGGGTC          66.182834   \n",
       "5            128  86.350028           GCAGCACAAGCACCGTTA          56.557603   \n",
       "6            128  86.996315           TGCAGCAGCATCTCCGAG          57.731341   \n",
       "7            128  86.673171           GTTCTGCAGAACCAGCGT          55.847940   \n",
       "8             78  84.977510           CAGCCGCTGGATGGCCTG          61.498122   \n",
       "9             73  78.900785           GCGCGCCCGGCAGTTGAA          64.895379   \n",
       "10           119  73.278854                         None                NaN   \n",
       "11           129  83.628370           CCGCGCCGAAAGATTTTA          54.090460   \n",
       "12           128  83.441734          GTGTATCGCCTGACCTATA          51.006043   \n",
       "13           128  85.057453           CGCGAGCTATGATGGCGA          57.967543   \n",
       "14           129  83.628370           GCCCGCATGGCCATGCGC          66.122712   \n",
       "15           129  84.910925          AGCTTTGATGAAGTGGATG          50.750810   \n",
       "16           128  83.118590          GGATTATCCGTATCTGAGC          49.967947   \n",
       "17           128  83.118590           TGCAGGCGATGAAAACCC          56.217177   \n",
       "18           129  78.818789           ACCGTGTTTGCGGCGTAT          58.255093   \n",
       "19            65  72.833305          TAAACAGCATCTGAAAGGC          51.093793   \n",
       "\n",
       "    overlap_length  overlap_homodimer_dg  overlap_hairpin_dg  \\\n",
       "0              NaN                   NaN                 NaN   \n",
       "1             18.0             -9.460174           -0.339302   \n",
       "2             18.0             -1.719598           -1.247846   \n",
       "3             18.0             -1.312299            0.316358   \n",
       "4             18.0             -4.525976           -2.580547   \n",
       "5             18.0             -0.875108            0.000000   \n",
       "6             18.0             -2.097456            0.000000   \n",
       "7             18.0             -6.742123            0.000000   \n",
       "8             18.0             -2.630546            0.039317   \n",
       "9             18.0             -5.908807            0.000000   \n",
       "10             NaN                   NaN                 NaN   \n",
       "11            18.0             -3.281008            0.000000   \n",
       "12            19.0             -0.325379            0.000000   \n",
       "13            18.0             -4.056448            0.000000   \n",
       "14            18.0            -12.838022           -1.587316   \n",
       "15            19.0             -1.013548            0.000000   \n",
       "16            19.0             -0.797296            0.000000   \n",
       "17            18.0             -2.037659            0.000000   \n",
       "18            18.0             -3.467608            0.035670   \n",
       "19            19.0             -0.061540            0.000000   \n",
       "\n",
       "                                    original_sequence  \n",
       "0   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "1   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "2   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "3   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "4   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "5   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "6   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "7   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "8   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "9   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "10  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "11  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "12  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "13  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "14  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "15  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "16  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "17  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "18  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "19  cccctctagaaataattttgtttaactttaagaaggagatatacat...  "
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import pandas as pd\n",
    "import random\n",
    "import numpy \n",
    "  \n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['id'] = ['Seq1', 'Seq2']\n",
    "min_gc = 0.25\n",
    "max_gc = 0.65\n",
    "min_tm = 10\n",
    "max_tm = 100\n",
    "min_segment_length = 40\n",
    "max_segment_length = 100\n",
    "max_length = 500\n",
    "\n",
    "SEED = i\n",
    "random.seed(SEED)\n",
    "numpy.random.seed(SEED)\n",
    "df['seqs'] = ['MRTSLLVAALGLALAAALPGGAPLAQPDPEATMDRSLLQRQDLPYRFSAVDLDSVDGQRHYRLWLGRPLQAPPAAGYPVVWMLDGNAAVGALDESTLRRLADGDAPLLVAIGYRTPLRIDRAGRTFDYTPASPGQADQRDPLNGLPSGGADAFLDLLRDGMRPAVAAQAPLDTARQTLWGHSYGGLLVLHALFTRPGEFARYAAASPSLWWRDGAILGERAGLEQRLRGKRAELLLWRGSAEPASPRGSLKAEPGQAMARLVDDLRRVAGLTLDFQPLDGLGHGETLGASLRLLLARPAVERQR', \n",
    "             'MQLFDLPLDQLQTYKPEKTAPKDFSEFWKLSLEELAKVQAEPDLQPVDYPADGVKVYRLTYKSFGNARITGWYAVPDKEGPHPAIVKYHGYNASYDGEIHEMVNWALHGYATFGMLVRGQQSSEDTSISPHGHALGWMTKGILDKDTYYYRGVYLDAVRALEVISSFDEVDETRIGVTGGSQGGGLTIAAAALSDIPKAAVADYPYLSNFERAIDVALEQPYLEINSFFRRNGSPETEVQAMKTLSYFDIMNLADRVKVPVLMSIGLIDKVTPPSTVFAAYNHLETKKELKVYRYFGHEYIPAFQTEKLAFFKQHLKG']\n",
    "oligo_df = get_oligos(df, 'seqs', 'id', '.', 'cccctctagaaataattttgtttaactttaagaaggagatatacat', 'gatccggctgctaacaaag', sequence_end='CTCGAGCACCACCACCACCACCACTGA',\n",
    "                     min_gc=min_gc, max_gc=max_gc, min_tm=min_tm, max_tm=max_tm, min_segment_length=min_segment_length, max_segment_length=max_segment_length, \n",
    "                     genbank_file=\"base-pet22b-base-anm.gb\", insert_position=5193)\n",
    "\n",
    "oligo_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "id": "e93c5950-4650-46c6-b9b6-b129515322ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def build_simple_oligos(seq_id: str, sequence: str, min_segment_length=40, max_segment_length=100, overlap_len=18):\n",
    "    # Basically get the splits size\n",
    "    seq_len = len(sequence)\n",
    "    num_splits = int(math.ceil(seq_len / max_segment_length))\n",
    "    # Make sure it is even\n",
    "    if num_splits % 2 != 0:\n",
    "        num_splits += 1\n",
    "    remainder = seq_len % num_splits\n",
    "    # Check if there is a remainder, if so we need to add one to the splits and then share the new remainder \n",
    "    print(num_splits, remainder)        \n",
    "    prev_overlap = ''\n",
    "    # Now we want to go through the new part length and while remainder is greater then 0 we distribute this across the splits\n",
    "    max_part_len = math.ceil(seq_len/num_splits)\n",
    "    \n",
    "    split_counts = {}\n",
    "    for i in range(0, num_splits + 1):\n",
    "        split_counts[i] = max_part_len\n",
    "    # Now distribute the remainder\n",
    "    split_count = 0\n",
    "    print(max_part_len, num_splits, seq_len, max_part_len * num_splits)\n",
    "    for i in range(0, remainder + 1):\n",
    "        split_counts[split_count] += 1\n",
    "        split_count += 1\n",
    "        # Iterate through this again\n",
    "        if split_count == num_splits + 1:\n",
    "            split_count = 0\n",
    "    prev_cut = 0\n",
    "    rows = []\n",
    "    print(split_counts)\n",
    "    for i in split_counts:\n",
    "        part_len = split_counts[i]\n",
    "        cut = prev_cut + part_len\n",
    "        oligo = sequence[prev_cut:cut]\n",
    "        print(prev_cut, part_len, cut, oligo)\n",
    "        rows.append([f'{seq_id}_{i}', oligo, sequence, prev_cut, cut, part_len])\n",
    "        prev_cut = cut - overlap_len\n",
    "    # Add in the last one \n",
    "    oligo = sequence[prev_cut:]\n",
    "    if len(oligo) > 0:\n",
    "        part_len = len(oligo)\n",
    "        cut = len(sequence)\n",
    "        if len(oligo) < 18:\n",
    "            u.warn_p([\"Last oligo very short,.... check this!\", f'{seq_id}_{i}', oligo, sequence, prev_cut, cut, part_len])\n",
    "        print(prev_cut, part_len, len(sequence), oligo)\n",
    "        rows.append([f'{seq_id}_{i}', oligo, sequence, prev_cut, cut, part_len])\n",
    "    return rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "id": "71437483-8e92-4cdd-8157-42ec64a6a3b3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10 4\n",
      "96 10 954 960\n",
      "{0: 97, 1: 97, 2: 97, 3: 97, 4: 97, 5: 96, 6: 96, 7: 96, 8: 96, 9: 96, 10: 96}\n",
      "0 97 97 ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGG\n",
      "79 97 176 TTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCT\n",
      "158 97 255 CGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATT\n",
      "237 97 334 GGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGA\n",
      "316 97 413 CGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGAT\n",
      "395 96 491 CCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGAT\n",
      "473 96 569 GCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGC\n",
      "551 96 647 TGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGT\n",
      "629 96 725 TGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGAT\n",
      "707 96 803 AACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGAT\n",
      "785 96 881 GATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCG\n",
      "863 91 954 AGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[['test_0',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGG',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  0,\n",
       "  97,\n",
       "  97],\n",
       " ['test_1',\n",
       "  'TTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCT',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  79,\n",
       "  176,\n",
       "  97],\n",
       " ['test_2',\n",
       "  'CGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATT',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  158,\n",
       "  255,\n",
       "  97],\n",
       " ['test_3',\n",
       "  'GGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGA',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  237,\n",
       "  334,\n",
       "  97],\n",
       " ['test_4',\n",
       "  'CGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGAT',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  316,\n",
       "  413,\n",
       "  97],\n",
       " ['test_5',\n",
       "  'CCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGAT',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  395,\n",
       "  491,\n",
       "  96],\n",
       " ['test_6',\n",
       "  'GCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGC',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  473,\n",
       "  569,\n",
       "  96],\n",
       " ['test_7',\n",
       "  'TGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGT',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  551,\n",
       "  647,\n",
       "  96],\n",
       " ['test_8',\n",
       "  'TGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGAT',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  629,\n",
       "  725,\n",
       "  96],\n",
       " ['test_9',\n",
       "  'AACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGAT',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  707,\n",
       "  803,\n",
       "  96],\n",
       " ['test_10',\n",
       "  'GATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCG',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  785,\n",
       "  881,\n",
       "  96],\n",
       " ['test_10',\n",
       "  'AGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC',\n",
       "  863,\n",
       "  954,\n",
       "  91]]"
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "build_simple_oligos('test', 'ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC', \n",
    "                    40, 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "id": "d7ecadfa-8b6f-4fca-8bb9-373a0fe2c2b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "objective:   0%|                                                                         | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-912](e_coli...]\n",
      "location:   0%|                                                                                                | 0/238 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                                                 | 0/238 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-912](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-912](mini:0.25, maxi:0.65, window:50)\n",
      "      │ Passed !\n",
      "✔PASS ┍ EnforceTranslation[0-912(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "✔PASS ┍ AvoidStopCodons[0-909]\n",
      "      │ All OK.\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:    -54.13\n",
      "    -54.13 ┍ MaximizeCAI[0-912](e_coli) \n",
      "           │ Codon opt. on window 0-912 scored -5.41E+01\n",
      "\n",
      "\n",
      "MRTSLLVAALGLALAAALPGGAPLAQPDPEATMDRSLLQRQDLPYRFSAVDLDSVDGQRHYRLWLGRPLQAPPAAGYPVVWMLDGNAAVGALDESTLRRLADGDAPLLVAIGYRTPLRIDRAGRTFDYTPASPGQADQRDPLNGLPSGGADAFLDLLRDGMRPAVAAQAPLDTARQTLWGHSYGGLLVLHALFTRPGEFARYAAASPSLWWRDGAILGERAGLEQRLRGKRAELLLWRGSAEPASPRGSLKAEPGQAMARLVDDLRRVAGLTLDFQPLDGLGHGETLGASLRLLLARPAVERQR\n",
      "ATGCGCACCAGCCTGCTGGTGGCGGCGCTGGGTTTAGCATTAGCAGCAGCATTACCGGGCGGTGCACCGTTAGCGCAGCCAGATCCAGAAGCGACTATGGATCGCAGCCTGCTGCAGCGCCAGGATCTGCCGTATCGCTTTAGCGCAGTGGATCTGGATAGCGTGGATGGCCAGCGCCATTATCGCCTGTGGCTGGGTCGTCCGTTACAAGCACCACCGGCAGCAGGTTATCCGGTGGTGTGGATGCTGGATGGCAACGCAGCGGTGGGTGCGTTAGATGAAAGCACCCTGCGCCGCTTAGCGGATGGCGATGCGCCGCTGTTAGTTGCGATTGGTTATCGCACCCCGCTGCGCATTGATCGCGCGGGCCGTACCTTTGATTATACCCCGGCAAGCCCGGGTCAAGCAGATCAGCGCGATCCGCTGAACGGTTTACCGAGCGGTGGCGCAGATGCGTTTCTGGATCTGCTGCGCGATGGTATGCGCCCAGCGGTTGCAGCACAAGCACCGTTAGATACCGCGCGCCAGACTTTATGGGGCCATAGCTATGGCGGCCTGCTGGTGTTACATGCGCTGTTTACCCGCCCGGGCGAATTTGCGCGCTATGCAGCAGCATCTCCGAGCTTATGGTGGCGCGATGGCGCGATTCTGGGCGAACGTGCAGGTCTGGAACAACGTCTGCGCGGCAAACGCGCAGAACTGCTGCTGTGGCGTGGTTCTGCAGAACCAGCGTCTCCACGCGGCAGCTTAAAAGCGGAACCGGGCCAAGCAATGGCACGCCTGGTGGATGATCTGCGTCGTGTTGCAGGCCTGACCCTGGATTTTCAGCCGCTGGATGGCCTGGGCCATGGTGAAACCCTGGGCGCATCTTTACGCTTACTGCTGGCGCGCCCGGCAGTTGAACGTCAGCGC\n",
      "{0: 79, 1: 79, 2: 79, 3: 78, 4: 78, 5: 78, 6: 78, 7: 78, 8: 78, 9: 78, 10: 78, 11: 78, 12: 78, 13: 78}\n",
      "0 79 79 cccctctagaaataattttgtttaactttaagaaggagatatacatATGCGCACCAGCCTGCTGGTGGCGGCGCTGGGT\n",
      "61 79 140 CTGGTGGCGGCGCTGGGTTTAGCATTAGCAGCAGCATTACCGGGCGGTGCACCGTTAGCGCAGCCAGATCCAGAAGCGA\n",
      "122 79 201 AGCCAGATCCAGAAGCGACTATGGATCGCAGCCTGCTGCAGCGCCAGGATCTGCCGTATCGCTTTAGCGCAGTGGATCT\n",
      "183 78 261 CTTTAGCGCAGTGGATCTGGATAGCGTGGATGGCCAGCGCCATTATCGCCTGTGGCTGGGTCGTCCGTTACAAGCACC\n",
      "243 78 321 TCGTCCGTTACAAGCACCACCGGCAGCAGGTTATCCGGTGGTGTGGATGCTGGATGGCAACGCAGCGGTGGGTGCGTT\n",
      "303 78 381 CGCAGCGGTGGGTGCGTTAGATGAAAGCACCCTGCGCCGCTTAGCGGATGGCGATGCGCCGCTGTTAGTTGCGATTGG\n",
      "363 78 441 GCTGTTAGTTGCGATTGGTTATCGCACCCCGCTGCGCATTGATCGCGCGGGCCGTACCTTTGATTATACCCCGGCAAG\n",
      "423 78 501 TGATTATACCCCGGCAAGCCCGGGTCAAGCAGATCAGCGCGATCCGCTGAACGGTTTACCGAGCGGTGGCGCAGATGC\n",
      "483 78 561 GAGCGGTGGCGCAGATGCGTTTCTGGATCTGCTGCGCGATGGTATGCGCCCAGCGGTTGCAGCACAAGCACCGTTAGA\n",
      "543 78 621 AGCACAAGCACCGTTAGATACCGCGCGCCAGACTTTATGGGGCCATAGCTATGGCGGCCTGCTGGTGTTACATGCGCT\n",
      "603 78 681 GCTGGTGTTACATGCGCTGTTTACCCGCCCGGGCGAATTTGCGCGCTATGCAGCAGCATCTCCGAGCTTATGGTGGCG\n",
      "663 78 741 TCCGAGCTTATGGTGGCGCGATGGCGCGATTCTGGGCGAACGTGCAGGTCTGGAACAACGTCTGCGCGGCAAACGCGC\n",
      "723 78 801 TCTGCGCGGCAAACGCGCAGAACTGCTGCTGTGGCGTGGTTCTGCAGAACCAGCGTCTCCACGCGGCAGCTTAAAAGC\n",
      "783 78 861 ACGCGGCAGCTTAAAAGCGGAACCGGGCCAAGCAATGGCACGCCTGGTGGATGATCTGCGTCGTGTTGCAGGCCTGAC\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "constraint:   0%|                                                                        | 0/3 [00:00<?, ?it/s, now=AvoidPattern[0-954](patte...]\n",
      "location:   0%|                                                                                                  | 0/8 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                                                  | 0/8 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "objective:   0%|                                                                         | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-954](e_coli...]\u001b[A\n",
      "location:   0%|                                                                                                | 0/202 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "location:   0%|                                                                                                 | 0/202 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                                                 \u001b[A\n",
      "location:   0%|                                                                                                 | 0/202 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                                                 \u001b[A"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-954](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-954](mini:0.25, maxi:0.65, window:50)\n",
      "      │ Passed !\n",
      "✔PASS ┍ EnforceTranslation[0-954(+)]\n",
      "      │ Enforced by nucleotides restrictions\n",
      "✔PASS ┍ AvoidStopCodons[0-951]\n",
      "      │ All OK.\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:     -6.31\n",
      "     -6.31 ┍ MaximizeCAI[0-954](e_coli) \n",
      "           │ Codon opt. on window 0-954 scored -6.31E+00\n",
      "\n",
      "\n",
      "MQLFDLPLDQLQTYKPEKTAPKDFSEFWKLSLEELAKVQAEPDLQPVDYPADGVKVYRLTYKSFGNARITGWYAVPDKEGPHPAIVKYHGYNASYDGEIHEMVNWALHGYATFGMLVRGQQSSEDTSISPHGHALGWMTKGILDKDTYYYRGVYLDAVRALEVISSFDEVDETRIGVTGGSQGGGLTIAAAALSDIPKAAVADYPYLSNFERAIDVALEQPYLEINSFFRRNGSPETEVQAMKTLSYFDIMNLADRVKVPVLMSIGLIDKVTPPSTVFAAYNHLETKKELKVYRYFGHEYIPAFQTEKLAFFKQHLKG\n",
      "ATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAAAAAAGAACTGAAAGTGTATCGCTATTTTGGCCATGAATATATTCCGGCGTTTCAGACCGAAAAACTGGCGTTTTTTAAACAGCATCTGAAAGGC\n",
      "{0: 82, 1: 82, 2: 82, 3: 82, 4: 82, 5: 82, 6: 81, 7: 81, 8: 81, 9: 81, 10: 81, 11: 81, 12: 81, 13: 81}\n",
      "0 82 82 cccctctagaaataattttgtttaactttaagaaggagatatacatATGCAGCTGTTTGATCTGCCGCTGGATCAGCTGCAG\n",
      "64 82 146 CCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAATTTTGGAAACTGAGCCTGGAAG\n",
      "128 82 210 GGAAACTGAGCCTGGAAGAACTGGCGAAAGTGCAGGCGGAACCGGATCTGCAGCCGGTGGATTATCCGGCAGATGGCGTGAA\n",
      "192 82 274 TCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCG\n",
      "256 82 338 GGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGAGCTATGATGGCG\n",
      "320 82 402 ACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCTGCATGGCTATGCGACCTTTGGCATGCTGGTGCGCGG\n",
      "384 81 465 TGGCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAGCCCGCATGGCCATGCGCTGGGCTGGATGACCAA\n",
      "447 81 528 GCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTATTATCGCGGCGTGTATCTGGATGCGGTGCGCGCGCT\n",
      "510 81 591 GGATGCGGTGCGCGCGCTGGAAGTGATTAGCAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCAGCCA\n",
      "573 81 654 CGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCAGCGCTGAGCGATATTCCGAAAGCGGCGGTGGCGGA\n",
      "636 81 717 GAAAGCGGCGGTGGCGGATTATCCGTATCTGAGCAACTTTGAACGCGCGATTGATGTGGCGCTGGAACAGCCGTATCTGGA\n",
      "699 81 780 GGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGCAGCCCGGAAACCGAAGTGCAGGCGATGAAAACCCT\n",
      "762 81 843 GCAGGCGATGAAAACCCTGAGCTATTTTGATATTATGAACCTGGCGGATCGCGTGAAAGTGCCGGTGCTGATGAGCATTGG\n",
      "825 81 906 GGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGCACCGTGTTTGCGGCGTATAACCATCTGGAAACCAA\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>oligo_id</th>\n",
       "      <th>oligo_sequence</th>\n",
       "      <th>oligo_length</th>\n",
       "      <th>oligo_tm</th>\n",
       "      <th>primer_overlap_with_previous</th>\n",
       "      <th>overlap_tm_5prime</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>overlap_homodimer_dg</th>\n",
       "      <th>overlap_hairpin_dg</th>\n",
       "      <th>original_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_0</td>\n",
       "      <td>ACCCAGCGCCGCCACCAGCAGGCTGGTGCGCATatgtatatctcct...</td>\n",
       "      <td>79</td>\n",
       "      <td>71.126314</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_1</td>\n",
       "      <td>CTGGTGGCGGCGCTGGGTTTAGCATTAGCAGCAGCATTACCGGGCG...</td>\n",
       "      <td>79</td>\n",
       "      <td>84.215682</td>\n",
       "      <td>CTGGTGGCGGCGCTGGGT</td>\n",
       "      <td>66.043739</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-6.604577</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_2</td>\n",
       "      <td>AGATCCACTGCGCTAAAGCGATACGGCAGATCCTGGCGCTGCAGCA...</td>\n",
       "      <td>79</td>\n",
       "      <td>82.644958</td>\n",
       "      <td>AGCCAGATCCAGAAGCGA</td>\n",
       "      <td>55.856688</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.037659</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_3</td>\n",
       "      <td>CTTTAGCGCAGTGGATCTGGATAGCGTGGATGGCCAGCGCCATTAT...</td>\n",
       "      <td>78</td>\n",
       "      <td>82.856361</td>\n",
       "      <td>CTTTAGCGCAGTGGATCT</td>\n",
       "      <td>52.267941</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.342888</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_4</td>\n",
       "      <td>AACGCACCCACCGCTGCGTTGCCATCCAGCATCCACACCACCGGAT...</td>\n",
       "      <td>78</td>\n",
       "      <td>83.916936</td>\n",
       "      <td>TCGTCCGTTACAAGCACC</td>\n",
       "      <td>55.237853</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.312299</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_5</td>\n",
       "      <td>CGCAGCGGTGGGTGCGTTAGATGAAAGCACCCTGCGCCGCTTAGCG...</td>\n",
       "      <td>78</td>\n",
       "      <td>84.447223</td>\n",
       "      <td>CGCAGCGGTGGGTGCGTT</td>\n",
       "      <td>64.535873</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.087284</td>\n",
       "      <td>-1.423561</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_6</td>\n",
       "      <td>CTTGCCGGGGTATAATCAAAGGTACGGCCCGCGCGATCAATGCGCA...</td>\n",
       "      <td>78</td>\n",
       "      <td>82.326074</td>\n",
       "      <td>GCTGTTAGTTGCGATTGG</td>\n",
       "      <td>52.094341</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.037659</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_7</td>\n",
       "      <td>TGATTATACCCCGGCAAGCCCGGGTCAAGCAGATCAGCGCGATCCG...</td>\n",
       "      <td>78</td>\n",
       "      <td>83.916936</td>\n",
       "      <td>TGATTATACCCCGGCAAG</td>\n",
       "      <td>51.513051</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.753178</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_8</td>\n",
       "      <td>TCTAACGGTGCTTGTGCTGCAACCGCTGGGCGCATACCATCGCGCA...</td>\n",
       "      <td>78</td>\n",
       "      <td>83.916936</td>\n",
       "      <td>GAGCGGTGGCGCAGATGC</td>\n",
       "      <td>62.457091</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.152218</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_9</td>\n",
       "      <td>AGCACAAGCACCGTTAGATACCGCGCGCCAGACTTTATGGGGCCAT...</td>\n",
       "      <td>78</td>\n",
       "      <td>82.326074</td>\n",
       "      <td>AGCACAAGCACCGTTAGA</td>\n",
       "      <td>54.362566</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.312299</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_10</td>\n",
       "      <td>CGCCACCATAAGCTCGGAGATGCTGCTGCATAGCGCGCAAATTCGC...</td>\n",
       "      <td>78</td>\n",
       "      <td>83.386649</td>\n",
       "      <td>GCTGGTGTTACATGCGCT</td>\n",
       "      <td>56.318877</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.833398</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_11</td>\n",
       "      <td>TCCGAGCTTATGGTGGCGCGATGGCGCGATTCTGGGCGAACGTGCA...</td>\n",
       "      <td>78</td>\n",
       "      <td>84.977510</td>\n",
       "      <td>TCCGAGCTTATGGTGGCG</td>\n",
       "      <td>57.418262</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.781808</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_12</td>\n",
       "      <td>GCTTTTAAGCTGCCGCGTGGAGACGCTGGTTCTGCAGAACCACGCC...</td>\n",
       "      <td>78</td>\n",
       "      <td>83.916936</td>\n",
       "      <td>TCTGCGCGGCAAACGCGC</td>\n",
       "      <td>65.750978</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-7.113216</td>\n",
       "      <td>-3.647187</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_13</td>\n",
       "      <td>ACGCGGCAGCTTAAAAGCGGAACCGGGCCAAGCAATGGCACGCCTG...</td>\n",
       "      <td>78</td>\n",
       "      <td>83.916936</td>\n",
       "      <td>ACGCGGCAGCTTAAAAGC</td>\n",
       "      <td>57.335961</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.242728</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_0</td>\n",
       "      <td>CTGCAGCTGATCCAGCGGCAGATCAAACAGCTGCATatgtatatct...</td>\n",
       "      <td>82</td>\n",
       "      <td>68.926776</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_1</td>\n",
       "      <td>CCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGA...</td>\n",
       "      <td>82</td>\n",
       "      <td>79.015166</td>\n",
       "      <td>CCGCTGGATCAGCTGCAG</td>\n",
       "      <td>58.833465</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.561997</td>\n",
       "      <td>0.182244</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_2</td>\n",
       "      <td>TTCACGCCATCTGCCGGATAATCCACCGGCTGCAGATCCGGTTCCG...</td>\n",
       "      <td>82</td>\n",
       "      <td>83.050522</td>\n",
       "      <td>GGAAACTGAGCCTGGAAG</td>\n",
       "      <td>52.810582</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.496558</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_3</td>\n",
       "      <td>TCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTT...</td>\n",
       "      <td>82</td>\n",
       "      <td>82.546103</td>\n",
       "      <td>TCCGGCAGATGGCGTGAA</td>\n",
       "      <td>59.618437</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.166426</td>\n",
       "      <td>-0.323592</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_4</td>\n",
       "      <td>CGCCATCATAGCTCGCGTTATAGCCATGATATTTCACAATCGCCGG...</td>\n",
       "      <td>82</td>\n",
       "      <td>81.537264</td>\n",
       "      <td>GGCTGGTATGCGGTGCCG</td>\n",
       "      <td>61.831203</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.564467</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_5</td>\n",
       "      <td>ACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCT...</td>\n",
       "      <td>82</td>\n",
       "      <td>82.041683</td>\n",
       "      <td>ACGCGAGCTATGATGGCG</td>\n",
       "      <td>58.272593</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-5.433718</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_6</td>\n",
       "      <td>TTGGTCATCCAGCCCAGCGCATGGCCATGCGGGCTAATGCTGGTAT...</td>\n",
       "      <td>81</td>\n",
       "      <td>83.769750</td>\n",
       "      <td>TGGCATGCTGGTGCGCGG</td>\n",
       "      <td>64.671249</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.233207</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_7</td>\n",
       "      <td>GCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTAT...</td>\n",
       "      <td>81</td>\n",
       "      <td>80.705869</td>\n",
       "      <td>GCTGGGCTGGATGACCAA</td>\n",
       "      <td>57.561188</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.303578</td>\n",
       "      <td>-0.101817</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_8</td>\n",
       "      <td>TGGCTGCCGCCGGTCACGCCAATGCGGGTTTCATCCACTTCATCAA...</td>\n",
       "      <td>81</td>\n",
       "      <td>83.769750</td>\n",
       "      <td>GGATGCGGTGCGCGCGCT</td>\n",
       "      <td>67.551371</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-9.894576</td>\n",
       "      <td>-0.175707</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_9</td>\n",
       "      <td>CGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCA...</td>\n",
       "      <td>81</td>\n",
       "      <td>84.791044</td>\n",
       "      <td>CGTGACCGGCGGCAGCCA</td>\n",
       "      <td>66.271975</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-6.467737</td>\n",
       "      <td>-0.702142</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_10</td>\n",
       "      <td>TCCAGATACGGCTGTTCCAGCGCCACATCAATCGCGCGTTCAAAGT...</td>\n",
       "      <td>81</td>\n",
       "      <td>81.727162</td>\n",
       "      <td>GAAAGCGGCGGTGGCGGA</td>\n",
       "      <td>63.977847</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.534007</td>\n",
       "      <td>0.327723</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_11</td>\n",
       "      <td>GGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGC...</td>\n",
       "      <td>81</td>\n",
       "      <td>80.705869</td>\n",
       "      <td>GGAACAGCCGTATCTGGA</td>\n",
       "      <td>53.896146</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-0.835849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_12</td>\n",
       "      <td>CCAATGCTCATCAGCACCGGCACTTTCACGCGATCCGCCAGGTTCA...</td>\n",
       "      <td>81</td>\n",
       "      <td>79.684575</td>\n",
       "      <td>GCAGGCGATGAAAACCCT</td>\n",
       "      <td>55.894536</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.037659</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_13</td>\n",
       "      <td>GGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGC...</td>\n",
       "      <td>81</td>\n",
       "      <td>80.705869</td>\n",
       "      <td>GGTGCTGATGAGCATTGG</td>\n",
       "      <td>54.040857</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.259944</td>\n",
       "      <td>-0.089766</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id oligo_id                                     oligo_sequence  \\\n",
       "0   Seq1   Seq1_0  ACCCAGCGCCGCCACCAGCAGGCTGGTGCGCATatgtatatctcct...   \n",
       "1   Seq1   Seq1_1  CTGGTGGCGGCGCTGGGTTTAGCATTAGCAGCAGCATTACCGGGCG...   \n",
       "2   Seq1   Seq1_2  AGATCCACTGCGCTAAAGCGATACGGCAGATCCTGGCGCTGCAGCA...   \n",
       "3   Seq1   Seq1_3  CTTTAGCGCAGTGGATCTGGATAGCGTGGATGGCCAGCGCCATTAT...   \n",
       "4   Seq1   Seq1_4  AACGCACCCACCGCTGCGTTGCCATCCAGCATCCACACCACCGGAT...   \n",
       "5   Seq1   Seq1_5  CGCAGCGGTGGGTGCGTTAGATGAAAGCACCCTGCGCCGCTTAGCG...   \n",
       "6   Seq1   Seq1_6  CTTGCCGGGGTATAATCAAAGGTACGGCCCGCGCGATCAATGCGCA...   \n",
       "7   Seq1   Seq1_7  TGATTATACCCCGGCAAGCCCGGGTCAAGCAGATCAGCGCGATCCG...   \n",
       "8   Seq1   Seq1_8  TCTAACGGTGCTTGTGCTGCAACCGCTGGGCGCATACCATCGCGCA...   \n",
       "9   Seq1   Seq1_9  AGCACAAGCACCGTTAGATACCGCGCGCCAGACTTTATGGGGCCAT...   \n",
       "10  Seq1  Seq1_10  CGCCACCATAAGCTCGGAGATGCTGCTGCATAGCGCGCAAATTCGC...   \n",
       "11  Seq1  Seq1_11  TCCGAGCTTATGGTGGCGCGATGGCGCGATTCTGGGCGAACGTGCA...   \n",
       "12  Seq1  Seq1_12  GCTTTTAAGCTGCCGCGTGGAGACGCTGGTTCTGCAGAACCACGCC...   \n",
       "13  Seq1  Seq1_13  ACGCGGCAGCTTAAAAGCGGAACCGGGCCAAGCAATGGCACGCCTG...   \n",
       "14  Seq2   Seq2_0  CTGCAGCTGATCCAGCGGCAGATCAAACAGCTGCATatgtatatct...   \n",
       "15  Seq2   Seq2_1  CCGCTGGATCAGCTGCAGACCTATAAACCGGAAAAAACCGCGCCGA...   \n",
       "16  Seq2   Seq2_2  TTCACGCCATCTGCCGGATAATCCACCGGCTGCAGATCCGGTTCCG...   \n",
       "17  Seq2   Seq2_3  TCCGGCAGATGGCGTGAAAGTGTATCGCCTGACCTATAAAAGCTTT...   \n",
       "18  Seq2   Seq2_4  CGCCATCATAGCTCGCGTTATAGCCATGATATTTCACAATCGCCGG...   \n",
       "19  Seq2   Seq2_5  ACGCGAGCTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGCT...   \n",
       "20  Seq2   Seq2_6  TTGGTCATCCAGCCCAGCGCATGGCCATGCGGGCTAATGCTGGTAT...   \n",
       "21  Seq2   Seq2_7  GCTGGGCTGGATGACCAAAGGCATTCTGGATAAAGATACCTATTAT...   \n",
       "22  Seq2   Seq2_8  TGGCTGCCGCCGGTCACGCCAATGCGGGTTTCATCCACTTCATCAA...   \n",
       "23  Seq2   Seq2_9  CGTGACCGGCGGCAGCCAAGGTGGTGGCTTAACTATTGCAGCAGCA...   \n",
       "24  Seq2  Seq2_10  TCCAGATACGGCTGTTCCAGCGCCACATCAATCGCGCGTTCAAAGT...   \n",
       "25  Seq2  Seq2_11  GGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCAACGGC...   \n",
       "26  Seq2  Seq2_12  CCAATGCTCATCAGCACCGGCACTTTCACGCGATCCGCCAGGTTCA...   \n",
       "27  Seq2  Seq2_13  GGTGCTGATGAGCATTGGCCTGATTGATAAAGTGACCCCGCCGAGC...   \n",
       "\n",
       "    oligo_length   oligo_tm primer_overlap_with_previous  overlap_tm_5prime  \\\n",
       "0             79  71.126314                         None                NaN   \n",
       "1             79  84.215682           CTGGTGGCGGCGCTGGGT          66.043739   \n",
       "2             79  82.644958           AGCCAGATCCAGAAGCGA          55.856688   \n",
       "3             78  82.856361           CTTTAGCGCAGTGGATCT          52.267941   \n",
       "4             78  83.916936           TCGTCCGTTACAAGCACC          55.237853   \n",
       "5             78  84.447223           CGCAGCGGTGGGTGCGTT          64.535873   \n",
       "6             78  82.326074           GCTGTTAGTTGCGATTGG          52.094341   \n",
       "7             78  83.916936           TGATTATACCCCGGCAAG          51.513051   \n",
       "8             78  83.916936           GAGCGGTGGCGCAGATGC          62.457091   \n",
       "9             78  82.326074           AGCACAAGCACCGTTAGA          54.362566   \n",
       "10            78  83.386649           GCTGGTGTTACATGCGCT          56.318877   \n",
       "11            78  84.977510           TCCGAGCTTATGGTGGCG          57.418262   \n",
       "12            78  83.916936           TCTGCGCGGCAAACGCGC          65.750978   \n",
       "13            78  83.916936           ACGCGGCAGCTTAAAAGC          57.335961   \n",
       "14            82  68.926776                         None                NaN   \n",
       "15            82  79.015166           CCGCTGGATCAGCTGCAG          58.833465   \n",
       "16            82  83.050522           GGAAACTGAGCCTGGAAG          52.810582   \n",
       "17            82  82.546103           TCCGGCAGATGGCGTGAA          59.618437   \n",
       "18            82  81.537264           GGCTGGTATGCGGTGCCG          61.831203   \n",
       "19            82  82.041683           ACGCGAGCTATGATGGCG          58.272593   \n",
       "20            81  83.769750           TGGCATGCTGGTGCGCGG          64.671249   \n",
       "21            81  80.705869           GCTGGGCTGGATGACCAA          57.561188   \n",
       "22            81  83.769750           GGATGCGGTGCGCGCGCT          67.551371   \n",
       "23            81  84.791044           CGTGACCGGCGGCAGCCA          66.271975   \n",
       "24            81  81.727162           GAAAGCGGCGGTGGCGGA          63.977847   \n",
       "25            81  80.705869           GGAACAGCCGTATCTGGA          53.896146   \n",
       "26            81  79.684575           GCAGGCGATGAAAACCCT          55.894536   \n",
       "27            81  80.705869           GGTGCTGATGAGCATTGG          54.040857   \n",
       "\n",
       "    overlap_length  overlap_homodimer_dg  overlap_hairpin_dg  \\\n",
       "0              NaN                   NaN                 NaN   \n",
       "1             18.0             -6.604577            0.000000   \n",
       "2             18.0             -2.037659            0.000000   \n",
       "3             18.0             -4.342888            0.000000   \n",
       "4             18.0             -1.312299            0.000000   \n",
       "5             18.0             -3.087284           -1.423561   \n",
       "6             18.0             -2.037659            0.000000   \n",
       "7             18.0             -2.753178            0.000000   \n",
       "8             18.0             -4.152218            0.000000   \n",
       "9             18.0             -1.312299            0.000000   \n",
       "10            18.0             -3.833398            0.000000   \n",
       "11            18.0             -1.781808            0.000000   \n",
       "12            18.0             -7.113216           -3.647187   \n",
       "13            18.0             -4.242728            0.000000   \n",
       "14             NaN                   NaN                 NaN   \n",
       "15            18.0             -3.561997            0.182244   \n",
       "16            18.0             -2.496558            0.000000   \n",
       "17            18.0             -3.166426           -0.323592   \n",
       "18            18.0             -1.564467            0.000000   \n",
       "19            18.0             -5.433718            0.000000   \n",
       "20            18.0             -4.233207            0.000000   \n",
       "21            18.0             -1.303578           -0.101817   \n",
       "22            18.0             -9.894576           -0.175707   \n",
       "23            18.0             -6.467737           -0.702142   \n",
       "24            18.0             -4.534007            0.327723   \n",
       "25            18.0             -0.835849            0.000000   \n",
       "26            18.0             -2.037659            0.000000   \n",
       "27            18.0             -3.259944           -0.089766   \n",
       "\n",
       "                                    original_sequence  \n",
       "0   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "1   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "2   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "3   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "4   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "5   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "6   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "7   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "8   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "9   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "10  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "11  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "12  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "13  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "14  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "15  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "16  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "17  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "18  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "19  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "20  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "21  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "22  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "23  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "24  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "25  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "26  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "27  cccctctagaaataattttgtttaactttaagaaggagatatacat...  "
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import pandas as pd\n",
    "import random\n",
    "import numpy \n",
    "  \n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['id'] = ['Seq1', 'Seq2']\n",
    "min_gc = 0.25\n",
    "max_gc = 0.65\n",
    "min_tm = 10\n",
    "max_tm = 100\n",
    "min_segment_length = 40\n",
    "max_segment_length = 100\n",
    "max_length = 500\n",
    "\n",
    "SEED = i\n",
    "random.seed(SEED)\n",
    "numpy.random.seed(SEED)\n",
    "df['seqs'] = ['MRTSLLVAALGLALAAALPGGAPLAQPDPEATMDRSLLQRQDLPYRFSAVDLDSVDGQRHYRLWLGRPLQAPPAAGYPVVWMLDGNAAVGALDESTLRRLADGDAPLLVAIGYRTPLRIDRAGRTFDYTPASPGQADQRDPLNGLPSGGADAFLDLLRDGMRPAVAAQAPLDTARQTLWGHSYGGLLVLHALFTRPGEFARYAAASPSLWWRDGAILGERAGLEQRLRGKRAELLLWRGSAEPASPRGSLKAEPGQAMARLVDDLRRVAGLTLDFQPLDGLGHGETLGASLRLLLARPAVERQR', \n",
    "             'MQLFDLPLDQLQTYKPEKTAPKDFSEFWKLSLEELAKVQAEPDLQPVDYPADGVKVYRLTYKSFGNARITGWYAVPDKEGPHPAIVKYHGYNASYDGEIHEMVNWALHGYATFGMLVRGQQSSEDTSISPHGHALGWMTKGILDKDTYYYRGVYLDAVRALEVISSFDEVDETRIGVTGGSQGGGLTIAAAALSDIPKAAVADYPYLSNFERAIDVALEQPYLEINSFFRRNGSPETEVQAMKTLSYFDIMNLADRVKVPVLMSIGLIDKVTPPSTVFAAYNHLETKKELKVYRYFGHEYIPAFQTEKLAFFKQHLKG']\n",
    "oligo2_df = get_oligos(df, 'seqs', 'id', '.', 'cccctctagaaataattttgtttaactttaagaaggagatatacat', 'gatccggctgctaacaaag', sequence_end='CTCGAGCACCACCACCACCACCACTGA',\n",
    "                     min_gc=min_gc, max_gc=max_gc, min_tm=min_tm, max_tm=max_tm, min_segment_length=min_segment_length, max_segment_length=max_segment_length, \n",
    "                     genbank_file=\"base-pet22b-base-anm.gb\", insert_position=5193, simple=True)\n",
    "oligo2_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "998f7ab7-d1a7-43c3-a445-d6bceaee2605",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>oligo_id</th>\n",
       "      <th>oligo_sequence</th>\n",
       "      <th>oligo_length</th>\n",
       "      <th>oligo_tm</th>\n",
       "      <th>primer_overlap_with_previous</th>\n",
       "      <th>overlap_tm_5prime</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>overlap_homodimer_dg</th>\n",
       "      <th>overlap_hairpin_dg</th>\n",
       "      <th>original_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_0</td>\n",
       "      <td>CTGCTAATGCTAAACCCAGCGCCGCCACCAGCAGGCTGGTGCGCAT...</td>\n",
       "      <td>92</td>\n",
       "      <td>72.671867</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_1</td>\n",
       "      <td>TGGGTTTAGCATTAGCAGCAGCATTACCGGGCGGTGCACCGTTAGC...</td>\n",
       "      <td>92</td>\n",
       "      <td>84.361241</td>\n",
       "      <td>TGGGTTTAGCATTAGCAG</td>\n",
       "      <td>49.684873</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.312299</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_2</td>\n",
       "      <td>AGCCACAGGCGATAATGGCGCTGGCCATCCACGCTATCCAGATCCA...</td>\n",
       "      <td>92</td>\n",
       "      <td>85.260424</td>\n",
       "      <td>CGCAGCCTGCTGCAGCGC</td>\n",
       "      <td>65.908190</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-5.560014</td>\n",
       "      <td>-0.680867</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_3</td>\n",
       "      <td>CCATTATCGCCTGTGGCTGGGTCGTCCGTTACAAGCACCACCGGCA...</td>\n",
       "      <td>91</td>\n",
       "      <td>85.015837</td>\n",
       "      <td>CCATTATCGCCTGTGGCT</td>\n",
       "      <td>55.011907</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-0.564170</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_4</td>\n",
       "      <td>GATAACCAATCGCAACTAACAGCGGCGCATCGCCATCCGCTAAGCG...</td>\n",
       "      <td>91</td>\n",
       "      <td>84.106773</td>\n",
       "      <td>GATGGCAACGCAGCGGTG</td>\n",
       "      <td>60.403112</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-5.846544</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_5</td>\n",
       "      <td>TAGTTGCGATTGGTTATCGCACCCCGCTGCGCATTGATCGCGCGGG...</td>\n",
       "      <td>91</td>\n",
       "      <td>83.652241</td>\n",
       "      <td>TAGTTGCGATTGGTTATC</td>\n",
       "      <td>47.172552</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.037659</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_6</td>\n",
       "      <td>GCGCATACCATCGCGCAGCAGATCCAGAAACGCATCTGCGCCACCG...</td>\n",
       "      <td>91</td>\n",
       "      <td>85.470369</td>\n",
       "      <td>CCCGGGTCAAGCAGATCA</td>\n",
       "      <td>56.927827</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.593417</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_7</td>\n",
       "      <td>CTGCGCGATGGTATGCGCCCAGCGGTTGCAGCACAAGCACCGTTAG...</td>\n",
       "      <td>91</td>\n",
       "      <td>85.470369</td>\n",
       "      <td>CTGCGCGATGGTATGCGC</td>\n",
       "      <td>60.297627</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.434948</td>\n",
       "      <td>-1.493750</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_8</td>\n",
       "      <td>CACCATAAGCTCGGAGATGCTGCTGCATAGCGCGCAAATTCGCCCG...</td>\n",
       "      <td>91</td>\n",
       "      <td>83.652241</td>\n",
       "      <td>ATAGCTATGGCGGCCTGC</td>\n",
       "      <td>57.568469</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.420447</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_9</td>\n",
       "      <td>ATCTCCGAGCTTATGGTGGCGCGATGGCGCGATTCTGGGCGAACGT...</td>\n",
       "      <td>91</td>\n",
       "      <td>85.015837</td>\n",
       "      <td>ATCTCCGAGCTTATGGTG</td>\n",
       "      <td>51.243452</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.781808</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>Seq1_10</td>\n",
       "      <td>GTGCCATTGCTTGGCCCGGTTCCGCTTTTAAGCTGCCGCGTGGAGA...</td>\n",
       "      <td>91</td>\n",
       "      <td>84.561305</td>\n",
       "      <td>AAACGCGCAGAACTGCTG</td>\n",
       "      <td>57.263676</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.078706</td>\n",
       "      <td>0.155873</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_0</td>\n",
       "      <td>TCCGGTTTATAGGTCTGCAGCTGATCCAGCGGCAGATCAAACAGCT...</td>\n",
       "      <td>96</td>\n",
       "      <td>71.107775</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_1</td>\n",
       "      <td>GCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAA...</td>\n",
       "      <td>95</td>\n",
       "      <td>80.738566</td>\n",
       "      <td>GCAGACCTATAAACCGGA</td>\n",
       "      <td>51.155715</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-4.905888</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_2</td>\n",
       "      <td>GCGCGCGTTGCCAAAGCTTTTATAGGTCAGGCGATACACTTTCACG...</td>\n",
       "      <td>95</td>\n",
       "      <td>83.350928</td>\n",
       "      <td>AAGTGCAGGCGGAACCGG</td>\n",
       "      <td>60.945871</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-2.013587</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_3</td>\n",
       "      <td>AGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATA...</td>\n",
       "      <td>95</td>\n",
       "      <td>82.915535</td>\n",
       "      <td>AGCTTTGGCAACGCGCGC</td>\n",
       "      <td>63.520112</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-7.158247</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_4</td>\n",
       "      <td>GACCGCGCACCAGCATGCCAAAGGTCGCATAGCCATGCAGCGCCCA...</td>\n",
       "      <td>95</td>\n",
       "      <td>82.044747</td>\n",
       "      <td>TCATGGCTATAACGCGAG</td>\n",
       "      <td>52.176018</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-5.433718</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_5</td>\n",
       "      <td>GCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAG...</td>\n",
       "      <td>95</td>\n",
       "      <td>83.350928</td>\n",
       "      <td>GCATGCTGGTGCGCGGTC</td>\n",
       "      <td>63.039125</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-3.586877</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_6</td>\n",
       "      <td>TCCACTTCATCAAAGCTGCTAATCACTTCCAGCGCGCGCACCGCAT...</td>\n",
       "      <td>95</td>\n",
       "      <td>79.432385</td>\n",
       "      <td>AAAGGCATTCTGGATAAA</td>\n",
       "      <td>46.185257</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-0.576037</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_7</td>\n",
       "      <td>CAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGC...</td>\n",
       "      <td>95</td>\n",
       "      <td>82.915535</td>\n",
       "      <td>CAGCTTTGATGAAGTGGA</td>\n",
       "      <td>49.970528</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.625988</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_8</td>\n",
       "      <td>ATACGGCTGTTCCAGCGCCACATCAATCGCGCGTTCAAAGTTGCTC...</td>\n",
       "      <td>95</td>\n",
       "      <td>82.915535</td>\n",
       "      <td>CAGCGCTGAGCGATATTC</td>\n",
       "      <td>54.084460</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-7.221216</td>\n",
       "      <td>0.125144</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_9</td>\n",
       "      <td>GCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCA...</td>\n",
       "      <td>95</td>\n",
       "      <td>81.609354</td>\n",
       "      <td>GCGCTGGAACAGCCGTAT</td>\n",
       "      <td>58.429067</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.504889</td>\n",
       "      <td>-1.822581</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>Seq2_10</td>\n",
       "      <td>GCGGGGTCACTTTATCAATCAGGCCAATGCTCATCAGCACCGGCAC...</td>\n",
       "      <td>95</td>\n",
       "      <td>80.303172</td>\n",
       "      <td>GAAAACCCTGAGCTATTT</td>\n",
       "      <td>47.131980</td>\n",
       "      <td>18.0</td>\n",
       "      <td>-1.928938</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id oligo_id                                     oligo_sequence  \\\n",
       "0   Seq1   Seq1_0  CTGCTAATGCTAAACCCAGCGCCGCCACCAGCAGGCTGGTGCGCAT...   \n",
       "1   Seq1   Seq1_1  TGGGTTTAGCATTAGCAGCAGCATTACCGGGCGGTGCACCGTTAGC...   \n",
       "2   Seq1   Seq1_2  AGCCACAGGCGATAATGGCGCTGGCCATCCACGCTATCCAGATCCA...   \n",
       "3   Seq1   Seq1_3  CCATTATCGCCTGTGGCTGGGTCGTCCGTTACAAGCACCACCGGCA...   \n",
       "4   Seq1   Seq1_4  GATAACCAATCGCAACTAACAGCGGCGCATCGCCATCCGCTAAGCG...   \n",
       "5   Seq1   Seq1_5  TAGTTGCGATTGGTTATCGCACCCCGCTGCGCATTGATCGCGCGGG...   \n",
       "6   Seq1   Seq1_6  GCGCATACCATCGCGCAGCAGATCCAGAAACGCATCTGCGCCACCG...   \n",
       "7   Seq1   Seq1_7  CTGCGCGATGGTATGCGCCCAGCGGTTGCAGCACAAGCACCGTTAG...   \n",
       "8   Seq1   Seq1_8  CACCATAAGCTCGGAGATGCTGCTGCATAGCGCGCAAATTCGCCCG...   \n",
       "9   Seq1   Seq1_9  ATCTCCGAGCTTATGGTGGCGCGATGGCGCGATTCTGGGCGAACGT...   \n",
       "10  Seq1  Seq1_10  GTGCCATTGCTTGGCCCGGTTCCGCTTTTAAGCTGCCGCGTGGAGA...   \n",
       "11  Seq2   Seq2_0  TCCGGTTTATAGGTCTGCAGCTGATCCAGCGGCAGATCAAACAGCT...   \n",
       "12  Seq2   Seq2_1  GCAGACCTATAAACCGGAAAAAACCGCGCCGAAAGATTTTAGCGAA...   \n",
       "13  Seq2   Seq2_2  GCGCGCGTTGCCAAAGCTTTTATAGGTCAGGCGATACACTTTCACG...   \n",
       "14  Seq2   Seq2_3  AGCTTTGGCAACGCGCGCATTACCGGCTGGTATGCGGTGCCGGATA...   \n",
       "15  Seq2   Seq2_4  GACCGCGCACCAGCATGCCAAAGGTCGCATAGCCATGCAGCGCCCA...   \n",
       "16  Seq2   Seq2_5  GCATGCTGGTGCGCGGTCAACAGAGCAGCGAAGATACCAGCATTAG...   \n",
       "17  Seq2   Seq2_6  TCCACTTCATCAAAGCTGCTAATCACTTCCAGCGCGCGCACCGCAT...   \n",
       "18  Seq2   Seq2_7  CAGCTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGC...   \n",
       "19  Seq2   Seq2_8  ATACGGCTGTTCCAGCGCCACATCAATCGCGCGTTCAAAGTTGCTC...   \n",
       "20  Seq2   Seq2_9  GCGCTGGAACAGCCGTATCTGGAAATTAACAGCTTTTTTCGCCGCA...   \n",
       "21  Seq2  Seq2_10  GCGGGGTCACTTTATCAATCAGGCCAATGCTCATCAGCACCGGCAC...   \n",
       "\n",
       "    oligo_length   oligo_tm primer_overlap_with_previous  overlap_tm_5prime  \\\n",
       "0             92  72.671867                         None                NaN   \n",
       "1             92  84.361241           TGGGTTTAGCATTAGCAG          49.684873   \n",
       "2             92  85.260424           CGCAGCCTGCTGCAGCGC          65.908190   \n",
       "3             91  85.015837           CCATTATCGCCTGTGGCT          55.011907   \n",
       "4             91  84.106773           GATGGCAACGCAGCGGTG          60.403112   \n",
       "5             91  83.652241           TAGTTGCGATTGGTTATC          47.172552   \n",
       "6             91  85.470369           CCCGGGTCAAGCAGATCA          56.927827   \n",
       "7             91  85.470369           CTGCGCGATGGTATGCGC          60.297627   \n",
       "8             91  83.652241           ATAGCTATGGCGGCCTGC          57.568469   \n",
       "9             91  85.015837           ATCTCCGAGCTTATGGTG          51.243452   \n",
       "10            91  84.561305           AAACGCGCAGAACTGCTG          57.263676   \n",
       "11            96  71.107775                         None                NaN   \n",
       "12            95  80.738566           GCAGACCTATAAACCGGA          51.155715   \n",
       "13            95  83.350928           AAGTGCAGGCGGAACCGG          60.945871   \n",
       "14            95  82.915535           AGCTTTGGCAACGCGCGC          63.520112   \n",
       "15            95  82.044747           TCATGGCTATAACGCGAG          52.176018   \n",
       "16            95  83.350928           GCATGCTGGTGCGCGGTC          63.039125   \n",
       "17            95  79.432385           AAAGGCATTCTGGATAAA          46.185257   \n",
       "18            95  82.915535           CAGCTTTGATGAAGTGGA          49.970528   \n",
       "19            95  82.915535           CAGCGCTGAGCGATATTC          54.084460   \n",
       "20            95  81.609354           GCGCTGGAACAGCCGTAT          58.429067   \n",
       "21            95  80.303172           GAAAACCCTGAGCTATTT          47.131980   \n",
       "\n",
       "    overlap_length  overlap_homodimer_dg  overlap_hairpin_dg  \\\n",
       "0              NaN                   NaN                 NaN   \n",
       "1             18.0             -1.312299            0.000000   \n",
       "2             18.0             -5.560014           -0.680867   \n",
       "3             18.0             -0.564170            0.000000   \n",
       "4             18.0             -5.846544            0.000000   \n",
       "5             18.0             -2.037659            0.000000   \n",
       "6             18.0             -4.593417            0.000000   \n",
       "7             18.0             -3.434948           -1.493750   \n",
       "8             18.0             -4.420447            0.000000   \n",
       "9             18.0             -1.781808            0.000000   \n",
       "10            18.0             -2.078706            0.155873   \n",
       "11             NaN                   NaN                 NaN   \n",
       "12            18.0             -4.905888            0.000000   \n",
       "13            18.0             -2.013587            0.000000   \n",
       "14            18.0             -7.158247            0.000000   \n",
       "15            18.0             -5.433718            0.000000   \n",
       "16            18.0             -3.586877            0.000000   \n",
       "17            18.0             -0.576037            0.000000   \n",
       "18            18.0             -1.625988            0.000000   \n",
       "19            18.0             -7.221216            0.125144   \n",
       "20            18.0             -1.504889           -1.822581   \n",
       "21            18.0             -1.928938            0.000000   \n",
       "\n",
       "                                    original_sequence  \n",
       "0   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "1   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "2   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "3   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "4   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "5   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "6   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "7   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "8   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "9   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "10  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "11  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "12  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "13  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "14  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "15  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "16  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "17  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "18  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "19  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "20  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "21  cccctctagaaataattttgtttaactttaagaaggagatatacat...  "
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oligo2_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "b6a9107d-3199-476c-aa3d-b803c7b7fc9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "objective:   0%|                                                  | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-912](e_coli...]\n",
      "location:   0%|                                                                         | 0/238 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                          | 0/238 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-912](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-912](mini:0.25, maxi:0.75, window:50)\n",
      "      │ Passed !\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:         0\n",
      "✔        0 ┍ MaximizeCAI[0-912](e_coli) \n",
      "           │ Codon opt. on window 0-912 scored -0.00E+00\n",
      "\n",
      "\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m              Warning: Seq1 does not end with a stop codon. \tCGC\t               \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[93m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[93mWarning: Seq1 does not end with a stop codon. AND you don't have a stop codon in your primer!!\tCTCGAGCACCACCACCACCACCACTGAgatccggctgctaacaaag\t\u001b[0m\n",
      "\u001b[93m--------------------------------------------------------------------------------\u001b[0m\n",
      "We expect the primer to be in 5 to 3 prime direction.\n",
      "ATGCGCACCTATTTTTTTGTGGCGGCGTTTGGCTTTGCGTTTGCGGCGGCGTTTCCGGGCGGCGCGCCGTTTGCGCAGCATGATCCGGAAGCGACCATGGATCGCTATTTTTTTCAGCGCCAGGATTTTCCGTATCGCTTTTATGCGGTGGATTTTGATTATGTGGATGGCCAGCGCCATTATCGCTTTTGGTTTGGCCGCCCGTTTCAGGCGCCGCCGGCGGCGGGCTATCCGGTGGATTGGATGTTTGATGGCAACGCGGCGGTGGGCGCGTTTGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTTTTTGTGGCGATTGGCTATCGCACCCCGTTTCGCATTGATCGCGCGGGCCGCACCTTTGATTATACCCCGGCGTATCCGGGCCAGGCGGATCAGCGCGATCCGTTTAACGGCTTTCCGTATGGCGGCGCGGATGCGTTTTTTGATTTTTTTCGCGATGGCATGCGCCCGGCGGTGGCGGCGCAGGCGCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCCATTATTATGGCGGCTTTTTTGTGTTTCATGCGTTTTTTACCCGCCCGGGCGAATTTGCGCGCTATGCGGCGGCGTATCCGTATTTTTGGTGGCGCGATGGCGCGATTTTTGGCGAACGCGCGGGCTTTGAACAGCGCTTTCGCGGCAAACGCGCGGAATTTTTTTTTTGGCGCGGCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACCGGGCCAGGCGATGGCGCGCTTTGTGGATGATTTTCGCCGCGTGGCGGGCTTTACCTTTGATTTTCAGCCGTTTGATGGCTTTGGCCATGGCGAAACCTTTGGCGCGTATTTTCGCTTTTTTTTTGCGCGCCCGGCGGTGGAACGCCAGCGC\n",
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n",
      "constraint:   0%|                                                 | 0/2 [00:00<?, ?it/s, now=AvoidPattern[0-954](patte...]\n",
      "location:   0%|                                                                           | 0/8 [00:00<?, ?it/s, now=None]\u001b[A\n",
      "location:   0%|                                                                           | 0/8 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "objective:   0%|                                                  | 0/1 [00:00<?, ?it/s, now=MaximizeCAI[0-954](e_coli...]\u001b[A\n",
      "location:   0%|                                                                         | 0/194 [00:00<?, ?it/s, now=0-99]\u001b[A\n",
      "location:   0%|                                                                          | 0/194 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\n",
      "location:   0%|                                                                          | 0/194 [00:00<?, ?it/s, now=3-6]\u001b[A\n",
      "                                                                                                                          \u001b[A\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===> SUCCESS - all constraints evaluations pass\n",
      "✔PASS ┍ AvoidPattern[0-954](pattern:BsaI(GGTCTC))\n",
      "      │ Passed. Pattern not found !\n",
      "✔PASS ┍ EnforceGCContent[0-954](mini:0.25, maxi:0.75, window:50)\n",
      "      │ Passed !\n",
      "\n",
      "\n",
      "===> TOTAL OBJECTIVES SCORE:         0\n",
      "✔        0 ┍ MaximizeCAI[0-954](e_coli) \n",
      "           │ Codon opt. on window 0-954 scored -0.00E+00\n",
      "\n",
      "\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[94m              Warning: Seq2 does not end with a stop codon. \tGGC\t               \u001b[0m\n",
      "\u001b[94m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[93m--------------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[93mWarning: Seq2 does not end with a stop codon. AND you don't have a stop codon in your primer!!\tCTCGAGCACCACCACCACCACCACTGAgatccggctgctaacaaag\t\u001b[0m\n",
      "\u001b[93m--------------------------------------------------------------------------------\u001b[0m\n",
      "We expect the primer to be in 5 to 3 prime direction.\n",
      "ATGCAGTTTTTTGATTTTCCGTTTGATCCGCTGCAGACCTATAAACCGGAAGAAACCGCGCCGAAAGAATTTTATGAATTTGGCAAATGGTATTGGGAAGAATTTGCGAAAGTGCAGGCGGAACCGGATTTTCAGCCGGTGGATTATCCGGCGGATGGCGTGAAAGTGTATCGCGTGACCTATAAATATTTTGGCAACGCGCGCACCACCGGCTGGTATGCGGTGCCGGATAAAGAAGGCCCGCATCCGGCGATTGTGAAATATCATGGCTATAACGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTTCATGGCTATGCGACCTTTGGCATGTTTGTGCGCGGCCAGCAGTATTATGAAGATACCTATATTTATCCGCATGGCCATGCGTTTGGCTGGATGACCAAAGGCATTTTTGGCAAAGATACCTATTATTATCGCGGCGTGTATTTTGATGCGGTGCGCGCGTTTGAAGTGATTTATTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGCTATCAGGGCGGCGGCTTTACCATTGCGGCGGCGGCGTTTTATGATATTCCGAAAGCGGCGGTGGCGGATTATCCGTATTTTTATAACTTTGAACGCGCGATTGATGCGGCGTTTGAACAGCCGTATTTTGAAATTAACTATTTTGTGCGCCGCAACGGCTATCCGGAAAACGAAGTGCAGGCGATGCAGACCTTTTATTATTTTGAAATTATGAACTTTGCGGATTGGGTGGAAGTGCCGGTGTTTATGTGGATTGGCTTTATTGATAAAGTGACCCCGCCGTGGATTGTGTTTGCGGCGTATAACCATTTTGAAACCAGCAAAGGCTTTATGGTGTATCGCTATTATGGCCATGAATATATTCCGGATTTTCAGACCGATAAATTTCCGTTTTTTAAACCGCATTTTAAAGGC\n",
      "Looking for the best assembly plan...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:489: UserWarning: Function deprecated please use \"calc_tm\" instead\n",
      "  return THERMO_ANALYSIS.calcTm(seq)\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:135: UserWarning: Function deprecated please use \"calc_hairpin\" instead\n",
      "  return THERMO_ANALYSIS.calcHairpin(seq, output_structure).check_exc()\n",
      "/Users/arianemora/miniconda3/envs/oligo/lib/python3.10/site-packages/primer3/bindings.py:217: UserWarning: Function deprecated please use \"calc_homodimer\" instead\n",
      "  return THERMO_ANALYSIS.calcHomodimer(seq, output_structure).check_exc()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>oligo_id</th>\n",
       "      <th>oligo_sequence</th>\n",
       "      <th>oligo_length</th>\n",
       "      <th>oligo_tm</th>\n",
       "      <th>primer_overlap_with_previous</th>\n",
       "      <th>overlap_tm_5prime</th>\n",
       "      <th>overlap_length</th>\n",
       "      <th>overlap_homodimer_tm</th>\n",
       "      <th>overlap_hairpin_tm</th>\n",
       "      <th>original_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00000</td>\n",
       "      <td>CAAACGGCGCGCCGCCCGGAAACGCCGCCGCAAACGCAAAGCCAAA...</td>\n",
       "      <td>119</td>\n",
       "      <td>76.754686</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00001</td>\n",
       "      <td>CGGGCGGCGCGCCGTTTGCGCAGCATGATCCGGAAGCGACCATGGA...</td>\n",
       "      <td>128</td>\n",
       "      <td>83.441734</td>\n",
       "      <td>CGGGCGGCGCGCCGTTTG</td>\n",
       "      <td>68.658228</td>\n",
       "      <td>18.0</td>\n",
       "      <td>56.011620</td>\n",
       "      <td>68.396815</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00002</td>\n",
       "      <td>GCGAAAGGTATATTCATCAAACGCGCCCACCGCCGCGTTGCCATCA...</td>\n",
       "      <td>129</td>\n",
       "      <td>86.834758</td>\n",
       "      <td>GATGGCCAGCGCCATTAT</td>\n",
       "      <td>55.811457</td>\n",
       "      <td>18.0</td>\n",
       "      <td>17.815718</td>\n",
       "      <td>68.998303</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00003</td>\n",
       "      <td>TGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.380596</td>\n",
       "      <td>TGATGAATATACCTTTCGC</td>\n",
       "      <td>47.775584</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-64.316560</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00004</td>\n",
       "      <td>AAACGGCGCCTGCGCCGCCACCGCCGGGCGCATGCCATCGCGAAAA...</td>\n",
       "      <td>128</td>\n",
       "      <td>88.288890</td>\n",
       "      <td>CCCCGGCGTATCCGGGCC</td>\n",
       "      <td>65.660649</td>\n",
       "      <td>18.0</td>\n",
       "      <td>30.462576</td>\n",
       "      <td>75.699579</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00005</td>\n",
       "      <td>GCGGCGCAGGCGCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCC...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.057453</td>\n",
       "      <td>GCGGCGCAGGCGCCGTTT</td>\n",
       "      <td>68.094164</td>\n",
       "      <td>18.0</td>\n",
       "      <td>53.361828</td>\n",
       "      <td>83.411766</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00006</td>\n",
       "      <td>ACGCCGGTTCCGCATAGCCGCGCCAAAAAAAAAATTCCGCGCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>86.026884</td>\n",
       "      <td>TGCGGCGGCGTATCCGTA</td>\n",
       "      <td>62.137239</td>\n",
       "      <td>18.0</td>\n",
       "      <td>32.349272</td>\n",
       "      <td>58.411563</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00007</td>\n",
       "      <td>GCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACC...</td>\n",
       "      <td>128</td>\n",
       "      <td>85.057453</td>\n",
       "      <td>GCTATGCGGAACCGGCGT</td>\n",
       "      <td>61.126936</td>\n",
       "      <td>18.0</td>\n",
       "      <td>15.010537</td>\n",
       "      <td>40.188310</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00008</td>\n",
       "      <td>TTCCACCGCCGGGCGCGCAAAAAAAAAGCGAAAATACGCGCCAAAG...</td>\n",
       "      <td>78</td>\n",
       "      <td>81.265500</td>\n",
       "      <td>CAGCCGTTTGATGGCTTT</td>\n",
       "      <td>54.166830</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10.864156</td>\n",
       "      <td>58.661393</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Seq1</td>\n",
       "      <td>S_00009</td>\n",
       "      <td>GCGCGCCCGGCGGTGGAACGCCAGCGCCTCGAGCACCACCACCACC...</td>\n",
       "      <td>73</td>\n",
       "      <td>80.600609</td>\n",
       "      <td>GCGCGCCCGGCGGTGGAA</td>\n",
       "      <td>69.523553</td>\n",
       "      <td>18.0</td>\n",
       "      <td>31.975298</td>\n",
       "      <td>63.355769</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00000</td>\n",
       "      <td>TAAAATTCTTTCGGCGCGGTTTCTTCCGGTTTATAGGTCTGCAGCG...</td>\n",
       "      <td>120</td>\n",
       "      <td>71.840745</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00001</td>\n",
       "      <td>CCGCGCCGAAAGAATTTTATGAATTTGGCAAATGGTATTGGGAAGA...</td>\n",
       "      <td>129</td>\n",
       "      <td>81.704538</td>\n",
       "      <td>CCGCGCCGAAAGAATTTTA</td>\n",
       "      <td>54.887527</td>\n",
       "      <td>19.0</td>\n",
       "      <td>-2.987647</td>\n",
       "      <td>39.271088</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00002</td>\n",
       "      <td>TTCGCCATCATAATACGCGTTATAGCCATGATATTTCACAATCGCC...</td>\n",
       "      <td>129</td>\n",
       "      <td>82.666454</td>\n",
       "      <td>GTGTATCGCGTGACCTATA</td>\n",
       "      <td>51.807772</td>\n",
       "      <td>19.0</td>\n",
       "      <td>5.768584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00003</td>\n",
       "      <td>CGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTT...</td>\n",
       "      <td>128</td>\n",
       "      <td>80.856584</td>\n",
       "      <td>CGCGTATTATGATGGCGAA</td>\n",
       "      <td>53.868963</td>\n",
       "      <td>19.0</td>\n",
       "      <td>10.800809</td>\n",
       "      <td>49.069751</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00004</td>\n",
       "      <td>TCATCCACTTCATCAAAATAATAAATCACTTCAAACGCGCGCACCG...</td>\n",
       "      <td>130</td>\n",
       "      <td>79.675896</td>\n",
       "      <td>ATCCGCATGGCCATGCGT</td>\n",
       "      <td>60.798906</td>\n",
       "      <td>18.0</td>\n",
       "      <td>57.093991</td>\n",
       "      <td>67.117879</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00005</td>\n",
       "      <td>TTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGC...</td>\n",
       "      <td>131</td>\n",
       "      <td>83.045865</td>\n",
       "      <td>TTATTTTGATGAAGTGGATGA</td>\n",
       "      <td>48.512171</td>\n",
       "      <td>21.0</td>\n",
       "      <td>-138.759018</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00006</td>\n",
       "      <td>AGGTCTGCATCGCCTGCACTTCGTTTTCCGGATAGCCGTTGCGGCG...</td>\n",
       "      <td>130</td>\n",
       "      <td>80.630413</td>\n",
       "      <td>GCGGATTATCCGTATTTTTATA</td>\n",
       "      <td>49.832919</td>\n",
       "      <td>22.0</td>\n",
       "      <td>30.590793</td>\n",
       "      <td>59.177368</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00007</td>\n",
       "      <td>TGCAGGCGATGCAGACCTTTTATTATTTTGAAATTATGAACTTTGC...</td>\n",
       "      <td>129</td>\n",
       "      <td>79.780705</td>\n",
       "      <td>TGCAGGCGATGCAGACCT</td>\n",
       "      <td>59.627099</td>\n",
       "      <td>18.0</td>\n",
       "      <td>10.850945</td>\n",
       "      <td>59.622681</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00008</td>\n",
       "      <td>GCCTTTAAAATGCGGTTTAAAAAACGGAAATTTATCGGTCTGAAAA...</td>\n",
       "      <td>129</td>\n",
       "      <td>76.894956</td>\n",
       "      <td>ATTGTGTTTGCGGCGTATA</td>\n",
       "      <td>53.466680</td>\n",
       "      <td>19.0</td>\n",
       "      <td>9.367922</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Seq2</td>\n",
       "      <td>S_00009</td>\n",
       "      <td>TAAACCGCATTTTAAAGGCCTCGAGCACCACCACCACCACCACTGA...</td>\n",
       "      <td>65</td>\n",
       "      <td>72.196960</td>\n",
       "      <td>TAAACCGCATTTTAAAGGC</td>\n",
       "      <td>49.857060</td>\n",
       "      <td>19.0</td>\n",
       "      <td>4.109652</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>cccctctagaaataattttgtttaactttaagaaggagatatacat...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      id oligo_id                                     oligo_sequence  \\\n",
       "0   Seq1  S_00000  CAAACGGCGCGCCGCCCGGAAACGCCGCCGCAAACGCAAAGCCAAA...   \n",
       "1   Seq1  S_00001  CGGGCGGCGCGCCGTTTGCGCAGCATGATCCGGAAGCGACCATGGA...   \n",
       "2   Seq1  S_00002  GCGAAAGGTATATTCATCAAACGCGCCCACCGCCGCGTTGCCATCA...   \n",
       "3   Seq1  S_00003  TGATGAATATACCTTTCGCCGCTTTGCGGATGGCGATGCGCCGTTT...   \n",
       "4   Seq1  S_00004  AAACGGCGCCTGCGCCGCCACCGCCGGGCGCATGCCATCGCGAAAA...   \n",
       "5   Seq1  S_00005  GCGGCGCAGGCGCCGTTTGATAACGCGCTGCAGACCTTTTGGGGCC...   \n",
       "6   Seq1  S_00006  ACGCCGGTTCCGCATAGCCGCGCCAAAAAAAAAATTCCGCGCGTTT...   \n",
       "7   Seq1  S_00007  GCTATGCGGAACCGGCGTATCCGCGCGGCTATTTTAAAGCGGAACC...   \n",
       "8   Seq1  S_00008  TTCCACCGCCGGGCGCGCAAAAAAAAAGCGAAAATACGCGCCAAAG...   \n",
       "9   Seq1  S_00009  GCGCGCCCGGCGGTGGAACGCCAGCGCCTCGAGCACCACCACCACC...   \n",
       "10  Seq2  S_00000  TAAAATTCTTTCGGCGCGGTTTCTTCCGGTTTATAGGTCTGCAGCG...   \n",
       "11  Seq2  S_00001  CCGCGCCGAAAGAATTTTATGAATTTGGCAAATGGTATTGGGAAGA...   \n",
       "12  Seq2  S_00002  TTCGCCATCATAATACGCGTTATAGCCATGATATTTCACAATCGCC...   \n",
       "13  Seq2  S_00003  CGCGTATTATGATGGCGAAATTCATGAAATGGTGAACTGGGCGTTT...   \n",
       "14  Seq2  S_00004  TCATCCACTTCATCAAAATAATAAATCACTTCAAACGCGCGCACCG...   \n",
       "15  Seq2  S_00005  TTATTTTGATGAAGTGGATGAAACCCGCATTGGCGTGACCGGCGGC...   \n",
       "16  Seq2  S_00006  AGGTCTGCATCGCCTGCACTTCGTTTTCCGGATAGCCGTTGCGGCG...   \n",
       "17  Seq2  S_00007  TGCAGGCGATGCAGACCTTTTATTATTTTGAAATTATGAACTTTGC...   \n",
       "18  Seq2  S_00008  GCCTTTAAAATGCGGTTTAAAAAACGGAAATTTATCGGTCTGAAAA...   \n",
       "19  Seq2  S_00009  TAAACCGCATTTTAAAGGCCTCGAGCACCACCACCACCACCACTGA...   \n",
       "\n",
       "    oligo_length   oligo_tm primer_overlap_with_previous  overlap_tm_5prime  \\\n",
       "0            119  76.754686                         None                NaN   \n",
       "1            128  83.441734           CGGGCGGCGCGCCGTTTG          68.658228   \n",
       "2            129  86.834758           GATGGCCAGCGCCATTAT          55.811457   \n",
       "3            128  85.380596          TGATGAATATACCTTTCGC          47.775584   \n",
       "4            128  88.288890           CCCCGGCGTATCCGGGCC          65.660649   \n",
       "5            128  85.057453           GCGGCGCAGGCGCCGTTT          68.094164   \n",
       "6            128  86.026884           TGCGGCGGCGTATCCGTA          62.137239   \n",
       "7            128  85.057453           GCTATGCGGAACCGGCGT          61.126936   \n",
       "8             78  81.265500           CAGCCGTTTGATGGCTTT          54.166830   \n",
       "9             73  80.600609           GCGCGCCCGGCGGTGGAA          69.523553   \n",
       "10           120  71.840745                         None                NaN   \n",
       "11           129  81.704538          CCGCGCCGAAAGAATTTTA          54.887527   \n",
       "12           129  82.666454          GTGTATCGCGTGACCTATA          51.807772   \n",
       "13           128  80.856584          CGCGTATTATGATGGCGAA          53.868963   \n",
       "14           130  79.675896           ATCCGCATGGCCATGCGT          60.798906   \n",
       "15           131  83.045865        TTATTTTGATGAAGTGGATGA          48.512171   \n",
       "16           130  80.630413       GCGGATTATCCGTATTTTTATA          49.832919   \n",
       "17           129  79.780705           TGCAGGCGATGCAGACCT          59.627099   \n",
       "18           129  76.894956          ATTGTGTTTGCGGCGTATA          53.466680   \n",
       "19            65  72.196960          TAAACCGCATTTTAAAGGC          49.857060   \n",
       "\n",
       "    overlap_length  overlap_homodimer_tm  overlap_hairpin_tm  \\\n",
       "0              NaN                   NaN                 NaN   \n",
       "1             18.0             56.011620           68.396815   \n",
       "2             18.0             17.815718           68.998303   \n",
       "3             19.0            -64.316560            0.000000   \n",
       "4             18.0             30.462576           75.699579   \n",
       "5             18.0             53.361828           83.411766   \n",
       "6             18.0             32.349272           58.411563   \n",
       "7             18.0             15.010537           40.188310   \n",
       "8             18.0             10.864156           58.661393   \n",
       "9             18.0             31.975298           63.355769   \n",
       "10             NaN                   NaN                 NaN   \n",
       "11            19.0             -2.987647           39.271088   \n",
       "12            19.0              5.768584            0.000000   \n",
       "13            19.0             10.800809           49.069751   \n",
       "14            18.0             57.093991           67.117879   \n",
       "15            21.0           -138.759018            0.000000   \n",
       "16            22.0             30.590793           59.177368   \n",
       "17            18.0             10.850945           59.622681   \n",
       "18            19.0              9.367922            0.000000   \n",
       "19            19.0              4.109652            0.000000   \n",
       "\n",
       "                                    original_sequence  \n",
       "0   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "1   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "2   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "3   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "4   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "5   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "6   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "7   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "8   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "9   cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "10  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "11  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "12  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "13  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "14  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "15  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "16  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "17  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "18  cccctctagaaataattttgtttaactttaagaaggagatatacat...  \n",
       "19  cccctctagaaataattttgtttaactttaagaaggagatatacat...  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "df = pd.DataFrame()\n",
    "df['id'] = ['Seq1', 'Seq2']\n",
    "min_gc = 0.25\n",
    "max_gc = 0.75\n",
    "min_tm = 10\n",
    "max_tm = 100\n",
    "min_segment_length = 40\n",
    "max_segment_length = 100\n",
    "max_length = 500\n",
    "\n",
    "df['seqs'] = ['MRTSLLVAALGLALAAALPGGAPLAQPDPEATMDRSLLQRQDLPYRFSAVDLDSVDGQRHYRLWLGRPLQAPPAAGYPVVWMLDGNAAVGALDESTLRRLADGDAPLLVAIGYRTPLRIDRAGRTFDYTPASPGQADQRDPLNGLPSGGADAFLDLLRDGMRPAVAAQAPLDTARQTLWGHSYGGLLVLHALFTRPGEFARYAAASPSLWWRDGAILGERAGLEQRLRGKRAELLLWRGSAEPASPRGSLKAEPGQAMARLVDDLRRVAGLTLDFQPLDGLGHGETLGASLRLLLARPAVERQR', \n",
    "             'MQLFDLPLDQLQTYKPEKTAPKDFSEFWKLSLEELAKVQAEPDLQPVDYPADGVKVYRLTYKSFGNARITGWYAVPDKEGPHPAIVKYHGYNASYDGEIHEMVNWALHGYATFGMLVRGQQSSEDTSISPHGHALGWMTKGILDKDTYYYRGVYLDAVRALEVISSFDEVDETRIGVTGGSQGGGLTIAAAALSDIPKAAVADYPYLSNFERAIDVALEQPYLEINSFFRRNGSPETEVQAMKTLSYFDIMNLADRVKVPVLMSIGLIDKVTPPSTVFAAYNHLETKKELKVYRYFGHEYIPAFQTEKLAFFKQHLKG']\n",
    "oligo_df = get_oligos(df, 'seqs', 'id', '.', 'cccctctagaaataattttgtttaactttaagaaggagatatacat', 'CTCGAGCACCACCACCACCACCACTGAgatccggctgctaacaaag', \n",
    "                     min_gc=min_gc, max_gc=max_gc, min_tm=min_tm, max_tm=max_tm, min_segment_length=min_segment_length, max_segment_length=max_segment_length)\n",
    "oligo_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
